{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "IE5fRRyeQ3Ev"
      ],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "e70369c20e234db89d689cffe041f0a0": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_dd48b2fd8d454382ab9265d877ab46f2",
              "IPY_MODEL_03e88a83f8524f81b7a1c8c861bd4bda",
              "IPY_MODEL_34783c6caaec4b8aa8b3d593d17acd3e"
            ],
            "layout": "IPY_MODEL_82eac9f4076d4315af4a3fb0d91dd964"
          }
        },
        "dd48b2fd8d454382ab9265d877ab46f2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8d9364e77cea421e8dcb593f35cd4c33",
            "placeholder": "​",
            "style": "IPY_MODEL_5c39dface15f4d4997d9c3b31c007317",
            "value": "config.json: 100%"
          }
        },
        "03e88a83f8524f81b7a1c8c861bd4bda": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_4c5e36df968847ae954cd2c78d36eb51",
            "max": 819,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_7d10fa9e3b1e475ba75b6fe71ff690b1",
            "value": 819
          }
        },
        "34783c6caaec4b8aa8b3d593d17acd3e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_667fa9f238df4ea99ea313ce7e1502f0",
            "placeholder": "​",
            "style": "IPY_MODEL_a4163cb68f7745c2a78685cea7dd553b",
            "value": " 819/819 [00:00&lt;00:00, 54.7kB/s]"
          }
        },
        "82eac9f4076d4315af4a3fb0d91dd964": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "8d9364e77cea421e8dcb593f35cd4c33": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "5c39dface15f4d4997d9c3b31c007317": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "4c5e36df968847ae954cd2c78d36eb51": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7d10fa9e3b1e475ba75b6fe71ff690b1": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "667fa9f238df4ea99ea313ce7e1502f0": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "a4163cb68f7745c2a78685cea7dd553b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "dcd4e1809a974cd3885182ed662acba2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_d47a3d5ed1ab41e4b303e3fe6e4d171d",
              "IPY_MODEL_35172f54aaec4db3868f0e9814aaef37",
              "IPY_MODEL_a3fc6f48aa704b169b9bfabab29337ce"
            ],
            "layout": "IPY_MODEL_074c5d5bc4fc4392b7f04ba6259ca945"
          }
        },
        "d47a3d5ed1ab41e4b303e3fe6e4d171d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8faa46b5ce4d44b485942726494e8574",
            "placeholder": "​",
            "style": "IPY_MODEL_bcfb44e9dd884d9da684efb59e90da74",
            "value": "tokenizer_config.json: 100%"
          }
        },
        "35172f54aaec4db3868f0e9814aaef37": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_17b2d4ed10a24e059632177aab99a131",
            "max": 393,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_ea675cdc27c94da3b4bfefd0e27fdec8",
            "value": 393
          }
        },
        "a3fc6f48aa704b169b9bfabab29337ce": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c71f4bc2e3e04d90b3163eed51d00aee",
            "placeholder": "​",
            "style": "IPY_MODEL_9b7f31b9016246e885c9cc6b192e7c9b",
            "value": " 393/393 [00:00&lt;00:00, 19.9kB/s]"
          }
        },
        "074c5d5bc4fc4392b7f04ba6259ca945": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "8faa46b5ce4d44b485942726494e8574": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "bcfb44e9dd884d9da684efb59e90da74": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "17b2d4ed10a24e059632177aab99a131": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ea675cdc27c94da3b4bfefd0e27fdec8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "c71f4bc2e3e04d90b3163eed51d00aee": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9b7f31b9016246e885c9cc6b192e7c9b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "1d9829ee81914aa18a21b3d41a0946ef": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_03a73a278ddd414dbd9303199980df33",
              "IPY_MODEL_243d7dbd7e9d4a3bad42e0c8fb2f0c05",
              "IPY_MODEL_4db46cc9d6de4d32b0284b9f80cdeebf"
            ],
            "layout": "IPY_MODEL_6637927afa7c435ea780e5db8305ea71"
          }
        },
        "03a73a278ddd414dbd9303199980df33": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_69e0f2da64a043a980c65b8c32461390",
            "placeholder": "​",
            "style": "IPY_MODEL_52e6d2770aad43c09cfc1b9dce7f7678",
            "value": "vocab.json: 100%"
          }
        },
        "243d7dbd7e9d4a3bad42e0c8fb2f0c05": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_acedaae9ec844c7792e949623c0895ac",
            "max": 798293,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_9ced533e44af4274bcb3a32a3d3c4807",
            "value": 798293
          }
        },
        "4db46cc9d6de4d32b0284b9f80cdeebf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_31110f637aa949afb9a2f3f6c57659ac",
            "placeholder": "​",
            "style": "IPY_MODEL_26bbc17621f94d36a0c0230617afd7cd",
            "value": " 798k/798k [00:00&lt;00:00, 4.01MB/s]"
          }
        },
        "6637927afa7c435ea780e5db8305ea71": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "69e0f2da64a043a980c65b8c32461390": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "52e6d2770aad43c09cfc1b9dce7f7678": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "acedaae9ec844c7792e949623c0895ac": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9ced533e44af4274bcb3a32a3d3c4807": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "31110f637aa949afb9a2f3f6c57659ac": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "26bbc17621f94d36a0c0230617afd7cd": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "39afc6cba7334315aaa755339ac48d65": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_ce83f3a10fd04838af005705642525e0",
              "IPY_MODEL_4f483615ce61442daf8160d21ed7e1b8",
              "IPY_MODEL_953016b0202b4ac5b5f22d527664b816"
            ],
            "layout": "IPY_MODEL_43dc123a1c684e50bd2001ddbddba85f"
          }
        },
        "ce83f3a10fd04838af005705642525e0": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_34cfc8aa6504404988d89001fb99493b",
            "placeholder": "​",
            "style": "IPY_MODEL_5106b67b29ff4ff5a9d059bd59a7a211",
            "value": "merges.txt: 100%"
          }
        },
        "4f483615ce61442daf8160d21ed7e1b8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_dd67ceebb3f646d7b687b2a29cf4fe1c",
            "max": 456356,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_39566b9586034d8e82315e5cb91752f4",
            "value": 456356
          }
        },
        "953016b0202b4ac5b5f22d527664b816": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_2a02f05fb26348e69288205500975af3",
            "placeholder": "​",
            "style": "IPY_MODEL_cbfafcd3351148abbbf7cd296e978949",
            "value": " 456k/456k [00:00&lt;00:00, 3.45MB/s]"
          }
        },
        "43dc123a1c684e50bd2001ddbddba85f": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "34cfc8aa6504404988d89001fb99493b": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "5106b67b29ff4ff5a9d059bd59a7a211": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "dd67ceebb3f646d7b687b2a29cf4fe1c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "39566b9586034d8e82315e5cb91752f4": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "2a02f05fb26348e69288205500975af3": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "cbfafcd3351148abbbf7cd296e978949": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "48faaf67784e48ba99992d7f27458038": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_97051258460241b8a6a5bfa7c8dde1d8",
              "IPY_MODEL_865540c2c7a9435ab0a007a747517f29",
              "IPY_MODEL_a80dca3724d1410698489d42f973a663"
            ],
            "layout": "IPY_MODEL_19950f33eb614d94b1009719dc583fd8"
          }
        },
        "97051258460241b8a6a5bfa7c8dde1d8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_33b1a558628e40e7bdbc12c89b195ef3",
            "placeholder": "​",
            "style": "IPY_MODEL_3e6fcacfa83845d48a463045dac39d5b",
            "value": "tokenizer.json: 100%"
          }
        },
        "865540c2c7a9435ab0a007a747517f29": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_e00d81391e71413181c1346eddb36b88",
            "max": 1355880,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_b2d1b3f6280945c99a828c73fcf42d3c",
            "value": 1355880
          }
        },
        "a80dca3724d1410698489d42f973a663": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_ed783ab8385f4a4cac986d9c0513cd1d",
            "placeholder": "​",
            "style": "IPY_MODEL_1a0c802dadb64a4fac52e52850f7dd69",
            "value": " 1.36M/1.36M [00:00&lt;00:00, 6.85MB/s]"
          }
        },
        "19950f33eb614d94b1009719dc583fd8": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "33b1a558628e40e7bdbc12c89b195ef3": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "3e6fcacfa83845d48a463045dac39d5b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "e00d81391e71413181c1346eddb36b88": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b2d1b3f6280945c99a828c73fcf42d3c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "ed783ab8385f4a4cac986d9c0513cd1d": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "1a0c802dadb64a4fac52e52850f7dd69": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "7bfb3c0bcb7845a8abf87498c4ba05e9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_786bb2a18db4436f9a0fe45d87c6a3ca",
              "IPY_MODEL_e8a3b7dc44b044e9bce7a58a9505688a",
              "IPY_MODEL_8fb159ad19c844bf9931d5052c2f2417"
            ],
            "layout": "IPY_MODEL_bbff96595c5a4c63a30a7e62966c13dd"
          }
        },
        "786bb2a18db4436f9a0fe45d87c6a3ca": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_91c1cc9cbb37438a8b90394f96063832",
            "placeholder": "​",
            "style": "IPY_MODEL_928123d454b74c7db0e10dc138c9f941",
            "value": "special_tokens_map.json: 100%"
          }
        },
        "e8a3b7dc44b044e9bce7a58a9505688a": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_4b5aa97b4f284234afad0b6fd4bc0b4c",
            "max": 239,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_128be60f86fe4c8091d2899233cf9385",
            "value": 239
          }
        },
        "8fb159ad19c844bf9931d5052c2f2417": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_a1cc45f72a30477ab5c3df6291e5f17c",
            "placeholder": "​",
            "style": "IPY_MODEL_b5853da605e04f9e8cb9a733d89af339",
            "value": " 239/239 [00:00&lt;00:00, 17.5kB/s]"
          }
        },
        "bbff96595c5a4c63a30a7e62966c13dd": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "91c1cc9cbb37438a8b90394f96063832": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "928123d454b74c7db0e10dc138c9f941": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "4b5aa97b4f284234afad0b6fd4bc0b4c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "128be60f86fe4c8091d2899233cf9385": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "a1cc45f72a30477ab5c3df6291e5f17c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b5853da605e04f9e8cb9a733d89af339": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "f13d96e5a33a405ea76281e2719604d7": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_05b46caac1d04feeb95141b776ff3252",
              "IPY_MODEL_6374b19ff00643798a41af827de1bd81",
              "IPY_MODEL_bbd5c25aa97e4b7087eff6a10575eaeb"
            ],
            "layout": "IPY_MODEL_b0abc178ed634a258debd4e1be513796"
          }
        },
        "05b46caac1d04feeb95141b776ff3252": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_804aabe0bab54d599d88a90eaeb0bd48",
            "placeholder": "​",
            "style": "IPY_MODEL_72d3cebf6b3848758099589c93a6ff1b",
            "value": "pytorch_model.bin: 100%"
          }
        },
        "6374b19ff00643798a41af827de1bd81": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_1eb15955e54e434db039bccd83f3fb8a",
            "max": 362162101,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_69c982cb01924a6a8913e0143f398d95",
            "value": 362162101
          }
        },
        "bbd5c25aa97e4b7087eff6a10575eaeb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c9c5245b3e93465aa24bb699d9813945",
            "placeholder": "​",
            "style": "IPY_MODEL_0c7f7800d0d14750bed72419b8a2cb89",
            "value": " 362M/362M [00:01&lt;00:00, 214MB/s]"
          }
        },
        "b0abc178ed634a258debd4e1be513796": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "804aabe0bab54d599d88a90eaeb0bd48": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "72d3cebf6b3848758099589c93a6ff1b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "1eb15955e54e434db039bccd83f3fb8a": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "69c982cb01924a6a8913e0143f398d95": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "c9c5245b3e93465aa24bb699d9813945": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "0c7f7800d0d14750bed72419b8a2cb89": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "78f7e9faeb354d41b5097c1e906d8926": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_1cc3431de0da4a17b6150b434d898c9e",
              "IPY_MODEL_b770309788914ef0bcdb6a58c99618c1",
              "IPY_MODEL_fa45d2ca6653413a81eeae3693700f52"
            ],
            "layout": "IPY_MODEL_e49623ff75bf497ca4985618e1262196"
          }
        },
        "1cc3431de0da4a17b6150b434d898c9e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_1798e88119d74ad6bf071fa849858c14",
            "placeholder": "​",
            "style": "IPY_MODEL_fe4e6ab40eb54099bc062472fd98ef22",
            "value": "config.json: 100%"
          }
        },
        "b770309788914ef0bcdb6a58c99618c1": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8fa9181848f740c3b2c49b3320846937",
            "max": 819,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_818f7c9f135e43d884635af9eeec31a8",
            "value": 819
          }
        },
        "fa45d2ca6653413a81eeae3693700f52": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_0f58231c0a9c4ab0b0613da8d8fc760e",
            "placeholder": "​",
            "style": "IPY_MODEL_a03d4879da2d4c5fae644e7ebb9f248b",
            "value": " 819/819 [00:00&lt;00:00, 56.3kB/s]"
          }
        },
        "e49623ff75bf497ca4985618e1262196": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "1798e88119d74ad6bf071fa849858c14": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "fe4e6ab40eb54099bc062472fd98ef22": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8fa9181848f740c3b2c49b3320846937": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "818f7c9f135e43d884635af9eeec31a8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "0f58231c0a9c4ab0b0613da8d8fc760e": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "a03d4879da2d4c5fae644e7ebb9f248b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "b59f144ff36c43e9a738900e9173f08b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_dfb71ef971ef4ae3af14321be9283b56",
              "IPY_MODEL_e09ab65f039d4d7a91575c409021fc61",
              "IPY_MODEL_0257cd89fa3f45129c9c7bfa184f2049"
            ],
            "layout": "IPY_MODEL_2d419d6719844396bcac4000f4dcb138"
          }
        },
        "dfb71ef971ef4ae3af14321be9283b56": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_7124f8c491f24363ae5fdd3364dc8a07",
            "placeholder": "​",
            "style": "IPY_MODEL_7cb630959f714cd088e202c7ed901e55",
            "value": "tokenizer_config.json: 100%"
          }
        },
        "e09ab65f039d4d7a91575c409021fc61": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_3818cd316f5040c9b8d522aba4e50a89",
            "max": 393,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_3703030778574de288ae5073475281cf",
            "value": 393
          }
        },
        "0257cd89fa3f45129c9c7bfa184f2049": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_42a944d62949408c943ddb0e234ab6a2",
            "placeholder": "​",
            "style": "IPY_MODEL_7145b20514bd40cdbbbb82109c67b1bf",
            "value": " 393/393 [00:00&lt;00:00, 27.0kB/s]"
          }
        },
        "2d419d6719844396bcac4000f4dcb138": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7124f8c491f24363ae5fdd3364dc8a07": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7cb630959f714cd088e202c7ed901e55": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "3818cd316f5040c9b8d522aba4e50a89": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "3703030778574de288ae5073475281cf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "42a944d62949408c943ddb0e234ab6a2": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7145b20514bd40cdbbbb82109c67b1bf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "e8e48ed233964d758c79f2241f2af3f4": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_f9fe372bcad345d19620c2a3862b589a",
              "IPY_MODEL_dcbfda2f8d204427a2ff0ee756d9c702",
              "IPY_MODEL_59600d0699bd43678f98a5274e0cb62c"
            ],
            "layout": "IPY_MODEL_501fefbf21ac4d44bcf7e17674a1c566"
          }
        },
        "f9fe372bcad345d19620c2a3862b589a": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_e47be1b5b9ba4b9e915f77945c051732",
            "placeholder": "​",
            "style": "IPY_MODEL_468bb1c0ecbd49bab8de977cd6c7f597",
            "value": "vocab.json: 100%"
          }
        },
        "dcbfda2f8d204427a2ff0ee756d9c702": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_cc69579b708049d38a1f4ea317ae7983",
            "max": 798293,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_68f8924445224b9ebe0986cee669839d",
            "value": 798293
          }
        },
        "59600d0699bd43678f98a5274e0cb62c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_4372c458d49047ae96859be793763031",
            "placeholder": "​",
            "style": "IPY_MODEL_3758484c358f444382629747ea7f998e",
            "value": " 798k/798k [00:00&lt;00:00, 9.39MB/s]"
          }
        },
        "501fefbf21ac4d44bcf7e17674a1c566": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "e47be1b5b9ba4b9e915f77945c051732": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "468bb1c0ecbd49bab8de977cd6c7f597": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "cc69579b708049d38a1f4ea317ae7983": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "68f8924445224b9ebe0986cee669839d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "4372c458d49047ae96859be793763031": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "3758484c358f444382629747ea7f998e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "ed35a3babab049d6b4153c224b0475c3": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_4cb48f81521a4047b8a4fd00cefadcc5",
              "IPY_MODEL_ac763f13a48d46028f989fba83402b22",
              "IPY_MODEL_6635573513fa4ef0ad38840dea43c066"
            ],
            "layout": "IPY_MODEL_fe203bbf3df4445eb657d22b4507fb41"
          }
        },
        "4cb48f81521a4047b8a4fd00cefadcc5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_75b45012a2b14af98e69ecbbf0bf6a6f",
            "placeholder": "​",
            "style": "IPY_MODEL_d54c7d9641904136aa44c43cee575fb4",
            "value": "merges.txt: 100%"
          }
        },
        "ac763f13a48d46028f989fba83402b22": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_5ae23f3db17445e19d562d8f09ceae41",
            "max": 456356,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_c939eb235a104687bf70c140359fa7cc",
            "value": 456356
          }
        },
        "6635573513fa4ef0ad38840dea43c066": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_7673334c0d3c417888fabcfa9f55aedf",
            "placeholder": "​",
            "style": "IPY_MODEL_5b938c57fb364adda61d8e2482238de5",
            "value": " 456k/456k [00:00&lt;00:00, 3.27MB/s]"
          }
        },
        "fe203bbf3df4445eb657d22b4507fb41": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "75b45012a2b14af98e69ecbbf0bf6a6f": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d54c7d9641904136aa44c43cee575fb4": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "5ae23f3db17445e19d562d8f09ceae41": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c939eb235a104687bf70c140359fa7cc": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "7673334c0d3c417888fabcfa9f55aedf": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "5b938c57fb364adda61d8e2482238de5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "e5d458db2aac4a11a13f1c934b1e8dcf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_8cd4bca9d48b4f8f98369a2dd5adf17b",
              "IPY_MODEL_a66e9995344648e5b77b0674156cc2fb",
              "IPY_MODEL_3a1a1947fdd54457a59da862268ba6fd"
            ],
            "layout": "IPY_MODEL_3fd30ba8c8534b5dbc29c7396597be95"
          }
        },
        "8cd4bca9d48b4f8f98369a2dd5adf17b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_ea28a193d2024bff8463cbb11d09d755",
            "placeholder": "​",
            "style": "IPY_MODEL_7587cd9037644ccb8925ee0524857f21",
            "value": "tokenizer.json: 100%"
          }
        },
        "a66e9995344648e5b77b0674156cc2fb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8aedcb205cb14c5ebc878e4c2689a10c",
            "max": 1355880,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_f5cc2a5959464e0fb72e3b420a58e5d9",
            "value": 1355880
          }
        },
        "3a1a1947fdd54457a59da862268ba6fd": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_21906dc6b3434bf29f62460890ec6288",
            "placeholder": "​",
            "style": "IPY_MODEL_582f2ef1a2714f59912cd0bcdbf1bb75",
            "value": " 1.36M/1.36M [00:00&lt;00:00, 5.07MB/s]"
          }
        },
        "3fd30ba8c8534b5dbc29c7396597be95": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ea28a193d2024bff8463cbb11d09d755": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7587cd9037644ccb8925ee0524857f21": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8aedcb205cb14c5ebc878e4c2689a10c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "f5cc2a5959464e0fb72e3b420a58e5d9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "21906dc6b3434bf29f62460890ec6288": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "582f2ef1a2714f59912cd0bcdbf1bb75": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d176bbfd2d464d7585bb65be65c31262": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_a94a3cec76fe47748a9a4db32b924cfa",
              "IPY_MODEL_2817e9a3ac6447e8981188717e927c9d",
              "IPY_MODEL_0141e4bc61c247c680fa4c05b3d04903"
            ],
            "layout": "IPY_MODEL_ad5edbd6ee20493f9a8e0833cfac7db5"
          }
        },
        "a94a3cec76fe47748a9a4db32b924cfa": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_534c09960afd40b0b3ccf4ec620aba51",
            "placeholder": "​",
            "style": "IPY_MODEL_d7ebd277b99143fdad80e1f4473d73fb",
            "value": "special_tokens_map.json: 100%"
          }
        },
        "2817e9a3ac6447e8981188717e927c9d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c06916c9e574464ebaa45d54b3aa2c92",
            "max": 239,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_9f900c0c253d461eaeeaede238093346",
            "value": 239
          }
        },
        "0141e4bc61c247c680fa4c05b3d04903": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_01ff09afb1af40bbb72a62826b81fc64",
            "placeholder": "​",
            "style": "IPY_MODEL_ef1e840a9f3d4a048c95c3865a5da922",
            "value": " 239/239 [00:00&lt;00:00, 9.17kB/s]"
          }
        },
        "ad5edbd6ee20493f9a8e0833cfac7db5": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "534c09960afd40b0b3ccf4ec620aba51": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d7ebd277b99143fdad80e1f4473d73fb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "c06916c9e574464ebaa45d54b3aa2c92": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9f900c0c253d461eaeeaede238093346": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "01ff09afb1af40bbb72a62826b81fc64": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ef1e840a9f3d4a048c95c3865a5da922": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "c4d57890e664453aa90a45c4bbb1d17f": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_b4e8802f985e443caee7f82f9e1d642f",
              "IPY_MODEL_e1bd9242da8043cda8a465415842d182",
              "IPY_MODEL_2013b5dea4794109b5d4e6932a3166a9"
            ],
            "layout": "IPY_MODEL_6a954028886f40c39ebbad3d33e81036"
          }
        },
        "b4e8802f985e443caee7f82f9e1d642f": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_22ffba8398a048e98c2628e2770f3141",
            "placeholder": "​",
            "style": "IPY_MODEL_5ce8a4e265374bdc87de7dc7ff0a6613",
            "value": "pytorch_model.bin: 100%"
          }
        },
        "e1bd9242da8043cda8a465415842d182": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d0a34c02e88a4d3fb2841e7fc696d5a0",
            "max": 362162101,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_cd389c7bca4a4ad39dce662ffb8e2300",
            "value": 362162101
          }
        },
        "2013b5dea4794109b5d4e6932a3166a9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_02186d45c99a463e82a3439bbd1b535a",
            "placeholder": "​",
            "style": "IPY_MODEL_41626856c41f40d58cee4207bdf9aee1",
            "value": " 362M/362M [00:02&lt;00:00, 207MB/s]"
          }
        },
        "6a954028886f40c39ebbad3d33e81036": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "22ffba8398a048e98c2628e2770f3141": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "5ce8a4e265374bdc87de7dc7ff0a6613": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d0a34c02e88a4d3fb2841e7fc696d5a0": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "cd389c7bca4a4ad39dce662ffb8e2300": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "02186d45c99a463e82a3439bbd1b535a": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "41626856c41f40d58cee4207bdf9aee1": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Abhijith-Nagarajan/CS_546_Project/blob/Abhijith_NER_TM/notebooks/RE_Inference.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lhtiu0g2UOBa"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XcEIIxukUp9M",
        "outputId": "d1d74032-af68-4af0-84e5-cd27653b79bd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n"
          ]
        }
      ],
      "source": [
        "drive.mount('/content/gdrive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bUvZtjtFUrmw"
      },
      "outputs": [],
      "source": [
        "package_directory = r'/content/gdrive/MyDrive/CS_546_Project_Dependencies/lib/python3.10/site-packages'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qhLl5vG_c5Ss"
      },
      "outputs": [],
      "source": [
        "import sys"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sys.path.append(package_directory)"
      ],
      "metadata": {
        "id": "FCa9x9HIzRvT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os"
      ],
      "metadata": {
        "id": "rQpVINA2ihTI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import copy"
      ],
      "metadata": {
        "id": "E9ovbxkaqTub"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WDNDUoTVVGUn"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import spacy\n",
        "from spacy import displacy\n",
        "import en_core_sci_md\n",
        "import re\n",
        "import torch"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "nb9Jkd0CbcM6"
      },
      "outputs": [],
      "source": [
        "from datasets import load_dataset, Dataset"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qPyHiOxcq82F"
      },
      "outputs": [],
      "source": [
        "from transformers import AutoTokenizer, AutoModelForTokenClassification, pipeline, Seq2SeqTrainingArguments, Seq2SeqTrainer, AutoModelForSeq2SeqLM, AutoModelForCausalLM, AutoModel, TrainingArguments, Trainer, IntervalStrategy, EarlyStoppingCallback"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "model_path = \"/content/gdrive/MyDrive/CS_546_Project_Dependencies/BioRED_Results/checkpoint-118\""
      ],
      "metadata": {
        "id": "PtEdXbJAjdSx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_model = AutoModelForSeq2SeqLM.from_pretrained(model_path)"
      ],
      "metadata": {
        "id": "xXkDQVFDHBqG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "device = \"cuda:0\" if torch.cuda.is_available() else \"cpu\""
      ],
      "metadata": {
        "id": "W9mw-t9jC0fW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_model.to(device)"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tYcXyA3o0SJb",
        "outputId": "729786d6-073e-47ef-e016-a4654abce5d3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "BartForConditionalGeneration(\n",
              "  (model): BartModel(\n",
              "    (shared): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "    (encoder): BartEncoder(\n",
              "      (embed_tokens): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024)\n",
              "      (layers): ModuleList(\n",
              "        (0-11): 12 x BartEncoderLayer(\n",
              "          (self_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (activation_fn): GELUActivation()\n",
              "          (fc1): Linear(in_features=1024, out_features=4096, bias=True)\n",
              "          (fc2): Linear(in_features=4096, out_features=1024, bias=True)\n",
              "          (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "        )\n",
              "      )\n",
              "      (layernorm_embedding): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "    )\n",
              "    (decoder): BartDecoder(\n",
              "      (embed_tokens): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024)\n",
              "      (layers): ModuleList(\n",
              "        (0-11): 12 x BartDecoderLayer(\n",
              "          (self_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (activation_fn): GELUActivation()\n",
              "          (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (encoder_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (encoder_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (fc1): Linear(in_features=1024, out_features=4096, bias=True)\n",
              "          (fc2): Linear(in_features=4096, out_features=1024, bias=True)\n",
              "          (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "        )\n",
              "      )\n",
              "      (layernorm_embedding): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "    )\n",
              "  )\n",
              "  (lm_head): Linear(in_features=1024, out_features=50272, bias=False)\n",
              ")"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_tokenizer = AutoTokenizer.from_pretrained(model_path)"
      ],
      "metadata": {
        "id": "-s2FZ5d0HBmk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_data = pd.read_json(\"/content/research_text_pubmed_new.json\")"
      ],
      "metadata": {
        "id": "-HSKiVkXkWCi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_data.head()"
      ],
      "metadata": {
        "id": "ZXP0BxnEs7wr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_data.shape"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "A6lYYXBJs9F5",
        "outputId": "1c2a7c23-d014-4988-98ee-6fe610ec9819"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(103, 473)"
            ]
          },
          "metadata": {},
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered = pubmed_inference_data[['title','abstract']]"
      ],
      "metadata": {
        "id": "czK9uk-ptEcm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "NKYY9fHbtJ4U",
        "outputId": "08fad130-d553-4882-a67b-41107fb0ea62"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                 title  \\\n",
              "0    Genetic and phenotypic diversity of NHE6 mutat...   \n",
              "1    LSD1-mediated epigenetic modification is requi...   \n",
              "2    TFBSshape: a motif database for DNA shape feat...   \n",
              "3    Histone H3K27ac separates active from poised e...   \n",
              "4    Novel C12orf65mutations in patients with axona...   \n",
              "..                                                 ...   \n",
              "98   MARS: improving multiple circular sequence ali...   \n",
              "99   JACUSA: site-specific identification of RNA ed...   \n",
              "100  RIblast: an ultrafast RNA–RNA interaction pred...   \n",
              "101  TIRR regulates 53BP1 by masking its histone me...   \n",
              "102  CarbonylDB: a curated data-resource of protein...   \n",
              "\n",
              "                                              abstract  \n",
              "0                                          abstract     \n",
              "1    abstract   TAL1 is a critical transcription fa...  \n",
              "2    abstract   Transcription factor binding sites ...  \n",
              "3    abstract   Developmental programs are controll...  \n",
              "4                                          abstract     \n",
              "..                                                 ...  \n",
              "98                                         abstract     \n",
              "99                                         abstract     \n",
              "100                                        abstract     \n",
              "101                                                NaN  \n",
              "102                                        abstract     \n",
              "\n",
              "[103 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5119c985-f944-49ba-9fad-1e4e66e677b4\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>title</th>\n",
              "      <th>abstract</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Genetic and phenotypic diversity of NHE6 mutat...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>LSD1-mediated epigenetic modification is requi...</td>\n",
              "      <td>abstract   TAL1 is a critical transcription fa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>TFBSshape: a motif database for DNA shape feat...</td>\n",
              "      <td>abstract   Transcription factor binding sites ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Histone H3K27ac separates active from poised e...</td>\n",
              "      <td>abstract   Developmental programs are controll...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Novel C12orf65mutations in patients with axona...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>MARS: improving multiple circular sequence ali...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>99</th>\n",
              "      <td>JACUSA: site-specific identification of RNA ed...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>RIblast: an ultrafast RNA–RNA interaction pred...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>101</th>\n",
              "      <td>TIRR regulates 53BP1 by masking its histone me...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>102</th>\n",
              "      <td>CarbonylDB: a curated data-resource of protein...</td>\n",
              "      <td>abstract</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>103 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5119c985-f944-49ba-9fad-1e4e66e677b4')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5119c985-f944-49ba-9fad-1e4e66e677b4 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5119c985-f944-49ba-9fad-1e4e66e677b4');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-45b63555-80a0-44ff-916a-4a74a6da5550\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-45b63555-80a0-44ff-916a-4a74a6da5550')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-45b63555-80a0-44ff-916a-4a74a6da5550 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_3848a2b2-b85c-47f2-acb0-02efc59e5a4e\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('pubmed_inference_filtered')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_3848a2b2-b85c-47f2-acb0-02efc59e5a4e button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('pubmed_inference_filtered');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "pubmed_inference_filtered",
              "summary": "{\n  \"name\": \"pubmed_inference_filtered\",\n  \"rows\": 103,\n  \"fields\": [\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 103,\n        \"samples\": [\n          \"Drosophila Ctf4 is essential for efficient DNA replication and normal cell cycle progression\",\n          \"Topologically associating domains are stable units of replication-timing regulation\",\n          \"Aramchol downregulates stearoyl CoA-desaturase 1 in hepatic stellate cells to attenuate cellular fibrogenesis\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"abstract\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 64,\n        \"samples\": [\n          \"abstract   Long noncoding RNAs (lncRNAs) are emerging as regulators of gene expression in pathogenesis, including cancer. Recently, lncRNAs have been implicated in progression of specific subtypes of breast cancer. One aggressive, basal\\u2010like subtype associates with increased EGFR signaling, while another, the HER2\\u2010enriched subtype, engages a kin of EGFR. Based on the premise that EGFR\\u2010regulated lncRNAs might control the aggressiveness of basal\\u2010like tumors, we identified multiple EGFR\\u2010inducible lncRNAs in basal\\u2010like normal cells and overlaid them with the transcriptomes of over 3,000 breast cancer patients. This led to the identification of 11 prognostic lncRNAs. Functional analyses of this group uncovered LINC01089 (here renamed LncRNA Inhibiting Metastasis; LIMT), a highly conserved lncRNA, which is depleted in basal\\u2010like and in HER2\\u2010positive tumors, and the low expression of which predicts poor patient prognosis. Interestingly, EGF rapidly downregulates LIMT expression by enhancing histone deacetylation at the respective promoter. We also find that LIMT inhibits extracellular matrix invasion of mammary cellsin\\u00a0vitroand tumor metastasisin\\u00a0vivo. In conclusion, lncRNAs dynamically regulated by growth factors might act as novel drivers of cancer progression and serve as prognostic biomarkers.\",\n          \"abstract   Hepadnaviral covalently closed circular DNA (cccDNA) exists as an episomal minichromosome in the nucleus of virus-infected hepatocytes, and serves as the transcriptional template for the synthesis of viral mRNAs. To obtain insight on the structure of hepadnaviral cccDNA minichromosomes, we utilized ducks infected with the duck hepatitis B virus (DHBV) as a model and determined thein vivonucleosome distribution pattern on viral cccDNA by the micrococcal nuclease (MNase) mapping and genome-wide PCR amplification of isolated mononucleosomal DHBV DNA. Several nucleosome-protected sites in a region of the DHBV genome [nucleotides (nt) 2000 to 2700], known to harbor variouscistranscription regulatory elements, were consistently identified in all DHBV-positive liver samples. In addition, we observed other nucleosome protection sites in DHBV minichromosomes that may vary among individual ducks, but the pattern of MNase mapping in those regions is transmittable from the adult ducks to the newly infected ducklings. These results imply that the nucleosomes along viral cccDNA in the minichromosomes are not random but sequence-specifically positioned. Furthermore, we showed in ducklings that a significant portion of cccDNA possesses a few negative superhelical turns, suggesting the presence of intermediates of viral minichromosomes assembled in the liver, where dynamic hepatocyte growth and cccDNA formation occur. This study supplies the initial framework for the understanding of the overall complete structure of hepadnaviral cccDNA minichromosomes.\",\n          \"abstract   \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 27
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GsEh-RVWydru"
      },
      "outputs": [],
      "source": [
        "def preprocess_text(text: str):\n",
        "  '''\n",
        "  The aim of this function is to preprocess text by converting it to lowercase, and punctuations.\n",
        "\n",
        "  This is not required.\n",
        "  '''\n",
        "  text = text.lower()\n",
        "  text = re.sub(r\",\\s?\",' ', text)\n",
        "  text = re.sub(r'\\.\\s*', ' ', text)\n",
        "  return text"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3xkDmoz9mqQT"
      },
      "outputs": [],
      "source": [
        "def filter_text(text:str):\n",
        "    try:\n",
        "        if type(text) != str:\n",
        "            return ''\n",
        "        else:\n",
        "            if 'abstract' not in text:\n",
        "                return text\n",
        "            else:\n",
        "                text = text.replace('abstract ', '')\n",
        "                return text\n",
        "    except:\n",
        "        print(f'Error text: {text}...{type(text)}')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "V8KWoR9xnJ_x",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "31fd7714-4389-41b5-d8a0-b7351c0baa88"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-31-3f59f4ba2418>:1: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  pubmed_inference_filtered['abstract_updated'] = pubmed_inference_filtered['abstract'].apply(lambda row: filter_text(row))\n"
          ]
        }
      ],
      "source": [
        "pubmed_inference_filtered['abstract_updated'] = pubmed_inference_filtered['abstract'].apply(lambda row: filter_text(row))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "njPGdcffXFTJ"
      },
      "outputs": [],
      "source": [
        "#pubmed_inference_filtered['abstract_updated'] = pubmed_inference_filtered['abstract_updated'].apply(lambda row: preprocess_text(row))"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered['abstract_updated'] = pubmed_inference_filtered['abstract_updated'].str.strip()"
      ],
      "metadata": {
        "id": "57E-mioZtw4k"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered['input'] = pubmed_inference_filtered['title']+\" \"+pubmed_inference_filtered['abstract_updated']"
      ],
      "metadata": {
        "id": "slV1WkgRuNeS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from spacy.tokens import Doc, Span\n",
        "from fastcoref import FCoref"
      ],
      "metadata": {
        "id": "IE28DFEuw8Fh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#nlp = spacy.load('en_core_web_sm')\n",
        "nlp_scispacy = spacy.load('en_core_sci_md')"
      ],
      "metadata": {
        "id": "Lw39edgW6HVz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fcoref_model = FCoref(device=device)"
      ],
      "metadata": {
        "id": "sHDduqo16HVz",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 400,
          "referenced_widgets": [
            "e70369c20e234db89d689cffe041f0a0",
            "dd48b2fd8d454382ab9265d877ab46f2",
            "03e88a83f8524f81b7a1c8c861bd4bda",
            "34783c6caaec4b8aa8b3d593d17acd3e",
            "82eac9f4076d4315af4a3fb0d91dd964",
            "8d9364e77cea421e8dcb593f35cd4c33",
            "5c39dface15f4d4997d9c3b31c007317",
            "4c5e36df968847ae954cd2c78d36eb51",
            "7d10fa9e3b1e475ba75b6fe71ff690b1",
            "667fa9f238df4ea99ea313ce7e1502f0",
            "a4163cb68f7745c2a78685cea7dd553b",
            "dcd4e1809a974cd3885182ed662acba2",
            "d47a3d5ed1ab41e4b303e3fe6e4d171d",
            "35172f54aaec4db3868f0e9814aaef37",
            "a3fc6f48aa704b169b9bfabab29337ce",
            "074c5d5bc4fc4392b7f04ba6259ca945",
            "8faa46b5ce4d44b485942726494e8574",
            "bcfb44e9dd884d9da684efb59e90da74",
            "17b2d4ed10a24e059632177aab99a131",
            "ea675cdc27c94da3b4bfefd0e27fdec8",
            "c71f4bc2e3e04d90b3163eed51d00aee",
            "9b7f31b9016246e885c9cc6b192e7c9b",
            "1d9829ee81914aa18a21b3d41a0946ef",
            "03a73a278ddd414dbd9303199980df33",
            "243d7dbd7e9d4a3bad42e0c8fb2f0c05",
            "4db46cc9d6de4d32b0284b9f80cdeebf",
            "6637927afa7c435ea780e5db8305ea71",
            "69e0f2da64a043a980c65b8c32461390",
            "52e6d2770aad43c09cfc1b9dce7f7678",
            "acedaae9ec844c7792e949623c0895ac",
            "9ced533e44af4274bcb3a32a3d3c4807",
            "31110f637aa949afb9a2f3f6c57659ac",
            "26bbc17621f94d36a0c0230617afd7cd",
            "39afc6cba7334315aaa755339ac48d65",
            "ce83f3a10fd04838af005705642525e0",
            "4f483615ce61442daf8160d21ed7e1b8",
            "953016b0202b4ac5b5f22d527664b816",
            "43dc123a1c684e50bd2001ddbddba85f",
            "34cfc8aa6504404988d89001fb99493b",
            "5106b67b29ff4ff5a9d059bd59a7a211",
            "dd67ceebb3f646d7b687b2a29cf4fe1c",
            "39566b9586034d8e82315e5cb91752f4",
            "2a02f05fb26348e69288205500975af3",
            "cbfafcd3351148abbbf7cd296e978949",
            "48faaf67784e48ba99992d7f27458038",
            "97051258460241b8a6a5bfa7c8dde1d8",
            "865540c2c7a9435ab0a007a747517f29",
            "a80dca3724d1410698489d42f973a663",
            "19950f33eb614d94b1009719dc583fd8",
            "33b1a558628e40e7bdbc12c89b195ef3",
            "3e6fcacfa83845d48a463045dac39d5b",
            "e00d81391e71413181c1346eddb36b88",
            "b2d1b3f6280945c99a828c73fcf42d3c",
            "ed783ab8385f4a4cac986d9c0513cd1d",
            "1a0c802dadb64a4fac52e52850f7dd69",
            "7bfb3c0bcb7845a8abf87498c4ba05e9",
            "786bb2a18db4436f9a0fe45d87c6a3ca",
            "e8a3b7dc44b044e9bce7a58a9505688a",
            "8fb159ad19c844bf9931d5052c2f2417",
            "bbff96595c5a4c63a30a7e62966c13dd",
            "91c1cc9cbb37438a8b90394f96063832",
            "928123d454b74c7db0e10dc138c9f941",
            "4b5aa97b4f284234afad0b6fd4bc0b4c",
            "128be60f86fe4c8091d2899233cf9385",
            "a1cc45f72a30477ab5c3df6291e5f17c",
            "b5853da605e04f9e8cb9a733d89af339",
            "f13d96e5a33a405ea76281e2719604d7",
            "05b46caac1d04feeb95141b776ff3252",
            "6374b19ff00643798a41af827de1bd81",
            "bbd5c25aa97e4b7087eff6a10575eaeb",
            "b0abc178ed634a258debd4e1be513796",
            "804aabe0bab54d599d88a90eaeb0bd48",
            "72d3cebf6b3848758099589c93a6ff1b",
            "1eb15955e54e434db039bccd83f3fb8a",
            "69c982cb01924a6a8913e0143f398d95",
            "c9c5245b3e93465aa24bb699d9813945",
            "0c7f7800d0d14750bed72419b8a2cb89"
          ]
        },
        "outputId": "4beecee2-cdfa-4b44-d0c2-a67dd95be7bc",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "config.json:   0%|          | 0.00/819 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "e70369c20e234db89d689cffe041f0a0"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer_config.json:   0%|          | 0.00/393 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "dcd4e1809a974cd3885182ed662acba2"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "1d9829ee81914aa18a21b3d41a0946ef"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "39afc6cba7334315aaa755339ac48d65"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "48faaf67784e48ba99992d7f27458038"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "7bfb3c0bcb7845a8abf87498c4ba05e9"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/spacy/util.py:1740: UserWarning: [W111] Jupyter notebook detected: if using `prefer_gpu()` or `require_gpu()`, include it in the same cell right before `spacy.load()` to ensure that the model is loaded on the correct device. More information: http://spacy.io/usage/v3#jupyter-notebook-gpu\n",
            "  warnings.warn(Warnings.W111)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "pytorch_model.bin:   0%|          | 0.00/362M [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "f13d96e5a33a405ea76281e2719604d7"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def core_logic_part(document: Doc, coref: list[int], resolved: list[str], mention_span: Span):\n",
        "    final_token = document[coref[1]]\n",
        "    if final_token.tag_ in [\"PRP$\", \"POS\"]:\n",
        "        resolved[coref[0]] = mention_span.text + \"'s\" + final_token.whitespace_\n",
        "    else:\n",
        "        resolved[coref[0]] = mention_span.text + final_token.whitespace_\n",
        "    for i in range(coref[0] + 1, coref[1] + 1):\n",
        "        resolved[i] = \"\"\n",
        "    return resolved"
      ],
      "metadata": {
        "id": "jiD9P6040oho"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_span_noun_indices(doc: Doc, cluster: list[list[int]]) -> list[int]:\n",
        "    try:\n",
        "      spans = [doc[span[0]:span[1]+1] for span in cluster]\n",
        "      spans_pos = [[token.pos_ for token in span] for span in spans]\n",
        "      span_noun_indices = [i for i, span_pos in enumerate(spans_pos)\n",
        "          if any(pos in span_pos for pos in ['NOUN', 'PROPN'])]\n",
        "      return span_noun_indices\n",
        "    except Exception as e:\n",
        "      print(f'Error received while getting span noun indices. {e}. Returning empty list')\n",
        "      return []"
      ],
      "metadata": {
        "id": "CD7ExESjxqE6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_cluster_head(doc: Doc, cluster: list[list[int]], noun_indices: list[int]):\n",
        "    head_idx = noun_indices[0]\n",
        "    head_start, head_end = cluster[head_idx]\n",
        "    head_span = doc[head_start:head_end+1]\n",
        "    return head_span, [head_start, head_end]"
      ],
      "metadata": {
        "id": "jMp5c4fHxuOE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def is_containing_other_spans(span: list[int], all_spans: list[list[int]]):\n",
        "    return any([s[0] >= span[0] and s[1] <= span[1] and s != span for s in all_spans])"
      ],
      "metadata": {
        "id": "1xfsUDpQxt-o"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def improved_replace_corefs(document, clusters):\n",
        "    resolved = list(tok.text_with_ws for tok in document)\n",
        "    all_spans = [span for cluster in clusters for span in cluster]  # flattened list of all spans\n",
        "\n",
        "    for cluster in clusters:\n",
        "        noun_indices = get_span_noun_indices(document, cluster)\n",
        "\n",
        "        if noun_indices:\n",
        "            mention_span, mention = get_cluster_head(document, cluster, noun_indices)\n",
        "\n",
        "            for coref in cluster:\n",
        "                if coref != mention and not is_containing_other_spans(coref, all_spans):\n",
        "                    core_logic_part(document, coref, resolved, mention_span)\n",
        "\n",
        "    return \"\".join(resolved)"
      ],
      "metadata": {
        "id": "8w_NWWTzxt2l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_fast_cluster_spans(doc, clusters):\n",
        "    fast_clusters = []\n",
        "    for cluster in clusters:\n",
        "        new_group = []\n",
        "        for tuple in cluster:\n",
        "            #print(type(tuple), tuple)\n",
        "            (start, end) = tuple\n",
        "            #print(\"start, end\", start, end)\n",
        "            span = doc.char_span(start, end)\n",
        "            #print('span', span.start, span.end)\n",
        "            new_group.append([span.start, span.end-1])\n",
        "        fast_clusters.append(new_group)\n",
        "    return fast_clusters"
      ],
      "metadata": {
        "id": "Dl-xExsXzr5b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_fastcoref_clusters(doc, text, fcoref_model):\n",
        "  try:\n",
        "    preds = fcoref_model.predict(texts=[text])\n",
        "    fast_clusters = preds[0].get_clusters(as_strings=False)\n",
        "    fast_cluster_spans = get_fast_cluster_spans(doc, fast_clusters)\n",
        "  except Exception as e:\n",
        "    print(f'Error received while getting coreference clusters. {e}. Returning original text')\n",
        "    return text\n",
        "  return fast_cluster_spans"
      ],
      "metadata": {
        "id": "jI6GwJ4izrm3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def perform_coreference_resolution(text):\n",
        "    doc = nlp_scispacy(text)\n",
        "    clusters = get_fastcoref_clusters(doc, text, fcoref_model)\n",
        "    coref_text = improved_replace_corefs(doc, clusters)\n",
        "    return coref_text"
      ],
      "metadata": {
        "id": "8QP_CENgubnr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered['input_coref'] = pubmed_inference_filtered['input'].apply(perform_coreference_resolution)"
      ],
      "metadata": {
        "collapsed": true,
        "id": "G2OVV4ykxOrx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "def preprocess_test_data(examples):\n",
        "    model_inputs = rebel_tokenizer(\n",
        "        examples[\"input_coref\"],  # Source text\n",
        "        max_length=512,\n",
        "        truncation=True,\n",
        "        padding=\"max_length\"\n",
        "    )\n",
        "    return model_inputs"
      ],
      "metadata": {
        "id": "GZ53r5iqy9LI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered['input_coref']"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 458
        },
        "id": "DQqjY9yR0gWY",
        "outputId": "e0ed176a-9602-492f-ca72-02f1fb48bc99"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0      Genetic and phenotypic diversity of NHE6 mutat...\n",
              "1      LSD1-mediated epigenetic modification is requi...\n",
              "2      TFBSshape: a motif database for DNA shape feat...\n",
              "3      Histone H3K27ac separates active from poised e...\n",
              "4      Novel C12orf65mutations in patients with axona...\n",
              "                             ...                        \n",
              "98     MARS: improving multiple circular sequence ali...\n",
              "99     JACUSA: site-specific identification of RNA ed...\n",
              "100    RIblast: an ultrafast RNA–RNA interaction pred...\n",
              "101    TIRR regulates 53BP1 by masking 53BP1's histon...\n",
              "102    CarbonylDB: a curated data-resource of protein...\n",
              "Name: input_coref, Length: 103, dtype: object"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>input_coref</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Genetic and phenotypic diversity of NHE6 mutat...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>LSD1-mediated epigenetic modification is requi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>TFBSshape: a motif database for DNA shape feat...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Histone H3K27ac separates active from poised e...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Novel C12orf65mutations in patients with axona...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>MARS: improving multiple circular sequence ali...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>99</th>\n",
              "      <td>JACUSA: site-specific identification of RNA ed...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>RIblast: an ultrafast RNA–RNA interaction pred...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>101</th>\n",
              "      <td>TIRR regulates 53BP1 by masking 53BP1's histon...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>102</th>\n",
              "      <td>CarbonylDB: a curated data-resource of protein...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>103 rows × 1 columns</p>\n",
              "</div><br><label><b>dtype:</b> object</label>"
            ]
          },
          "metadata": {},
          "execution_count": 46
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered.to_csv('pubmed_inderence_filtered.csv')"
      ],
      "metadata": {
        "id": "qW8tc7E9EPhZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def extract_triplets(decoded_string):\n",
        "    decoded_str_copy = decoded_string\n",
        "    if '<s>' in decoded_string:\n",
        "      decoded_str_copy = ''.join(decoded_string.split('<s>')[1:])\n",
        "      decoded_str_copy = decoded_str_copy.split('</s>')[0]\n",
        "\n",
        "    triplets = []\n",
        "    parts = decoded_str_copy.split(\"<triplet>\")\n",
        "\n",
        "    for part in parts[1:]:  # Skip the first empty split\n",
        "        # Extract the head entity (the first part before <subj>)\n",
        "        head, *relations = part.split(\"<subj>\")\n",
        "        head = head.strip()\n",
        "\n",
        "        # Parse each <subj> and <obj> pair\n",
        "        for relation in relations:\n",
        "            if \"<obj>\" in relation:\n",
        "                subj, obj = relation.split(\"<obj>\", 1)\n",
        "                triplets.append((head, subj.strip(), obj.strip()))\n",
        "\n",
        "    return triplets"
      ],
      "metadata": {
        "id": "DAmuk6SqxYwt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def perform_inference(text):\n",
        "  extracted_triplets = []\n",
        "  for sentence in text.split('. '):\n",
        "    inputs = rebel_tokenizer(sentence, return_tensors=\"pt\", truncation=True, max_length=512).to(device)\n",
        "    outputs = rebel_model.generate(**inputs)\n",
        "    decoded_output = rebel_tokenizer.decode(outputs[0], skip_special_tokens=False)\n",
        "    triplets = list(set(extract_triplets(decoded_output)))\n",
        "    if len(triplets)==0:\n",
        "      continue\n",
        "    else:\n",
        "      extracted_triplets.extend(triplets)\n",
        "\n",
        "  return extracted_triplets"
      ],
      "metadata": {
        "id": "VHYPqGMV0-gR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered['triplets'] = pubmed_inference_filtered['input_coref'].apply(perform_inference)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NstgPsISGID4",
        "outputId": "7d7ee530-f637-4b55-c0b5-5ebdec9cac0f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-68-620b74cd144d>:1: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  pubmed_inference_filtered['triplets'] = pubmed_inference_filtered['input_coref'].apply(perform_inference)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pubmed_inference_filtered.to_csv('pubmed_inference_filtered_with_triplets.csv')"
      ],
      "metadata": {
        "id": "DjRACwEMGH1c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets = list(set().union(*pubmed_inference_filtered['triplets']))"
      ],
      "metadata": {
        "id": "kmULvvQ8MLFM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "kr0NDUMrMK9D",
        "outputId": "4055bac9-b7a2-4f35-c69d-9eb96c03e679"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('p53', 'MDM2', 'Association'),\n",
              " ('RNA polymerase II', '5S ribosomal RNA', 'Association'),\n",
              " ('LSD1', 'p4.2 promoter', 'Positive_Correlation'),\n",
              " ('Preeclampsia', 'hyperinflammatory', 'Association'),\n",
              " ('DNA damage', 'H2A', 'Positive_Correlation'),\n",
              " ('Ezh2', 'H3K27 methyltransferase', 'Association'),\n",
              " ('p53', 'cancer', 'Positive_Correlation'),\n",
              " ('TMPDB', 'web site', 'instance of'),\n",
              " ('FRDA', 'Friedreich’s PsyNetataxia', 'Association'),\n",
              " ('erythropoiesis', 'GATA-1', 'Association'),\n",
              " ('Gfi-1b', 'hematopoietic factor', 'subclass of'),\n",
              " ('Ebola Virus Infection', 'ZMapp', 'Positive_Correlation'),\n",
              " ('GATA-1', 'hematopoietic factor', 'subclass of'),\n",
              " ('Pta1 Scaffold Protein', 'Pta1 Scaffold Protein', 'Association'),\n",
              " ('pineal region', 'spinal cord', 'connects with'),\n",
              " ('5S ribosomal RNA', 'RNA polymerase II', 'Association'),\n",
              " ('cell cycle progression', 'DNA replication', 'Positive_Correlation'),\n",
              " ('TAL1', 'LSD1', 'Positive_Correlation'),\n",
              " ('Eco1', 'sister chromatid cohesion', 'Association'),\n",
              " ('hyperinflammatory', 'Preeclampsia', 'Association'),\n",
              " ('CGRP', 'galcanezumab', 'Positive_Correlation'),\n",
              " ('K27M-H3', 'HIST1H3B', 'Positive_Correlation'),\n",
              " ('HCNEs', 'noncoding element', 'subclass of'),\n",
              " ('galcanezumab', 'CGRP', 'Positive_Correlation'),\n",
              " ('hSET1', 'TAL1', 'Association'),\n",
              " ('H3K4 methylation', 'H3K4 methylation', 'Positive_Correlation'),\n",
              " ('TMPDB', 'TMPDB', 'field of work'),\n",
              " ('GATA-1', 'Gfi-1b', 'Positive_Correlation'),\n",
              " ('Hst 5', 'Histatin 5', 'instance of'),\n",
              " ('thermosensitive growth', 'thermosensitive growth', 'Positive_Cor'),\n",
              " ('LSD1', 'thymidine kniase minimal promoter', 'Positive_Correlation'),\n",
              " ('HCNEs', 'GRBs', 'Association'),\n",
              " ('IDH1/2', 'G34V/R-H3', 'Positive_Correlation'),\n",
              " ('UniPROBE', 'JASPAR', 'Association'),\n",
              " ('cohesin acetylation', 'cohesin acetylation', 'Positive_Correlation'),\n",
              " ('TAL1', 'hematopoiesis', 'Association'),\n",
              " ('TAL1', 'epigenetic modification', 'Association'),\n",
              " ('fork slowing', 'H2A', 'Positive_Correlation'),\n",
              " ('C.albicans', 'C.albicans', 'parent taxon'),\n",
              " ('Cannotatr', 'genomic', 'field of work'),\n",
              " ('Pfh1', '5S rRNA', 'Association'),\n",
              " ('STAT1', 'interferon-gamma stimulationin vivo', 'Association'),\n",
              " ('1500 kb variant of chromosome XII', 'chromosome XII', 'subclass of'),\n",
              " ('dystrophin', 'DAPC', 'Association'),\n",
              " ('ATMIN', 'ATM cofactor', 'Association'),\n",
              " ('TNF', 'hematopoietic regulatory genes', 'Association'),\n",
              " ('p53', 'MDM2', 'Positive_Correlation'),\n",
              " ('DNA helicase', 'DNA damage', 'Positive_Correlation'),\n",
              " ('ZMapp', 'Ebola Virus Infection', 'Positive_Correlation'),\n",
              " ('TF', 'genomic regions', 'Association'),\n",
              " ('H3', 'IDH1/2', 'Positive_Correlation'),\n",
              " ('histone histone H3',\n",
              "  'histone histone histone H3.3 mutations',\n",
              "  'Association'),\n",
              " ('phosphoryl transfer reactions',\n",
              "  'phosphoryl transfer reactions',\n",
              "  'Association'),\n",
              " ('DNA damage', 'Pfh1', 'Positive_Correlation'),\n",
              " ('Amitriptyline',\n",
              "  'Childhood and Adolescent Migraine',\n",
              "  'Positive_Correlation'),\n",
              " ('TAL1', 'histone lysine demethylase LSD1', 'Positive_Correlation'),\n",
              " ('ATR', 'PR-Set7', 'Positive_Correlation'),\n",
              " ('ATRXmutations', 'histone H3', 'Association'),\n",
              " ('choroid plexus', 'transthyretin', 'has part'),\n",
              " ('STAT1', 'STAT1', ''),\n",
              " ('Anifrolumab',\n",
              "  'Anti–Interferon‐α Receptor Monoclonal Antibody',\n",
              "  'Positive_Interferon‐α Receptor Monoclonal Antibody <obj> Positive_Interferon‐α Receptor'),\n",
              " ('Christianson syndrome', 'NHE6', 'Association'),\n",
              " ('LADs', 'LADs', 'Association'),\n",
              " ('H3', 'G34V/R-H3', 'Positive_Correlation'),\n",
              " ('2-aminobenzamide', 'Huntington’s disease (HD)', 'Positive_Correlation'),\n",
              " ('compound 106', 'HDAC inhibitors', 'Positive_Correlation'),\n",
              " ('H3K27 methyltransferase', 'Ezh2', 'Association'),\n",
              " ('TAL1', 'p4.2 promoter', 'Positive_Correlation'),\n",
              " ('H3', 'G34', 'Association'),\n",
              " ('transthyretin', 'transthyretin', 'Association'),\n",
              " ('proteins', 'mass spectrometry', 'Positive_Correlation'),\n",
              " ('Dur31p', 'polyamine transporters', 'Positive_Correlation'),\n",
              " ('DNA polymerase', 'H2A', 'Positive_Correlation'),\n",
              " ('histone H4 lysine 20', 'ATR', 'Association'),\n",
              " ('Metazoan', 'CNEs', 'Association'),\n",
              " ('GENCODE', 'lncRNA', 'Association'),\n",
              " ('erythropoiesis', 'MeCP1', 'Association'),\n",
              " ('JACUSA', 'RNA editing', 'field of work'),\n",
              " ('protein carbonylation sites', 'protein carbonylation sites', 'Association'),\n",
              " ('mast', 'megakaryocytic', 'Association'),\n",
              " ('β-dystroglycan', 'DAPC', 'Association'),\n",
              " ('eliglustat', 'Gaucher disease type 1', 'Positive_Correlation'),\n",
              " ('TADs', 'TADs', 'subclass of'),\n",
              " ('HDAC', 'FRDA', 'Positive_Correlation'),\n",
              " ('mammary cellsin', 'tumor metastasisin', 'Positive_Correlation'),\n",
              " ('β-1,3-glucans', 'β-1,3-glucans', 'Association'),\n",
              " ('histone methyl-lysine binding', '53BP1', 'Association'),\n",
              " ('lncRNA', 'long noncoding RNA', 'Association'),\n",
              " ('Drosophila Ctf4', 'cell cycle progression', 'Association'),\n",
              " ('long noncoding RNA', 'lncRNA', 'Association'),\n",
              " ('lysine residues', 'lysine residues', 'Association'),\n",
              " ('lncRNA', 'prognostic', 'subclass of'),\n",
              " ('Christianson syndrome', 'Christianson syndrome', 'Association'),\n",
              " ('QTcF', 'QTcF', 'Association'),\n",
              " ('Gaucher disease type 1', 'eliglustat', 'Positive_Correlation'),\n",
              " ('NBS1', 'ATM cofactor', 'Association'),\n",
              " ('ZCCHC17', 'Alzheimer’s disease', 'Association'),\n",
              " ('snoRNA termination',\n",
              "  'RNA polymerase II C-terminal domain Ser5-P dephosphorylation',\n",
              "  'Positive_Correlation'),\n",
              " ('Swi1', 'TIMELESS homolog', 'Association'),\n",
              " ('gynecological cancers', 'cancers', 'subclass of'),\n",
              " ('CNEE', 'CNEE', 'subclass of'),\n",
              " ('IDH', 'K27M-H3', 'Positive_Correlation'),\n",
              " ('spinal cord', 'cerebellum', 'connects with'),\n",
              " ('Cancer Genome Atlas', 'TCGA2BED', 'derivative work'),\n",
              " ('ATR', 'histone H4 lysine 20', 'Positive_Correlation'),\n",
              " ('TAL1', 'hSET1', 'Positive_Correlation'),\n",
              " ('GATA-1', 'hematopoietic factor', 'Association'),\n",
              " ('GATA-1', 'TAL-1', 'Positive_Correlation'),\n",
              " ('K27M-H3', 'IDH', 'Positive_Correlation'),\n",
              " ('HDAC inhibitors', 'compound 106', 'Positive_Correlation'),\n",
              " ('webSDA', 'macromolecular diffusional association', 'use'),\n",
              " ('RPCs', 'Mrc1', 'has part'),\n",
              " ('Mrc1', 'RPCs', 'Association'),\n",
              " ('themutSLoperon', 'themutSLoperon', 'Positive_Correlation'),\n",
              " ('thermosensitive growth', 'thermosensitive growth', 'Positive_Correlation'),\n",
              " ('hematopoiesis', 'TAL1', 'Association'),\n",
              " ('Anifrolumab',\n",
              "  'Anti–Interferon‐α Receptor Monoclonal Antibody',\n",
              "  'Positive_Interferon‐α Receptor'),\n",
              " ('mass spectrometry', 'proteins', 'Positive_Correlation'),\n",
              " ('JASPAR', 'UniPROBE', 'Association'),\n",
              " ('ATRXandTP53mutations', 'K27M-H3', 'Positive_Correlation'),\n",
              " ('MCM2-7', 'DNA polymerase α', 'Association'),\n",
              " ('RPCs', 'DNA damage checkpoint', 'Positive_Correlation'),\n",
              " ('breast cancer', 'lncRNAs', 'Association'),\n",
              " ('MeCP1', 'erythropoiesis', 'Association'),\n",
              " ('hepadnaviral cccDNA', 'minichromosomes', 'subclass of'),\n",
              " ('cancer', 'p53', 'Positive_Correlation'),\n",
              " ('UniPROBE', 'UniPROBE', 'Association'),\n",
              " ('GRBs', 'HCNEs', 'Association'),\n",
              " ('GAL4 reporter',\n",
              "  'thymidine kniase minimal promoter',\n",
              "  'Positive_Correlation'),\n",
              " ('GATA-1', 'MeCP1', 'Positive_Correlation'),\n",
              " ('p4.2 promoter', 'LSD1', 'Positive_Correlation'),\n",
              " ('β-1,3-glucans', 'β-1,3-glucans', 'Positive_Correlation'),\n",
              " ('α-sarcoglycan', 'DAPC', 'Association'),\n",
              " ('LADs', 'TADs', 'Association'),\n",
              " ('transthyretin', 'choroid plexus', 'Association'),\n",
              " ('polyamine transporters', 'polyamine transporters', 'Positive_Correlation'),\n",
              " ('thalamus', 'spinal cord', 'connects with'),\n",
              " ('tRNA', 'RNA polymerase II', 'Association'),\n",
              " ('DNA', 'COUGER', 'Association'),\n",
              " ('protein carbonylation sites',\n",
              "  'protein carbonylation sites',\n",
              "  'Association <obj> Association'),\n",
              " ('EBF1', 'hematopoietic regulatory genes', 'Association'),\n",
              " ('cerebellum', 'spinal cord', 'connects with'),\n",
              " ('lncRNA', 'GENCODE', 'Association'),\n",
              " ('sister chromatid cohesion', 'Eco1', 'Association'),\n",
              " ('RNA polymerase II', 'tRNA', 'Association'),\n",
              " ('H3.3', 'disease development', 'Association'),\n",
              " ('heart failure (HF)1', 'SERCA2a', 'Positive_Correlation'),\n",
              " ('mast erythroid', 'megakaryocytic', 'Association'),\n",
              " ('GATA-1', 'erythroid differentiation', 'Association'),\n",
              " ('GATA-1', 'ACF', 'Positive_Correlation'),\n",
              " ('DNA replication', 'cell cycle progression', 'Association'),\n",
              " ('RPCs', 'Mrc1', 'Association'),\n",
              " ('LSD1', 'TAL1', 'Association'),\n",
              " ('Pfh1', 'DNA damage', 'Positive_Correlation'),\n",
              " ('LIMT', 'tumor metastasisin', 'Positive_Correlation'),\n",
              " ('TP53mutations', 'histone H3', 'Association'),\n",
              " ('RNA polymerase II C-terminal domain Ser5-P dephosphorylation',\n",
              "  'snoRNA termination',\n",
              "  'Positive_Correlation'),\n",
              " ('H3', 'H3', 'Positive_Correlation'),\n",
              " ('genomic regions', 'TF', 'Association'),\n",
              " ('tumor metastasisin', 'mammary cellsin', 'Positive_Correlation'),\n",
              " ('cohesin', 'Smc3', 'has part'),\n",
              " ('GATA-1', 'WCRF', 'Positive_Correlation'),\n",
              " ('Topiramate', 'Childhood and Adolescent Migraine', 'Positive_Correlation'),\n",
              " ('spinal cord', 'thalamus', 'connects with'),\n",
              " ('DNA damage checkpoint', 'chromosome replication', 'Positive_Correlation'),\n",
              " ('IEMs', 'IEMs', 'Association'),\n",
              " ('MCM2-7 helicase to DNA polymerase α',\n",
              "  'MCM2-7 helicase to MCM2-7 helicase',\n",
              "  'connects with'),\n",
              " ('TADs', 'LADs', 'Association'),\n",
              " ('MCM2-7 helicase to DNA polymerase α',\n",
              "  'MCM2-7 helicase to DNA polymerase α',\n",
              "  'connects with'),\n",
              " ('SERCA2a', 'heart failure (HF)1', 'Positive_Correlation'),\n",
              " ('HIST1H3B', 'K27M-H3', 'Positive_Correlation'),\n",
              " ('Mrc1', 'RPCs', 'part of'),\n",
              " ('iPSC', 'Friedreich’s PsyNetataxia', 'Positive_Correlation'),\n",
              " ('mitochondrial respiratory chain complexes I and IV',\n",
              "  'mitochondrial respiratory chain complexes I and IV',\n",
              "  'has part'),\n",
              " ('PR-Set7', 'ATR', 'Positive_Correlation'),\n",
              " ('Placebo', 'Childhood and Adolescent Migraine', 'Positive_Correlation'),\n",
              " ('histone H3', 'histone H3.3 wild type', 'Association'),\n",
              " ('ATM', 'T cell development', 'Association'),\n",
              " ('gynecological cancers', 'gynecological cancers', 'Association'),\n",
              " ('polyadenylation site analysis',\n",
              "  'polyadenylation site analysis',\n",
              "  'subclass of'),\n",
              " ('JADE', 'Adverse Drug Events', 'field of work'),\n",
              " ('HDAC', 'Huntington’s disease (HD)', 'Positive_Correlation'),\n",
              " ('Ozanimod', 'Ozanimod', 'Association'),\n",
              " ('TADs', 'TADs', 'Association'),\n",
              " ('Friedreich’s PsyNetataxia', 'iPSC', 'Positive_Correlation'),\n",
              " ('p53', 'tumor suppressor', 'Positive_Correlation'),\n",
              " ('spinal cord', 'pineal region', 'connects with'),\n",
              " ('HDAC1', 'histone lysine demethylase LSD1', 'Positive_Correlation'),\n",
              " ('NanoStringNormCNV', 'NanoString CNV', 'based on'),\n",
              " ('tumorigenesis', 'tumorigenesis', 'Association'),\n",
              " ('LIMT', 'mammary cellsin', 'Positive_Correlation'),\n",
              " ('NBS1', 'ATM', 'Association'),\n",
              " ('GATA-1', 'erythropoiesis', 'Association'),\n",
              " ('proteasome', '@@@@@@@@p53', 'Positive_Correlation'),\n",
              " ('2-aminobenzamide', 'FRDA', 'Positive_Correlation'),\n",
              " ('transthyretin', 'choroid plexus', 'part of'),\n",
              " ('450 kb chromosome XII variant', 'chromosome XII', 'subclass of'),\n",
              " ('STAT1', 'STAT1', 'Association'),\n",
              " ('2-aminobenzamide', 'neurodegenerative diseases', 'Positive_Correlation'),\n",
              " ('K27M-H3', 'ATRXandTP53mutations', 'Positive_Correlation'),\n",
              " ('ATM', 'NBS1', 'Positive_Correlation'),\n",
              " ('ASCIZ', 'ATM cofactor', 'Association'),\n",
              " ('TCGA2BED', 'Cancer Genome Atlas', 'based on'),\n",
              " ('GATA1', 'hematopoietic regulatory genes', 'Association'),\n",
              " ('Pfh1', 'Pfh1', 'Association'),\n",
              " ('TAL1', 'thymidine kniase minimal promoter', 'Positive_Correlation'),\n",
              " ('H3', 'K36M', 'Association'),\n",
              " ('MDM2', 'p53', 'Positive_Correlation'),\n",
              " ('Smc3', 'cohesin', 'part of'),\n",
              " ('Alzheimer’s disease', 'ZCCHC17', 'Association'),\n",
              " ('53BP1', 'histone methyl-lysine binding', 'Association'),\n",
              " ('RAPID', 'UK', 'country'),\n",
              " ('H3', 'K27M', 'Association'),\n",
              " ('megakaryocytic', 'megakaryocytic', 'Association'),\n",
              " ('RG7112', 'p53', 'Positive_Correlation'),\n",
              " ('DNA damage', 'DNA helicase', 'Positive_Correlation'),\n",
              " ('CNEs', 'Metazoan', 'Association'),\n",
              " ('themutSLoperon', 'themutSLoperon', 'Positive'),\n",
              " ('breast cancer', 'breast cancer', 'Association')]"
            ]
          },
          "metadata": {},
          "execution_count": 74
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open(\"aggregated_triplets.txt\", \"w\") as f:\n",
        "    for triplet in aggregated_triplets:\n",
        "        f.write(f\"{triplet}\\n\")"
      ],
      "metadata": {
        "id": "p4-S0Q4EMnH1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "df = pd.read_csv('clustered_data.csv')"
      ],
      "metadata": {
        "id": "LfKI3DpS3IjI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 414
        },
        "id": "qjYi5dCx3Ifw",
        "outputId": "13e1fbd8-948d-4a39-b27e-35b14a5d8fce"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                            abstract  \\\n",
              "0   Germline pathogenic variants associated with ...   \n",
              "1                                                NaN   \n",
              "2                                                NaN   \n",
              "3                                                NaN   \n",
              "4                                                NaN   \n",
              "\n",
              "                                        introduction  \\\n",
              "0   With the rise of the genomic era, a growing n...   \n",
              "1   Globally, colorectal cancer (CRC) is the thir...   \n",
              "2   The advances in sequencing technology have im...   \n",
              "3   Breast carcinoma is the leading cause of canc...   \n",
              "4   Characterising the genomic landscape of somat...   \n",
              "\n",
              "                                             results  \\\n",
              "0   pCPS genes with monoallelic risk phenotypes s...   \n",
              "1   Establishment of CRC organoids from primary t...   \n",
              "2   Selection of 33 patients with suspected hered...   \n",
              "3   3.1. Patient demographics and clinicopatholog...   \n",
              "4                                                NaN   \n",
              "\n",
              "                                          discussion  \\\n",
              "0   Here we have explored the theory that AD(LoF)...   \n",
              "1   Current strategies for precision medicine are...   \n",
              "2   In this study, we demonstrated that TAS-LRS c...   \n",
              "3   We addressed two goals in this study. First, ...   \n",
              "4                                                NaN   \n",
              "\n",
              "                                             methods  cluster_id  \\\n",
              "0   In order to ensure reproducibility the full, ...           0   \n",
              "1   Establishment of patient-derived organoid cul...           4   \n",
              "2   Ethics approval and consent to participateThe...           3   \n",
              "3                                                NaN           2   \n",
              "4                                                NaN           0   \n",
              "\n",
              "                                               title  \n",
              "0                                                NaN  \n",
              "1  A functional personalised oncology approach ag...  \n",
              "2  Assessing the efficacy of target adaptive samp...  \n",
              "3  Genomic disparity impacts variant classificati...  \n",
              "4  COSMIC: a curated database of somatic variants...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-456c667b-391e-4c90-8ae5-323775629d2c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>abstract</th>\n",
              "      <th>introduction</th>\n",
              "      <th>results</th>\n",
              "      <th>discussion</th>\n",
              "      <th>methods</th>\n",
              "      <th>cluster_id</th>\n",
              "      <th>title</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Germline pathogenic variants associated with ...</td>\n",
              "      <td>With the rise of the genomic era, a growing n...</td>\n",
              "      <td>pCPS genes with monoallelic risk phenotypes s...</td>\n",
              "      <td>Here we have explored the theory that AD(LoF)...</td>\n",
              "      <td>In order to ensure reproducibility the full, ...</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Globally, colorectal cancer (CRC) is the thir...</td>\n",
              "      <td>Establishment of CRC organoids from primary t...</td>\n",
              "      <td>Current strategies for precision medicine are...</td>\n",
              "      <td>Establishment of patient-derived organoid cul...</td>\n",
              "      <td>4</td>\n",
              "      <td>A functional personalised oncology approach ag...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NaN</td>\n",
              "      <td>The advances in sequencing technology have im...</td>\n",
              "      <td>Selection of 33 patients with suspected hered...</td>\n",
              "      <td>In this study, we demonstrated that TAS-LRS c...</td>\n",
              "      <td>Ethics approval and consent to participateThe...</td>\n",
              "      <td>3</td>\n",
              "      <td>Assessing the efficacy of target adaptive samp...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Breast carcinoma is the leading cause of canc...</td>\n",
              "      <td>3.1. Patient demographics and clinicopatholog...</td>\n",
              "      <td>We addressed two goals in this study. First, ...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2</td>\n",
              "      <td>Genomic disparity impacts variant classificati...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Characterising the genomic landscape of somat...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>COSMIC: a curated database of somatic variants...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-456c667b-391e-4c90-8ae5-323775629d2c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-456c667b-391e-4c90-8ae5-323775629d2c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-456c667b-391e-4c90-8ae5-323775629d2c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-c3f2feba-efff-4f68-8caf-1828f84574e1\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-c3f2feba-efff-4f68-8caf-1828f84574e1')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-c3f2feba-efff-4f68-8caf-1828f84574e1 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 28,\n  \"fields\": [\n    {\n      \"column\": \"abstract\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \" Head and neck squamous cell carcinomas (HNSCCs) are the most devastating diseases in India and southeast Asia. It is a preventable and curable disease if detected early. Tobacco and alcohol consumption are the two major risk-factors but infection of high-risk HPVs are also associated with development of predominantly oral and oropharyngeal carcinomas. Interestingly, unlike cervical cancer, HPV-induced HNSCCs show good prognosis and better survival in contrast, majority of tobacco-associated HPV\\u2212veHNSCCs are highly aggressive with poor clinical outcome. Biomarker analysis in circulatory body-fluids for early cancer diagnosis, prognosis and treatment monitoring are becoming important in clinical practice. Early diagnosis using non-invasive saliva for oral or other diseases plays an important role in successful treatment and better prognosis. Saliva mirrors the body's state of health as it comes into direct contact with oral lesions and needs no trained manpower to collect, making it a suitable bio-fluid of choice for screening. Saliva can be used to detect not only virus, bacteria and other biomarkers but variety of molecular and genetic markers for an early detection, treatment and monitoring cancer and other diseases. The performance of saliva-based diagnostics are reported to be highly (\\u226595\\u00a0%) sensitive and specific indicating the test's ability to correctly identify true positive or negative cases. This review focuses on the potentials of saliva in the early detection of not only HPV or other pathogens but also identification of highly reliable gene mutations, oral-microbiomes, metabolites, salivary cytokines, non-coding RNAs and exosomal miRNAs. It also discusses the importance of saliva as a reliable, cost-effective and an easy alternative to invasive procedures.Keywords:Head and neck cancer, Saliva, Non-invasive, Liquid biopsy, Biomarker, Early detection, HPV screening\",\n          \" Germline pathogenic variants associated with increased childhood mortality must be subject to natural selection. Here, we analyze publicly available germline genetic metadata from 4,574 children with cancer [11 studies; 1,083 whole exome sequences (WES), 1,950 whole genome sequences (WGS), and 1,541 gene panel] and 141,456 adults [125,748 WES and 15,708 WGS]. We find that pediatric cancer predisposition syndrome (pCPS) genes [n\\u2009=\\u200985] are highly constrained, harboring only a quarter of the loss-of-function variants that would be expected. This strong indication of selective pressure on pCPS genes is found across multiple lines of germline genomics data from both pediatric and adult cohorts. For six genes [ELP1,GPR161,VHLandSDHA/B/C], a clear lack of mutational constraint calls the pediatric penetrance and/or severity of associated cancers into question. Conversely, out of 23 known pCPS genes associated with biallelic risk, two [9%,DIS3L2andMSH2] show significant constraint, indicating that they may monoallelically increase childhood cancer risk. In summary, we show that population genetic data provide empirical evidence that heritable childhood cancer leads to natural selection powerful enough to have significantly impacted the present-day gene pool.Subject terms:Cancer genomics, Evolutionary genetics, Paediatric cancer, Cancer genetics\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"introduction\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 23,\n        \"samples\": [\n          \" For over two decades, HER2 has been classified in a binary fashion in breast oncology: positive, if overexpressed (immunohistochemistry [IHC] 3+) or amplified (positive in situ hybridization [ISH]), and negative in the absence of these alterations1. Based on this paradigm, 80\\u201385% of breast cancers were traditionally defined as HER2-negative, despite over half of these having detectable HER2 protein by IHC2. According to the latest American Society of Clinical Oncology/College of American Pathologists (ASCO/CAP) Guidelines for HER2 testing, HER2-negative breast cancers include tumors with HER2 IHC scores of 0, 1+ or 2+/ISH-negative1, for which treatment with traditional HER2-targeting agents\\u00a0has not demonstrated meaningful clinical benefits3,4. A paradigm shift in this setting has however occurred with the emergence of anti-HER2 antibody-drug conjugates (ADCs)5, several of which have demonstrated relevant activity among patients with HER2-negative metastatic breast cancer (MBC) exhibiting HER2-low expression, defined as IHC 1+ or 2+/ISH-negative6\\u201310. In this setting, the phase 3 DESTINY-Breast04 trial demonstrated that trastuzumab deruxtecan (T-DXd) is associated with substantial improvements in progression-free survival and overall survival compared with traditional chemotherapy among patients with HER2-low MBC11, leading to the approval of the drug by the U.S. Food and Drug Administration and European Medicines Agency, and reshaping treatment algorithms in breast oncology.Currently, patients with HER2-low MBC may be offered T-DXd, whereas T-DXd is not considered the standard of care in patients with HER2 IHC 0 (hereafter referred to as HER2-0) MBC. However, it remains unclear whether HER2-low breast cancer\\u00a0should be considered a distinct molecular entity, with different genomic underpinnings compared with HER2-0 tumorsbreast cancer12,13.In this study, we aim to characterize the genomic profile of HER2-low tumors across a large population of patients with MBC evaluated at Dana-Farber Cancer Institute and to compare it with the genomic profile of HER2-0 tumors after correcting for potential confounding variables.\",\n          \" ThePOLD1gene encodes the 125 kDa catalytic subunit of the heterotetrameric POL\\u03b4 complex.POLD1catalyzes two essential functions of POL\\u03b4: synthesizing Okazaki fragments with its DNA polymerase activity and DNA proofreading with its 3\\u2032 to 5\\u2032 exonuclease activity [1]. In humans, germline mutations inPOLD1resulting in defective DNA proofreading exonuclease function are associated with hereditary colorectal and endometrial cancer and colonic polyposis. Clinically, these tumors a exhibit high tumor mutation burden (TMB), increased presence of tumor infiltrating lymphocytes (TIL), and favorable response to immune checkpoint inhibitors [2]. Therefore, accurate detection of all affected patients has the potential to significantly impact their clinical management, including cancer surveillance, genetic counseling of family members, and precision oncology for immunotherapy.POLD1contains several important conserved motifs: Amino acid residues 4\\u201319 form the nuclear localization signal, 304\\u2013533 form the exonuclease domain, and 579\\u2013974 form the polymerase domains [3].POLD1\\u2019s exonuclease domain is a member of the DEDD-type exonuclease (Exo) superfamily characterized by three sequence motifs: Exo1, Exo2, and Exo3 [3]. Within the three sequence motifs, there are four highly conserved, invariant catalytic amino acids, DEDD, which have carboxylic sidechains. The DEDD motif is composed of p.D316 (located in Exo1), pE318 (located in Exo1), p.D402 (located in Exo2), and p.D515 (located in Exo3) (Figure 1A). The DEDD amino acids use their negatively charged carboxylate sidechains to coordinate the positively charged divalent metal ions to catalyze a phosphoryl-transfer reaction, thereby removing misincorporated nucleotides [4,5]. In yeast, mice, and human normal intestinal cells,POLD1\\u2019s DNA proofreading function is disrupted when an amino acid in the DEDD motif is mutated with a non-acidic amino acid lacking a carboxylate sidechain (charge-discordant AA substitution) [6,7,8,9].Open in a separate windowFigure 1(A) GermlinePOLD1mutations in humans in relation to protein structure. Currently, two pathogenic variants have been identified (green circles) within the exonuclease domain; thirteen charge-altering mutations are found within the DEDD motif (red circles), all currently classified as variants of uncertain significance. (B) Histology (H&E) of endometrial cancer carrying the p.D402NPOLD1mutation. Note the high amount of tumor infiltrating lymphocytes in the microenvironment, in keeping with the tumor\\u2019s high mutational status. (C) The family pedigree tree of the index patient with germline POLD1 p.D402N mutation. Black filled symbols indicate endometrial cancer (EC) and breast cancer (BC), respectively. Ages at diagnosis are indicated for each patient at the top of each shape. Circle and square shapes in the pedigree diagramph represent female and male gender, respectively. Number inside the shape represent number of sibling of the same gender. Black arrow denotes the index patient.Currently, virtually all clinically actionable pathogenic/likely pathogenicPOLD1variants (p.S478N and p.L474P) are missense mutations within the exonuclease DNA proofreading domain and spare the DEDD motif (Figure 1A). Our search of the ClinVar database [10] for germline missense mutations causing charge-discordant AA substitution in the DEDD motif yielded thirteen variant entries: all are classified as variants of uncertain significance (VUS;Supplementary Table S1). This designation is clinically inactionable, and additional supporting evidence is required for their reclassification. Due to the mechanistic significance of these four carboxylic acid sidechains in the DEDD motif, we hypothesize charge-discordant AA substitutions would result in defective exonuclease function, DNA proofreading errors, and be tumorigenic in humans.Herein, we studied an index case of a young woman with early onset endometrial cancer, who was found to be a carrier of a charge-discordant AA substitution in the DEDD motif (p.D402N). Using a multiprong approach, we systematically investigated the pathogenic effect of this mutation class. Our goal is to provide evidence to support its reclassification to pathogenic/likely pathogenic, so that these patients will obtain appropriate genetic counseling and cancer screening. Furthermore, recent data have shown thatPOLEandPOLD1mutation status is a predictive biomarker for immune checkpoint inhibitor response. Failure to reclassify these variants could lead to opportunity loss for personalized immune checkpoint inhibitor therapy.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"results\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 25,\n        \"samples\": [\n          \" Patients and tumour descriptionHere, we present the results of the first 200 consecutive patients diagnosed with NSCLC that were sequenced with the TSO500 NGS panel. Clinical and pathological features are summarised inTable\\u00a01. The median age at diagnosis was 64 years (range 35-88 years). Most patients were male (69.5%,n= 139) and current or former smokers (46.0%,n= 92 and 40.5%,n= 81, respectively). The most common histological subtype was lung adenocarcinoma (81.0%,n=\\u00a0162), followed by lung squamous cell carcinoma (6.5%,n= 13). Most patients were diagnosed with stage IV (63.5%,n= 127) or III (23.5%,n= 47).Table\\u00a01Baseline characteristicsAll patientsn= 200 (%)Age, median (range), years64 (35-88)GenderMale139 (69.5)Female61 (30.5)Smoking statusNever smoker27 (13.5)Former smoker81 (40.5)Current smoker92 (46.0)ECOG PS030 (15.0)1135 (67.5)\\u2265235 (17.5)HistologyAdenocarcinoma162 (81.0)Squamous cell carcinoma13 (6.5)NOS10 (5.0)Other15 (7.5)Stage at diagnosisI11 (5.5)II15 (7.5)III47 (23.5)IV127 (63.5)PD-L1 status<1%78 (39.0)1%-49%52 (26.0)\\u226550%50 (25.0)Missing20 (10.0)Type of sampleBiopsy102 (51.3)Surgical sample53 (26.6)Cytology44 (22.1)Anatomical site of biopsyLung98 (49.0)Pleura/pleural effusion16 (8.0)Adenopathy44 (22.0)Central nervous system12 (6.0)Bone11 (5.5)Other19 (9.5)Open in a separate windowECOG PS, Eastern Cooperative Oncology Group performance status; NOS, not otherwise specified; PD-L1, programmed death-ligand\\u00a01.The origin of samples utilised for sequencing was diverse: 102 biopsy (51.3%), 44 cytology (22.1%) and 53 surgical specimens (26.6%). Most specimens were obtained through\\u00a0computed tomography-guided lung or bronchoscopy biopsies\\u00a0of the primary lung lesion (49.0%,n= 98), or by endobronchial ultrasound-guided transbronchial needle aspiration of a mediastinal lymphadenopathy (22.0%,n= 44). All samples were assessed by an expert lung pathologist and 178 specimens (89.0%) showed at least 20% of tumour cells.Library quality failure rateThe quality metrics of the DNA and RNA libraries were not reached (NR) only in 2 (1.0%) and 11 (5.5%) patients, respectively. The failure rate of RNA libraries was significantly higher in surgical specimens, accounting for 7 out of 53 samples (13.2%), compared with 3 out of 102 biopsies (2.9%) and 1 out of 44 cytological samples (2.3%;P= 0.017). The mean time from sample collection to molecular analysis was significantly longer for RNA libraries failed (20.1 versus 7.1 months;P= 0.021). Low tumour content in the tumour sample or the anatomical site of the biopsy was not associated with failure of RNA libraries. To evaluate the contribution of these factors, we conducted a logistic regression which showed that surgical specimens and longer time from collection were associated with a higher risk of RNA failure in the univariate analysis (Supplementary Table\\u00a0S1, available athttps://doi.org/10.1016/j.esmoop.2023.102197). However, in the multivariate analysis, none of the variables remained statistically significant.Detection of actionable genomic alterations with the TSO500 NGS panelOverall, we identified actionable GAs in 53 patients (26.5%) using the TSO500 NGS panel taking into consideration nine oncogenic drivers defined by the ESMO guidelines consisting ofEGFR, ALK, ROS1, BRAF, KRAS, MET, RET, NTRKandERBB2(Figure\\u00a01A). The most recurrent oncogenic alteration wasKRASmutation found in 55 (27.5%) tumour samples, and the G12C variant was observed in 19 (9.5%) patients (Supplementary Figure\\u00a0S1A, available athttps://doi.org/10.1016/j.esmoop.2023.102197).EGFRactivating mutations were found in nine (4.5%) tumour samples, consisting of five (2.5%) patients harbouring common sensitisingEGFRmutations (exon 19 deletion or L858R), one patient (0.5%) had anEGFR-S768Imutation, one patient (0.5%) with exon 20 insertion and two patients (1.0%) with multiple mutations. Additionally, two patients (1.0%) harbouredEGFRtruncating mutations that were not considered actionable.BRAFmutations were found in 16 (8%) tumour samples and the actionable V600E variant was seen in 5 (2.5%) tumours. Actionable gene fusions were detected in 11 (5.5%) patients, consisting ofALKin 3 (1.5%),RETin 5 (2.5%),ROS1in 2 (1.0%) andNTRK1rearrangement in 1 (0.5%) patient. All these alterations were mutually exclusive.METexon 14 skipping alterations were detected in two (1.0%) patients. Also, other potentially actionableMETalterations were found:METamplification in two patients (1.0%) and one patient (0.5%) harboured aCD47-METfusion.ERBB2actionable alterations were detected in seven (3.5%) patients: five (2.5%) exon 20 insertions, one (0.5%) exon 19 mutation and one (0.5%) exon 17 mutation. Furthermore,ERBB2amplification was detected in two patients (1.0%) and a non-actionable mutation in exon 24 was also found in another patient (0.5%).Open in a separate windowFigure\\u00a01Actionable alterations found in non-small-cell lung cancer (NSCLC) specimens using the TruSight\\u2122 Oncology 500 (TSO500) panel in this cohort of 200 patients.(A) Distribution of actionable driver alterations in clinically relevant genes according to the European Society for Medical Oncology (ESMO) guidelines. (B) Oncoplot showing all pathogenic or likely pathogenic genomic alterations in clinically relevant genes, regardless of being actionable or not. Missing tumour mutational burden (TMB) values are depicted in red.CN, copy number; indel, insertions and deletions; SNV, single nucleotide variants.On the contrary, two actionable alterations were detected in two patients with an allele frequency below 3%. These alterations were further confirmed by an orthogonal method and consisted of anEML4-ALKfusion (allele frequency 2.64%) which was validated by IHC, and aMETexon 14 skipping mutation (allele frequency 1.23%) confirmed by NGS in liquid biopsy. None of the samples had MSI and the median TMB was 10.2 mutations/Mb (range 0.0-119.7 mutations/Mb). An oncoplot showing the frequency of all pathogenic or likely pathogenic GAs observed in these genes is shown inFigure\\u00a01B, regardless of being actionable or not.Detection of additional clinically relevant genomic alterations with the TSO500 NGS panelPathogenic or likely pathogenic variants in seven genes associated with cancer predisposition were identified in 10 tumours (5.0%):MUTYH(n= 4, 2.0%),BRCA2(n= 2, 1.0%),PALB2(n= 1, 0.5%),BAP1(n= 1, 0.5%),SDHA(n= 1, 0.5%) and 1 patient (0.5%) with alterations in bothBRIP1andFLCN. More details about the variant allele frequency and the specific variants are provided inSupplementary Table\\u00a0S2, available athttps://doi.org/10.1016/j.esmoop.2023.102197.BRCA2mutations were confirmed at the germline level by Sanger sequencing. Both patients had familiar history of cancer and were referred to the Genetic Counselling Unit, where the variant was studied in multiple relatives who were carriers that will be followed up in the upcoming years. AllMUTYHmutations were found in heterozygosis and were not deemed candidates for subsequent germline testing. Two patients died before being referred to the Genetic Counselling Unit (one harbouring anSDHAmutation and another with aBAP1mutation).PALB2mutation was not validated by Sanger sequencing at the germline level andBRIP1mutation is still being studied.The utilisation of TSO500 panel was helpful to identify additional oncogenic alterations involving genes that have not been yet established as actionable targets in NSCLC, such asHRAS,PIK3CA and FGFR1(Supplementary Figure\\u00a0S1B, available athttps://doi.org/10.1016/j.esmoop.2023.102197). Specifically, anFGFR1fusion and anHRASmutation were detected in this cohort, but alsoPIK3CAmutations in nine patients (4.5%), of which five (2.5%) involved the catalytic domain. Those genes could be classified according to the ESCAT as ESCAT IIIA, since clinical benefit has been demonstrated in patients with these specific alterations albeit in other tumour types. Despite the limited evidence of clinical benefit of drugs targeting these GAs, their identification is relevant since there are current clinical trials or expanded access programmes ongoing that could potentially benefit those patients.An additional advantage of using a large panel is the ability to account for concurrent mutations in tumour suppressor genes, as for example the presence ofTP53, STK11orKEAP1mutations which is gaining clinical relevance in the context ofKRAS-mutant tumours. In this regard,\\u00a0we analysed which mutations are more likely to co-occur focusing on a predefined list of actionable drivers and tumour suppressor genes (Supplementary Table\\u00a0S3, available athttps://doi.org/10.1016/j.esmoop.2023.102197). Only concurrent mutations ofSTK11andKRASremained statistically significant, after adjusting for multiple comparisons.Comparison of TSO500 panel results with alternate single-gene testingSingle-gene testing was conducted before NGS in 131 patients (65.5%). Specifically, an RT\\u2013PCR for assessingEGFRmutation was carried out in 130 patients (65.0%), while IHC or FISH forALKandROS1rearrangement in 126 (63.0%) and 103 (51.5%) patients, respectively. Additional information about the single-gene testing is provided inSupplementary Table\\u00a0S4, available athttps://doi.org/10.1016/j.esmoop.2023.102197.We obtained discordant results between single-gene testing and NGS in four cases. Firstly, a patient with anEGFR L861Qmutation detected by RT\\u2013PCR, but not confirmed by NGS due to a DNA library failure. This patient received frontline osimertinib and achieved a partial response lasting for only 6 months. Secondly, a patient harboured anRETrearrangement diagnosed by FISH, which was not confirmed by NGS, despite testing the same specimen and the good quality of DNA and RNA libraries. This patient achieved a confirmed partial response to pralsetinib lasting for 12 months. Thirdly, a patient with a negative FISH forRETrearrangement, but with aKIF5B-RETfusion detected by NGS. This patient had an ongoing partial response to pralsetinib lasting for >18 months. Lastly, in a patient with a previously negativeROS1expression by IHC and anROS1-EZRfusion detected by NGS. This patient had an ongoing partial response to entrectinib lasting for >25 months. In all these cases, NGS and single-gene testing were conducted in the same specimen.Access to targeted therapy in patients with NSCLCWe analysed the impact on survival of having access to targeted therapies in this cohort, including not only 53 harbouring actionable oncogenic alterations detected by the TSO500 panel, but also two patients with low allele frequency alterations that were validated by orthogonal methods (ALKfusion andMETexon 14 skipping) and six additional patients who carried driver alterations that might be potentially actionable (METfusions and amplifications,ERBB2amplifications andFGFR1fusions). We also considered two additional cases detected only by single-gene testing (EGFRL861Q andRETrearrangement). Of those 63 patients, only 28 (44.4%) were eligible to receive matched targeted therapy. The access to targeted therapy is highly influenced not only by the specific GA and tumour stage but also by regulatory constraints. More details about the percentage of patients receiving targeted therapy according to the oncogenic driver are shown inTable\\u00a02.Table\\u00a02Access to targeted therapies in 63 patients harbouring an actionable oncogenic driver alteration detected by TSO500 panel or by other testing methodsAccess to targeted therapyReasons for not receiving targeted therapyClinical deteriorationNo access in the first lineClinical trial/expanded access not availableStage I-IIIn= 63 (%)28 (44.4)12 (19.1)12 (19.1)4 (6.3)7 (11.1)EGFRmutationsn= 10 (%)10 (100.0)\\u2014\\u2014\\u2014\\u2014NTRK1fusionn= 1 (%)1 (100.0)\\u2014\\u2014\\u2014\\u2014FGFR1fusionn= 1 (%)1 (100.0)\\u2014\\u2014\\u2014\\u2014RETfusionn= 6 (%)5 (83.3)0 (0.0)1 (16.7)0 (0.0)0 (0.0)ALKfusionn= 4 (%)2 (50.0)0 (0.0)0 (0.0)0 (0.0)2 (50.0)ROS1fusionn= 2 (%)1 (50.0)0 (0.0)0 (0.0)0 (0.0)1 (50.0)ERBB2alterationsn= 9 (%)3 (33.4)2 (22.2)2 (22.2)2 (22.2)0 (0.0)BRAF-V600En= 5 (%)1 (20.0)2 (40.0)0 (0.0)1 (20.0)1 (20.0)METalterationsn= 6 (%)1 (16.7)2 (33.3)0 (0.0)2 (33.3)1 (16.7)KRAS-G12Cn= 19 (%)3 (15.8)6 (31.6)8 (42.1)0 (0.0)2 (10.5)Open in a separate windowReasons for not receiving personalised treatment are described according to the type of driver alteration.TSO500, TruSight\\u2122 Oncology\\u00a0500.In this cohort, patients withEGFRmutations and those withRETorNTRK1rearrangements had a higher access rate to targeted therapy, compared to those withKRAS-G12C mutation,BRAF-V600E mutation,METandERBB2alterations. Thirty-five patients (55.6%) did not receive targeted therapy, despite harbouring a druggable GA. The main reasons for not receiving matched targeted therapy were clinical deterioration (n= 12, 19.1%), lack of access to specific therapy in the first line (n= 12, 19.1%), no therapies approved and lack of clinical trials or expanded access programmes (n= 4, 6.3%) or no evidence of metastatic disease (n= 7, 11.1%). Patients withALKandROS1rearrangements had a lower access rate to targeted therapy than expected because half received therapy with curative intent and were free of recurrence during the study follow-up.Survival analysis according to the presence of actionable oncogenic alterations and targeted therapy in advanced NSCLCWe defined three groups of patients based on the presence of actionable alterations and whether they received targeted therapy. The patients\\u2019 characteristics are shown inTable\\u00a03. Patients harbouring actionable alterations were more likely to be women, never smokers and had lower TMB. The number of treatment lines was higher among patients who received targeted therapy, while the proportion of patients who received chemotherapy and immunotherapy was similar between these three groups. The survival analysis was restricted to 183 patients with advanced NSCLC with sufficient clinicopathological information; among this group, median follow-up was 14.5 months (interquartile range 7.5-32.6 months). The median OS for all patients was 20.7 months [95% confidence interval (CI) 15-58.6 months]. Patients harbouring a druggable GA receiving a targeted therapy had significant longer median OS (not reached, NR) compared to patients without an actionable driver [27.7 months, 95% CI 15.02 months-NR; hazard ratio (HR)\\u00a0= 0.32; 95% CI 0.14-0.75;Figure\\u00a02]. Patients with an oncogenic driver not receiving a targeted therapy had a trend toward shorter median OS (10 months, 95% CI 9.18 months-NR) compared to patients without an actionable driver (27.7 months, 95% CI 15.02 months-NR; HR\\u00a0= 1.71, 95% CI 0.93-3.19).Table\\u00a03Association between clinicopathological variables with distinct subgroup of advanced patients based on the presence of genomic driver alteration (gDA) and access to targeted therapiesnNo gDA foundgDA found (untreated with targeted therapy)gDA found (treated with targeted therapy)Pvaluen= 127n= 28n= 28Age, median (range), years18363.90 (35.28-82.82)66.37 (38.64-87.92)59.57 (42.26-85.58)0.240Sex,n(%)1830.012Male94 (74.02)15 (53.57)14 (50.00)Female33 (25.98)13 (46.43)14 (50.00)Stage,n(%)1830.463I10 (7.87)0 (0.00)1 (3.57)II5 (3.94)1 (3.57)2 (7.14)III29 (22.83)4 (14.29)4 (14.29)IV83 (65.35)23 (82.14)21 (75.00)Histology,n(%)1830.120Adenocarcinoma99 (77.95)26 (92.86)23 (82.14)Squamous12 (9.45)0 (0.00)0 (0.00)Other16 (12.60)2 (7.14)5 (17.86)ECOG PS,n(%)1830.076021 (16.54)0 (0.00)7 (25.00)180 (62.99)22 (78.57)18 (64.29)\\u2265226 (20.47)6 (21.43)3 (10.71)Smoking status,n(%)183<0.001Never smoker6 (4.72)8 (28.57)9 (32.14)Former smoker54 (42.52)10 (35.71)10 (35.71)Current smoker67 (52.76)10 (35.71)9 (32.14)Smoking (pack-years), median (range)15140.00 (2.00-100.00)40.00 (10.00-76.00)27.00 (5.00-75.00)0.028PD-L1 category,n(%)1690.610<1%50 (43.10)12 (46.15)13 (48.15)1%-49%32 (27.59)8 (30.77)10 (37.04)\\u226550%34 (29.31)6 (23.08)4 (14.81)Sample type,n(%)1820.748Biopsy68 (53.54)15 (53.57)11 (40.74)Surgical piece33 (25.98)6 (21.43)9 (33.33)Cytology26 (20.47)7 (25.00)7 (25.93)Tumour purity,n(%)1830.275<20%10 (7.87)3 (10.71)5 (17.86)\\u226520%117 (92.13)25 (89.29)23 (82.14)TMB status,n(%)1810.002Low51 (40.48)17 (60.71)20 (74.07)High75 (59.52)11 (39.29)7 (25.93)Treatment lines received,n(%)1830.002029 (22.83)6 (21.43)0 (0.00)155 (43.31)15 (53.57)8 (28.57)221 (16.54)5 (17.86)14 (50.00)\\u2265322 (17.32)2 (7.14)6 (21.43)Received chemotherapy,n(%)1830.778No45 (35.43)8 (28.57)10 (35.71)Yes82 (64.57)20 (71.43)18 (64.29)Received ICI,n(%)1830.072No50 (39.37)15 (53.57)17 (60.71)Yes77 (60.63)13 (46.43)11 (39.29)Open in a separate windowECOG, Eastern Cooperative Oncology Group; ICI, immune checkpoint inhibitors; PD-L1, programmed death-ligand 1; TMB, tumour mutational burden.Open in a separate windowFigure\\u00a02Kaplan\\u2013Meier plot of overall survival in patients with advanced non-small-cell lung cancer (NSCLC) according to the presence or absence of an actionable alteration and its treatment with a targeted therapy.Pvalue from a Cox proportional hazards model adjusted for sex, stage, age, histology, smoking status and tumour mutational burden (TMB) status.gDA, genomic driver alteration; TKI, tyrosine kinase inhibitor.\",\n          \" Integrated analysis of genomic and transcriptomic data from patients with MIBC undergoing TMTThere were 92 patients who met initial criteria for inclusion in the study. Seventy-six cases passed quality control metrics for WES and comprise our final cohort (Supplementary Fig. S1). Clinical characteristics of the cohort are summarized inTable 1. The median age of the cohort was 72.3 years, with 24% female patients. All patients had MIBC (86% T2, 14% T3/T4) and the majority were treated with cisplatin-based chemotherapy (66%) concurrent with radiation (median dose 64.3 Gy, interquartile range, 60.7\\u201364.3).Table 1.Clinical characteristics of bladder TMT cohort.CharacteristicsNumber%Age at Diagnosis72.3 yr (IQR, 63.8\\u201377.8)SexMale5876Female1824Clinical T stageT26586T3811T434Concurrent chemotherapyCisplatin/5-FU3343Cisplatin/taxol1114Mitomycin/5-FU1114Cisplatin79Paclitaxel575-FU34Gemcitabine11NA57GradeModerate34Poor7396HydronephrosisPresent1114Absent6586Tumor-associated CISPresent1216Absent6484TURBTVisibly complete5978Visibly incomplete1722Pathologic response to CRTComplete5978Incomplete1722Radiation dose (median)64.3 Gy (IQR 60.7\\u201364.3)HistologyPure urothelial5572Squamous differentiation811Glandular differentiation68Micropapillary23Neuroendocrine differentiation11Sarcomatoid differentiation11Spindle cell11NA23Open in a separate windowAbbreviations: CIS, carcinomain situ; CRT, chemoradiation therapy; 5-FU, 5-fluorouracil; Gy, Gray; IQR, interquartile range; NA, not available; yr, years.The genomic tumor landscape of the cohort is summarized inFig. 1and Supplementary Fig. S2.Figure 1displays all genes comprising known oncogenic signaling pathways and known frequently mutated MIBC genes (Methods) with at least 5% mutation frequency. Mean target coverage was 109X for somatic tumor DNA (Supplementary Table S1). Mean nonsynonymous TMB for the entire cohort was 12.4 mutations/Mb (range, 3.4\\u201331.5). Overall,TP53was the most frequently mutated gene with 52.6% of samples having an alteration, followed byKMT2Dat 40.8%. Mutational signature analysis identified four active signatures: Signatures 1, 2, 5, and 13. Signature 1, present in 62/76 (81.6%) samples, is associated with age-associated spontaneous or enzymatic deamination of 5-methylcystosine to thymine (54). Signature 5, which is the signature previously associated withERCC2mutations in urothelial tumors and with smoking history, was present in 68/76 (89.5%) samples (55). Signatures 2 and 13, present in 44/76 (57.9%) and 39/76 (51.3%) cases, respectively, are associated with APOBEC-mediated mutations (54), and have been described previously in MIBC (45). There were significant differences in the relative contribution of mutations across mutational signatures (Supplementary Fig. S3; Kruskal\\u2013Wallis H,P< 0.001), similar to previous observations in MIBC cohorts (56). Signature 5 had a significantly higher contribution than other signatures (Wilcoxon rank sum,P< 0.001), but patients withERCC2-mutant tumors did not have higher contribution from signature 5 compared with patients withERCC2WT tumors (P= 0.813).Open in a separate windowFigure 1.Integrative molecular characterization of the MIBC cohort. Mutational signatures, TMB, gene expression molecular subtypes (GSC, TCGA, and MIBC consensus), transcriptomic T-cell\\u2013inflamed immune signature score, select clinical characteristics, and somatic alternations in select genes in pathways (Methods) for each tumor in the cohort (n= 76). TURBT complete refers to whether the initial TURBT was complete versus incomplete. Complete response defined as no visible tumor on cystoscopy and a negative tumor site rebiopsy following TMT after a period of induction therapy (16).Sixty-seven samples had transcriptomic data available (Fig. 1). Among the GSC subtypes (23, 48), Luminal was most common (21/67, 31.3%), while the Luminal papillary subtype was most common among TCGA subtypes (21/67, 31.3%), and the Basal squamous subtype was most common among the MIBC consensus class (19/67, 28.4%). The distribution of TCGA subtypes in our cohort was not significantly different from the distribution in the TCGA MIBC cohort (Supplementary Fig. S4; ref.45). Overall, the mean T-cell\\u2013inflamed signature score was 0.23 (range, \\u22120.04 to 0.98).To verify that the genomic data correlates with the transcriptomic data, we examined how a few genomic alterations associated with corresponding transcriptomic signature scores.FGFR3mutations are common in non\\u2013MIBC as well as in tumors of the upper tract, but are also present at lower frequencies in MIBC (57). In this cohort, 6 of 76 tumors (7.9%) harboredFGFR3mutations (Supplementary Fig. S5A). A lncRNA-based genomic classifier (58) revealed significant enrichment for FGFR3+ (\\u201cFGFR3 active\\u201d) cases inFGFR3mutant tumors (4/6 cases; 66.7%) compared withFGFR3WT tumors (4/61 cases; 6.6%,P< 0.0001). Similarly, tumors withFGFR3mutations also had higherFGFR3transcriptional signature scores (59) compared withFGFR3WT tumors (median 0.528 versus 0.191;P< 0.001; Supplementary Fig. S5B).Mutations inTP53were most common in the cohort (40/76 samples mutated). Gene expression profiling revealed higher hallmark p53 pathway scores (60) among those withTP53WT tumors compared with mutant (median 0.3593 versus 0.1171;P= 0.0001; Supplementary Fig. S6A). Separately, upon evaluation of a previously described novel TP53 mutation signature in prostate cancers (61), tumors withTP53mutations had higher signature scores compared with WT tumors (median 0.9945 versus 0.9786;P= 0.020; Supplementary Fig. S6B), consistent with prior associations (61).Genomic predictors of favorable mBI-EFS outcomesMedian follow-up for the cohort was 74.6 (range, 20.0\\u2013266.4) months in alive patients. Thirty-seven of 76 patients (49%) were characterized as having favorable long-term outcomes for mBI-EFS, and 39 (51%) as having unfavorable long-term outcomes for mBI-EFS (see study endpoints in Methods). Five-year mBI-EFS for the entire cohort was 45.4% (Supplementary Fig. S7A). Five-year BI-EFS for the entire cohort was 37.9% (Supplementary Fig. S7B), and 5-year OS for the entire cohort was 48.6% (Supplementary Fig. S7C). There were 19 locoregional failures and 22 distant failures.TMB has been shown to be associated with response to immune checkpoint inhibitors (62, 63), and TMB has also been suggested to be associated with radiation response in other cancers (64). We therefore evaluated whether there were differences in TMB in patients with MIBC with favorable compared with unfavorable clinical outcomes after TMT in our cohort. Interestingly, there was no significant difference in TMB between those with favorable (13.3 mutations/Mb, range 3.4\\u201331.5) and unfavorable mBI-EFS outcomes (11.5 mutations/Mb, range 6.4\\u201326.8),P= 0.1584 (Supplementary Fig. S8).We then performed an exploratory analysis of DDR genes as defined in the Alliance A031701 ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT03609216\\\",\\\"term_id\\\":\\\"NCT03609216\\\"}}NCT03609216) and RETAIN ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT02710734\\\",\\\"term_id\\\":\\\"NCT02710734\\\"}}NCT02710734) clinical trials of risk-adapted MIBC treatment to determine whether mutations in any of these 10 DDR genes correlated with clinical outcomes after TMT (Fig. 2A). The following mutations were identified in our cohort:ATM (n = 13), RB1(n= 17),FANCC(n= 4),ERCC2(n= 8),ATR(n= 9),BRCA1(n= 7),BRCA2(n= 11),RECQL4(n= 0),RAD51C(n= 0), andERCC5(n= 2; Supplementary Table S2). TMB was higher among DDR mutant tumors (13.4 mutations/Mb, range 3.4\\u201331.5) compared with DDR WT tumors (10.8 mutations/Mb, range 5.5\\u201324.7,P= 0.04). In the TCGA cohort which included patients with MIBC who underwent surgical resection (45), the presence of these DDR mutations was not significantly associated with improved survival [median months survival in patients with DDR mutations: 41.72 (95% confidence interval (CI), 26.91\\u201392.9) versus 33.11 (95% CI, 22.86\\u201354.86);P= 0.442; Supplementary Fig. S9]. However, in our TMT cohort, having any DDR gene mutation present was associated with improved mBI-EFS (HR, 0.51; 95% CI, 0.27\\u20130.97;P= 0.040;Fig. 2B), as well as improved locoregional control (HR, 0.31; 95% CI, 0.12\\u20130.80; log-rankP= 0.009;Fig. 2C). This was also consistent when evaluating cumulative incidence of locoregional failures (n= 19); patients without locoregional failure were enriched for the presence of any DDR gene mutation present [39/57 (68.4%) with mutations] compared with patients with locoregional failure [7/19 (37%) with mutations;\\u03c72P= 0.014]. There were no associations between DDR gene alterations and distant failures (n= 22 total distant failures, 11 among patients with DDR gene-altered tumors). In addition, there was no association between patients withTP53mutations and favorable or unfavorable mBI-EFS.Open in a separate windowFigure 2.Association of select DDR genes and long-term clinical outcomes following TMT.A,TMB, concurrent cisplatin with radiation, locoregional failure (LRF), mBI-EFS outcome (F, favorable; UF, unfavorable), along with select DDR gene alterations. Samples are ordered by favorable mBI-EFS outcome and unfavorable mBI-EFS outcome. Patients without locoregional failure were enriched for presence of any DDR mutation compared to patients with locoregional failure (\\u03c720.014). More patients with favorable mBI-EFS outcomes had mutations inERCC2compared with those with unfavorable mBI-EFS outcomes (\\u03c72P= 0.020).B,Having any DDR mutation present associated with improved locoregional control (HR, 0.31; 95% CI, 0.12\\u20130.80; log-rankP= 0.0088).C,Having any DDR mutation present associated with improved mBI-EFS (HR, 0.51; 95% CI, 0.27\\u20130.97;P= 0.004).Next, because missense mutations in DDR genes are often non-functional passenger events, we further evaluated whether the presence of known loss of function (LOF) mutations in DDR genes (as annotated in OncoKB; ref.46) impacts outcomes after TMT (Supplementary Tables S3 and S4). We found that patients without locoregional failure continued to be enriched for the presence of any DDR (LOF) mutation present [26/57 (66.7%) with mutations] compared with patients with locoregional failure [2/19 (11%) with mutations;\\u03c72P= 0.006]. Patients with favorable mBI-EFS outcomes were enriched for the presence of anERCC2LOF mutation [4/37 (11%) with mutations] compared with patients with unfavorable mBI-EFS outcomes [1/39 (3%) with mutations;\\u03c72P< 0.001].We then evaluated the proportion of patients with favorable mBI-EFS outcomes after TMT who would have met DDR gene mutation-specific inclusion criteria for either of the risk-adapted bladder-sparing clinical trials that are testing whether chemotherapy alone is sufficient to omit upfront local therapy (cystectomy or radiation) in patients with MIBC with DDR gene mutations (Supplementary Fig. S10A and S10B). Upon applying these inclusion criteria [Alliance A031701 included any truncating mutations including frameshift, nonsense, splice site in 8 genes (ATM, ATR, FANCC, BRCA1/2, RECQL4, RAD51C, ERCC5) or any mutation inERCC2; RETAIN included any alteration present inATM, RB1, FANCC, orERCC2], it was found that 20/37 (54%) patients with favorable mBI-EFS had DDR gene mutations necessary for eligibility for one or both trials (9 RETAIN only, 2 Alliance only, 9 both trials), while 17/37 (46%) patients did not meet any of the mutation eligibility criteria for these trials (Supplementary Fig. S10A; Supplementary Table S4). Thus, nearly half of patients who had favorable mBI-EFS outcomes after bladder-sparing TMT would not have been eligible for participation in chemotherapy-only\\u2013based risk-adapted bladder preservation clinical trials.Of the DDR gene alterations evaluated,ERCC2mutations stood out as being significantly associated with favorable mBI-EFS outcomes (n= 7) as compared with unfavorable outcomes (n= 1, locoregional failure; Fisher exactP= 0.027;Fig. 2A). ERCC2 is a DNA helicase that plays an important role in the NER pathway (65). Upon examination of the distribution of 8ERCC2mutations, 7 were located within one of the two conserved helicase domains (Fig. 3A; Supplementary Table S5). The distribution ofERCC2mutations in this cohort reflects the overall mutation distribution present in the TCGA bladder cohort (Supplementary Fig. S11; ref.45). Six of the 8 patients withERCC2mutations received cisplatin-based chemotherapy concurrently with radiation (all with favorable mBI-EFS outcomes); of the 2 that did not, 1 received mitomycin C, which also creates DNA lesions that require the NER pathway for repair of DNA damage, and was also associated with favorable mBI-EFS outcome. The 1 patient with an unfavorable mBI-EFS outcome received concurrent 5-FU. Therefore, among patients who received cisplatin-based chemoradiotherapy, all 6 patients with an ERCC2 mutation had favorable mBI-EFS outcomes. Patients withERCC2mutations had significantly improved mBI-EFS (HR, 0.15; 95% CI, 0.06\\u20130.37;P= 0.030;Fig. 3B), and improved BI-EFS, an endpoint that includes all-cause mortality (HR, 0.33; 95% CI, 0.15\\u20130.68;P= 0.044;Fig. 3C). A trend was also noted toward improved OS in patients withERCC2mutations (HR, 0.36; 95% CI, 0.17\\u20130.80;P= 0.075;Fig. 3D). There was no significant difference in TMB amongERCC2mutation tumors (14.0 mutations/Mb, range 4.2\\u201324.3) versusERCC2WT tumors (12.2 mutations/Mb, range 3.4\\u201331.5;P= 0.38).Open in a separate windowFigure 3.ERCC2mutations associate with favorable mBI-EFS outcomes following TMT.A,Lollipop figure ofERCC2gene depicting the location of missense mutations in the 6 patients with a favorable mBI-EFS outcome (black), as well as the location of the missense mutation in the patient with an unfavorable mBI-EFS outcome (red). One patient with a favorable outcome had I17T and Y24C compound mutations (denoted by bracket). One patient with a favorable outcome had a splice site mutation (not shown).B,Patients withERCC2mutations had improved mBI-EFS (HR, 0.15; 95% CI, 0.06\\u20130.37;P= 0.030).C,Patients withERCC2mutations had improved BI-EFS, which includes all-cause mortality (HR, 0.33; 95% CI, 0.15\\u20130.68;P= 0.044).D,Patients withERCC2mutations had a trend toward improved OS (HR, 0.36; 95% CI, 0.17\\u20130.80;P= 0.075).The mBI-EFS endpoint includes cystectomy due to treatment toxicity as an event, and in our cohort, 2 patients had their bladders removed due to toxicity (Supplementary Table S4). To evaluate whether excluding these events would affect our analysis, we evaluated a new endpoint that does not include cystectomy for toxicity (called \\u201cre-modified BI-EFS\\u201d). We found that having any DDR mutation was significantly associated with improved re-modified BI-EFS (HR, 0.51; 95% CI, 0.26\\u20130.99;P= 0.040; Supplementary Fig. S12) and patients withERCC2mutations had significantly improved re-modified BI-EFS (HR, 0.15; 95% CI, 0.06\\u20130.37;P= 0.037; Supplementary Fig. S13).ERCC2mutations were associated with a Luminal papillary subtypeWe then evaluated the transcriptional landscape ofERCC2mutant tumors compared withERCC2WT tumors. Patient samples withERCC2mutations (n= 6 with transcriptomic data available) had differential enrichment of several transcriptomic signature scores and molecular subtypes compared withERCC2WT tumors (n= 61;Fig. 4A). Samples withERCC2mutations had higher hallmark DNA repair scores (ref.60; median 0.110 versus \\u22120.318;P= 0.020), consistent with dysregulation of DDR (Supplementary Fig. S14). In addition,ERCC2mutant tumors were enriched for TCGA Luminal papillary subtype compared withERCC2WT (5/6, 83% versus 26% Luminal papillary;P= 0.004;Fig. 4B). These same patients were also enriched for GSC Luminal subtype compared withERCC2WT tumors (4/6, 67% versus 28% Luminal;P= 0.050, Supplementary Fig. S15).ERCC2mutant tumors were not associated with higher or lower T-cell\\u2013inflamed signature scores compared withERCC2WT tumors.Open in a separate windowFigure 4.Gene expression subtypes and signature scores associate withERCC2mutations.A,TCGA and GSC expression subtypes in the TMT cohort with transcriptomic data (n= 67), hallmark DNA repair signature score, andERCC2alterations in a TMT cohort.B,ERCC2mutation tumors were enriched for TCGA Luminal papillary subtype compared withERCC2WT (83% versus 26% Luminal papillary,P= 0.004).Functional characterization ofERCC2mutations in response to cisplatin and radiationWe next tested the functional impact ofERCC2mutations on cisplatin and radiation sensitivity in bladder cancer cell lines. KU19\\u201319 and SW1710 are bladder cancer cell lines with WTERCC2, while KE1 and SC14 are derivatives of KU19\\u201319 and SW1710, respectively, withERCC2mutations introduced by CRISPR-mediated gene editing (52). We have previously shown that theERCC2-mutant KE1 and SC14 lines are NER-deficient and have increased cisplatin sensitivity compared with their respective WTERCC2parental lines (51, 52). To model the impact of ERCC2 functional loss on cisplatin-based chemoradiotherapy response, we compared the sensitivity of the KU19\\u201319/KE1 and SW1710/SC14 cell pairs to varying doses of combined cisplatin and radiation treatment. As expected, theERCC2-mutant lines were significantly more sensitive to cisplatin monotherapy than the WTERCC2parental lines (Fig. 5A\\u2013D). The addition of sublethal doses of cisplatin to radiation did not significantly impact sensitivity in the parental KU19\\u201319 and SW1710 cell lines; however, there was a significant increase in sensitivity with dose enhancement factors of 1.55 and 1.40 when cisplatin was added to radiation in theERCC2-mutant SC14 and KE1 cell lines, respectively (Fig. 5E\\u2013J). The ERCC2 mutant clones demonstrated mild increase in radiation sensitivity compared with WT, but this was much less pronounced than the increase in cisplatin sensitivity (Fig. 5Eand\\u200bandF).F). Taken together, these data demonstrate that the presence of a deleteriousERCC2mutation is sufficient to drive sensitivity to combined radiation and cisplatin treatment in bladder cancer preclinical models.Open in a separate windowFigure 5.ERCC2mutations confer sensitivity to cisplatin-based chemoradiotherapy in bladder cancer cell lines.A,Representative images of clonogenic cell survival following treatment with cisplatin and ionizing radiation (IR) over a range of concentrations and doses in SW1710 (ERCC2 WT) and SC14 (ERCC2 mutant, mut) bladder cancer cell lines.B,Representative images of clonogenic cell survival following treatment with cisplatin and IR for KU19\\u201319 (ERCC2 WT) and KE1 (ERCC2 mut) bladder cancer cell lines.CandD,Clonogenic survival fractions for SW1710 versus SC14 and KU19\\u201319 vs. KE1 cell lines following cisplatin treatment at the concentrations shown.E,Clonogenic survival fractions for ERCC2 WT cell lines following treatment with IR with versus without cisplatin at \\u223cIC70 (SW1710) or \\u223cIC50(KU19\\u201319) concentrations. Data points were fitted with the LQ formula with the IR dose range determined by the sensitivity of the ERCC2 mut lines to combined IR/cisplatin.F,Analogous to PanelE, clonogenic survival fractions of ERCC2 mut cell lines using cisplatin at \\u223cIC70 (SC14) and \\u223cIC50(KE1) concentrations. Dose enhancement factors (DEF) for cisplatin were estimated at 0.5 survival fraction.G\\u2013J, Synergy analysis was performed over the full range of treatment doses and concentrations (Combenefit). All data points represent three independent repeats. Comparison of data points byttest, two-sided, and comparison of fitted survival curves by F-test, two-sided (GraphPad Prism 6.1). NS, not significant.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"discussion\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 24,\n        \"samples\": [\n          \" In this study, we proved the feasibility of using the large hybrid capture-based NGS panel TSO500 to characterise the GAs in a cohort of 200 patients with NSCLC. Our study demonstrates that this NGS panel, even after conducting single-gene testing forEGFRandALK, was able to detect genomic druggable alterations in 26.5% of the patients, with major impact in treatment decisions. This result is consistent with previous works in advanced NSCLC. Drilon et\\u00a0al. sequenced 31 samples from patients diagnosed with NSCLC without any alteration by single-gene testing and using a hybrid capture NGS panel was able to detect a druggable alteration in 26% of patients.13Suh et\\u00a0al. analysed 6832 NSCLC samples using hybrid capture NGS panel and found a druggable alteration in 33.6% of the samples.26Noteworthy, when this study was published, neitherKRAS-G12C mutation norNTRKrearrangement was considered actionable alterations. In a more recent study, Perdrizet et\\u00a0al. evaluated 134 patients (107 not harbouringEGFRmutations orALKrearrangements by single-gene testing) using the Oncomine Comprehensive Assay NGS panel and found a potentially actionable alteration in 31% of patients. This cohort was more enriched for females (59.7%) and never smokers (53.0%) than our series.27In our study, the failure rate of the TSO500 panel was very low and was mostly related to the unsuccessful preparation of the RNA libraries. Failure of the RNA libraries was more likely in surgical specimens and in older samples, as previously reported.28RNA library failure may reduce the ability to detect gene rearrangements.28,29A previous work validating the TSO500 panel using 170 FFPE tumour samples from distinct tumour types and specimens, including a small\\u00a0subset of NSCLC, reported a 7% failure rate in the RNA library preparation, which was slightly higher than in our study. The analytical sensitivity and specificity were high for\\u00a0detecting SNVs, indels, CNVs, MSI and gene rearrangements.30A recent study assessed the feasibility of the TSO500 panel using Diff-Quick cytology smears from 27 patients with NSCLC, but sequencing failed in 5 cases (1 due to library preparation failure and 4 due to low median coverage).31In our study, we obtained discordant results between NGS and single-gene testing only in four patients, despite using the same specimen for the analysis. AnEGFRmutation was not detected due to failure in the DNA library and oneRETfusion found by FISH was not validated by NGS. These discordances occur in the real clinical practice and should always be discussed in the multidisciplinary molecular tumour board.A significant number of patients with actionable GAs were able to receive matched targeted therapy in this series. As previously reported, patients receiving tailored targeted therapies had improved survival outcomes.3,4The major reasons for not getting access to targeted therapy were diverse, but some were avoidable in principle such as the lack of access to personalised therapies especially in the frontline setting due to hurdles in the drug approval process.When using large NGS panels to sequence somatic tumour DNA, it is likely to detect incidental findings as pathogenic variants in cancer susceptibility genes. In our study, we found pathogenic variants in those genes at an allele frequency \\u226520% for indels or \\u226530% for SNVs in 12 patients (6%). Those cases were discussed in our multidisciplinary molecular tumour board and we followed the ESMO recommendations24to manage those cases and decide which patients should be referred to the Genetic Counselling Unit.32Our study has several limitations. Firstly, the single-gene testing overlap may introduce a bias selection in our cohort. During the first period of this study, only patients withoutEGFRmutations andALKorROS1rearrangements were sequenced; this partly explains the lower rate of these alterations observed in our study. Secondly, our study was carried out at a single institution with a limited sample size but included consecutive cases. The limited sample size did not allow us to assess the prognostic impact of concurrent mutations such asTP53mutations inEGFR-orALK-positive patients receiving targeted therapy orSTK11-KEAP1co-mutations inKRAS-mutant patients receiving immunotherapy. Moreover, the study population was heterogeneous and received a wide range of different treatments. This variability could yield biases in the survival analysis. Finally, the limited access to specific targeted therapies which are not reimbursed in our country could restrain the expected survival benefit obtained by NGS.In conclusion, the utilisation of the TSO500 hybrid capture NGS panel is feasible and highly informative in clinical practice, with reasonable turnaround time and low failure rate in this cohort of patients with NSCLC. Additionally, detection of oncogenic driver alterations and access to targeted therapies have a major impact on patients\\u2019 life expectancy. Ensuring access to NGS testing and to targeted therapy is currently essential to treat patients with NSCLC.\",\n          \" The RPA heterotrimer is an essential protein for binding ssDNA encountered in cellular transactions to facilitate DNA-DNA and DNA-protein interactions during DNA replication, repair, recombination, RNA transcription, and telomere maintenance. As such, mutations in this genome maintenance protein have been linked to cancer formation in mice (35) and are acquired in up to ~1% of human cancers (14). We recently demonstrated that heterozygous germlineRPA1mutationsRPA1c.680T>C p.V227A, c.718G>A p.E240K and c.808A>G p.T270A in DBD-A are associated with TBD, which predisposes to hematologic and solid tumors. In this study, one patient with RPA1-related TBD developed MDS (10). Based on these data, we reasoned that germline defects in RPA1 and possibly also the other 2 components of the RPA heterotrimer (RPA2 and RPA3) might be associated with cancer development. To this end, we investigated comprehensive germline genomic data for the presence of heterozygous variants inRPA1,RPA2andRPA3across a large series of pediatric hematologic, solid and CNS malignancies. We discovered significant enrichment of ultra-rare and novelRPA1germline variants in our pediatric cancer cohort compared to non-cancer controls, positioningRPA1as a novel candidate predisposition gene. Moreover, in an additional cohort of 41 patients with AML, we identified 3 heterozygous germlineRPA1variants (c460G>A, p.T154A; c.1397C>G, p.A466G; c.1538G>A, p.R513H) with potential pathogenic effect.RPA1harbored the most variants likely due to its larger size compared toRPA2andRPA3. Although we did not observe a statistically significant enrichment of putative damaging variants inRPA2andRPA3, some of the identified variants were novel or ultra-rare and could possibly have a deleterious effect. Thus,RPA2andRPA3could be considered as genes of unknown significance (GUS) yet potentially important in tumor formation. All 3 proteins are required to fold properly to form a functional RPA heterotrimer (13). For this reason, we calculated stabilities of the RPA modular domains harboring mutations to gain insight into the possible effect of identified germline variants on RPA heterotrimer function. Scores greater than 1.5 are highly predictive of protein instability and dysfunction. High protein folding scores were found for 4 uniqueRPA1variants in 4 cases, 3 identified in patients with B-ALL and one in a patient with ependymoma. All were either ultra-rare or novel with CADD scores suggesting high likelihood of pathogenicity. TwoRPA3variants also harbored high protein folding scores in patients with B-ALL. This suggests that dysfunctional folding of the RPA heterotrimer may lead to genomic instability in these patients.In our discovery cohort, we identified 5 AML cases with germlineRPA1variants. One had an ultra-rareRPA1p.L58F variant in DBD-F and the remaining 4 had variants affecting nucleotide 856 within DBD-A domain (c.G856A, p.V286I in 3 cases and c.G856T, p.V286F in one case). The resulting amino acid changes do not differ in size or charge from wild-type valine and have a neutral protein folding score of 0.6. However, these mutations may disrupt protein-protein, protein-DNA interactions, or post-translational modifications, which are known mechanisms implicated in pathogenicity ofRPA1variants in various experimental models (10\\u201313,35). Additionally, TBD-associated pathogenicRPA1variants, p.V227A, p.E240K and p.T270A, have protein folding scores of 1.4, 0.1 and 0.2 (consistent with normal protein folding shown in biochemical assays) yet were shown to exert gain-of-function effect on DNA binding and melting of telomeric G-quadruplexes (10). Three of the 4 AML cases withRPA1variants in DBD-A domain had additional germline variants in genes (NOTCH2, FANCD2, MLL, HIP1) which, together withRPA1may have an epistatic effect to cause overall genomic instability. Corroborating data from a small cohort of 41 AML patients in which 3 patients carriedRPA1variants (p.T154A in linker region; p.A466G and p.R513H in DBD-C) deserves further investigation. BeyondRPA1in the AML cohort, we also found a novel germline missense variant inRPA3in an infant with AML who also harbored a germline truncating variant in the DNA helicase,RTEL1, which is associated with TBD (36,37). More functional studies are needed to determine the pathogenicity ofRPA1V286I/F alterations and their role in hematologic malignancy.Among the 13 CNS tumors with variants in RPA heterotrimer genes, 9 cases were high grade neoplasms, including medulloblastoma and high-grade glioma. Interestingly, 3 of the 5 medulloblastoma cases had novel and one very rare germlineRPA1, as well as one ultra-rareRPA3variant. Notably, even though variants in other unrelated genes were also found in 4 of the 5 medulloblastoma cases, none of these genes have been previously associated with medulloblastomas in the literature. Other studies have identified germline defects in DNA repair genes in medulloblastoma (38,39). It would stand to reason that germline mutations in the RPA heterotrimer, which functions in almost all DNA repair pathways, could potentiate oncogenic transformation. Further investigation should focus on assessing the function of RPA mutant proteins in DNA repair and their contribution to tumor biology.Our study has several limitations. Although all cases were assessed using a uniform pipeline, the cohort is skewed towards cases with B-ALL (~4-fold higher number of B-ALL compared to solid and CNS cancers). We included all germline and somatic mutations per case that were reported in previously published studies; however, this information was unavailable for a proportion of cases and therefore we cannot make definitive conclusions aboutRPAvariants being the sole germline driver in these cancers. Although ultra-rare and novel heterozygous germline variants inRPA1were significantly enriched in pediatric cancers, it is difficult to ascertain pathogenicity and clinical relevance without functional follow-up, which falls beyond the scope of this study. It is plausible that variants with high in-silico protein folding energy, ultra-rare and/or novel allelic frequency and high pathogenicity scores may be clinically relevant and should be among the top variants to explore in future studies.In summary, evasion of DNA repair mechanisms is a common theme among cancers. RPA is an essential protein for DNA replication and repair. Our study describes novel and rare variants with potentially deleterious effect in theRPA1,RPA2andRPA3genes in pediatric malignancies. Moreover, we have identified enrichment ofRPA1variants in cancer cases compared to non-cancer controls, suggesting that this gene potentially acts as a novel cancer driver. We plan to exploit our findings and perform further functional and biochemical characterization of recurrent cancer associatedRPA1variants to assess their potential use as targets for future cancer therapies.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"methods\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12,\n        \"samples\": [\n          \" Study participantsBlood samples from 220 female breast cancer patients from the indigenous Arab population unselected for early age of onset or family history of cancer were collected from 2 participating institutions in Eastern Saudi Arabia: King Fahd Hospital - Alhafouf and King Fahd University Hospital - Dammam. All individuals in this study consented to institutional review board-approved protocols that allowed for comprehensive genetic analysis of germline samples.Sequencing and library preparationAll samples (n=220) prepared for lpWGS had sufficient starting material (100 ng of double-stranded gDNA). Normalized DNA was fragmented (Covaris sonication) to 350 bp and then ligated to specific adapters during automated library preparation (Roche/KAPA, Hyper KK8504) using the Beckman FXp liquid handling robot. Libraries were pooled in equal volume and sequenced on an Illumina nano flow cell to estimate each library\\u2019s concentration based on the number of index reads per sample. Library construction is considered successful if the yield is larger than or equal to 250. All samples were successful. Libraries were normalized, pooled, and sequenced using Illumina platforms. Pooled samples were demultiplexed using the Picard tools version 1.130 [15].For whole-exome sequencing (WES), a total amount of 1.0\\u03bcg genomic DNA per sample was used as input material for the DNA sample preparation. Whole-exome capture libraries were generated using Agilent SureSelect Human All ExonV6 kit, and fragmentation was carried out by a hydrodynamic shearing system (Covaris, Massachusetts, USA) to generate 180\\u2013280bp fragments. Products were purified using the AMPure XP system (Beckman Coulter, Beverly, USA) and quantified using the Agilent high-sensitivity DNA assay on the Agilent Bioanalyzer 2100 system. The qualified libraries were fed into Illumina sequencers after pooling according to their effective concentration and expected data volume. All case samples had satisfactory effective read rates (> 97%) and error rates (< 0.03%) and are included in further analysis.AlignmentAll raw sequencing data were uploaded to Terra (https://firecloud.terra.bio/), a collaborative cloud-computing platform utilized for genomic analyses, developed as part of the NCI Cloud Pilot program and supported by the Broad Institute [16]. Using Genome Analysis Toolkit (GATK) version 4.1.8.1 [17], all FASTQ files were first converted into unaligned Binary Alignment Map (uBAM) files, then aligned to the human reference genome b38 using BWA (version 0.7.15), as recommended by the GATK best practice workflows [18].Sequencing coverageThe average sequencing coverage of all lpWGS and WES samples was calculated using the GATK\\u2019s (version 3.7) tool \\u201cDepthofCoverage\\u201d. A sample-wide mean coverage of 0.1\\u00d7 was considered the minimum acceptable coverage for lpWGS, and a 15\\u00d7 average coverage over exon intervals was considered the minimum acceptable coverage for WES.Whole-exome variant callingDeepVariant (version 1.0.0) [19], a deep learning-based variant calling method that has been demonstrated to have superior performance at detecting pathogenic variants compared to the standard joint-genotyping approach [20,21], was used to call germline variants from WES data (docker image: gcr.io/deepvariant-docker/deepvariant:1.0.0). All variants annotated with \\u201cPASS\\u201d in the FILTER column of the VCF were deemed high quality. Variants passing QC from all samples were then merged into one VCF file using GATK\\u2019s (version 3.7) tool \\u201cCombineVariants\\u201d. Subsequently, the \\u201cvt\\u201d tool (version 3.13) was used on the cohort VCF file to normalize and decompose multiallelic variants.Functional and clinical annotation of germline variantsThe cohort VCF file was annotated using Variant Effect Predictor (VEP, release 104.3) [22] with the publicly available GRCh38 cache file with a custom plug-in to include a recent \\u201cClinVar\\u201d database release (accessed in June 2021). Using the tier criteria used by the Catalogue of Somatic Mutation in Cancer (COSMIC) [23], only variants in \\u201cgermline tier 1\\u201d genes were considered. All detected variants were then classified into five categories: benign, likely benign, variants of unknown significance, likely pathogenic, and pathogenic, using the American College of Medical Genetics (ACMG) guidelines [24]. Moderately or highly penetrant variants classified as likely pathogenic or pathogenic are then collectively referred to as pathogenic variants (PV).Low-pass whole genome imputationTo obtain variant calls from lpWGS, GLIMPSE v1.1.1 [25] was used to perform genome-wide variant imputation. Following the recommended steps, the genome-wide genotype likelihood was first calculated on each sample using bcftools then separated into smaller genomic intervals before imputation. To maximize the number of variants imputed, we used Eagle v2.4.1 [26] to computationally phase the publicly available 1000 Genome [27] (1KG) WGS VCFs that were called using DeepVariant [28] (v1.0.0, GLnexus v1.2.7, GRCh38 reference), and used the output as the reference panels for imputation. After imputation was carried out on each genomic chunk, they were combined using the \\u201cGLIMPSE_ligate\\u201d command with default arguments, producing the final imputed VCF.Imputed variant quality controlTo assess variant imputation accuracy and to select a proper filtering threshold, the concordance of exonic variants was calculated based on the intersection of variants called both by DeepVariant using WES data and imputed by GLIMPSE using lpWGS. The \\u201cINFO\\u201d score outputted by GLIMPSE, which is a value that ranges from 0 to 1 where 1 indicates high confidence in the variant call, is referred to as the imputation quality score. The linear transformation of the posterior genotype probabilities generated by imputation, which is a number ranging from 0 to 2 where a number close to 1 indicates confidence in a single alternate allele at the location while a number close to 2 reflects confidence in having 2 alternate alleles at the location, is referred to as variant dosage. Variants were binned based on minor allele frequency, and the correlation between variant dosage and the number of alternate alleles (0, 1, or 2) called by DeepVariant, referred to as dosage correlation, was calculated within each bin for every sample. Allele frequencies were calculated based on allele counts in the cohort.Relatedness inferenceTo control for confounding effects from related individuals, PLINK 1.9 [29] was first used to extract biallelic single-nucleotide polymorphisms (SNPs) from the merged WES VCF file. Subsequently, LDAK 5.2 [30] was used to compute a kinship matrix assuming the LDAK-Thin heritability model with a correlation squared threshold of 0.98 and window size of 100 kb, as recommended (https://dougspeed.com/calculate-kinships/). Samples were then removed until no pairs have a kinship value greater than 0.125. Five samples were removed after this step.Polygenic risk score calculationTo assess the clinical applicability of PRS, we adopted a similar PRS calculation methodology proposed by Hao L. et al. [31] and curated the initial sets of PRS weights from \\u201cCancerPRSWeb\\u201d [32], a repository that contains PRS coefficients for major cancer traits derived from multiple large population databases such as the UK BioBank (UKB) [33], Michigan Genomics Initiative (MGI) [34], and GWAS Catalog [1]. To pick PRS sets most relevant to breast cancer, we selected \\u201cBreast Cancer [Female]\\u201d as the cancer trait and manually curated 20 sets of non-subtype-specific weights which had validation performance in either MGI or UKB. The number of SNPs in the selected weights ranged from 79 to 1,120,410, and they were derived using various methods with different performances in UKB or MGI, as measured by area under the receiver-operator characteristic curve (AUCpopulation). After downloading the associated weight file and metadata, SNPs with hg19 coordinates were lifted over to hg38 using the python liftover library [35] for downstream compatibility. For each set of PRS weights, we calculated the unadjusted raw PRS in PLINK 1.9 [29] by using the \\u201c--score\\u201d command with the \\u201cscore-no-mean-imputation\\u201d option enabled.PRS population stratification adjustmentTo obtain the genetic principal components (PCs) of every sample, we first merged the Arab breast cancer cohort WES VCF with the WES VCF from the 1000 Genomes Project [27]. The merged WES VCF was then loaded using Hail v0.2 [36] and filtered for variants with allele frequency > 0.05 andp-value greater than 1e\\u22126 from the Hardy-Weinberg Equilibrium test. LD-pruning was then performed on passing variants with greater than 0.1 correlation within a 1 million base pair window. The Hail function \\u201chwe_normalized_pca\\u201d was then applied to the resulting set of common variants (n=58,286), and the top 10 PCs were kept for further analysis.To create a population-adjusted PRS, an ordinary least square model was fitted using the top 10 PCs as features with the raw PRS as the output variable. The difference between the predicted PRS and raw PRS was then standardized, creating the population-adjusted, residualized PRS. This process was then repeated for every set of PRS weights, and the CancerPRSWeb ID of the PRS weights with the highest performance at detecting breast cancer in first-degree relatives (AUCfamily) as well as any degree relatives (AUCfamily-any) was \\u201cPRSWEB_PHECODE174.1_Onco-iCOGS-Overall-BRCA_LASSOSUM_MGI_20200608.\\u201dStatistical analysisUnless otherwise specified, all odds ratios, 95% confidence intervals, andp-values were computed based on the two-sided Fisher\\u2019s exact test, as implemented in the exact2x2 R package [37] with the argument \\u201cminimum likelihood correction.\\u201d Confidence intervals of the area under the receiver-operator characteristic curves (AUC) were calculated based on the formulation by J. Hanley and B. McNeil [38]. Statistical diagrams were visualized using Seaborn v0.11.2 [39]. Statistical models were constructed using the python package \\u201cstatsmodel\\u201d [40]. The clinical characteristics table was constructed using the package \\u201ctableone\\u201d [41]. The effect of PRS standard deviation increase on the age of onset is estimated by fitting a generalized linear model (GLM) using the statsmodel package with standardized ancestry-adjusted PRS as the feature and age of presentation as the outcome. The coefficient of the PRS and its 95% confidence intervals as indicated by the GLM is then reported.To evaluate the change in PRS AUC after removing pathogenic variant carriers from the cohort, thep-values were obtained by calculating the proportion of samples in which the AUC was lower after removing pathogenic variant carriers using 10,000 bootstrapped samples of the analysis cohort.\",\n          \" Study cohortCandidate study participants with a personal history of tumour-associated phenotype(s) were identified through 11 Australian FCCs (located within Queensland, New South Wales, Victoria, South Australia, Tasmania and Western Australia) between 2017 and 2020 as part of their standard clinical care. For the purposes of this study (hereafter referred to as the ICCon study), a tumour-associated phenotype was inclusive of both primary cancers and cancer gene syndrome-associated benign tumours. Specific inclusion criteria applied for this study are detailed in Table\\u200bTable1;1; however, if the individual met these inclusion criteria solely on the basis of a family history of: breast cancer, ovarian cancer (epithelial), colorectal cancer (in absence of unexplained polyposis), melanoma (in populations with high occurrence of melanoma), or prostate cancer, they were deemed ineligible. Eligibility was determined by the FCC clinicians and/or genetic counsellors, with additional consultation by the ICCon study leads if required. A total of 195 individuals (hereafter referred to as index cases) accepted enrolment into the study (Additional file1: Table S1). The tumour-associated phenotype diagnoses for index cases were categorised and recorded into divisions as per Table\\u200bTable2.2. Tumour-associated phenotypes in the same tissue type and organ were considered separate primary diagnoses if the medical record denoted them as distinct.Table 1Eligibility criteria for recruitment of index casesEligibility criteriaIndex cases (number / percent)dAn index case in a family that fulfils clinical criteria for a hereditary cancer syndrome28 (14.4)An index case with two primary cancers (either synchronous or metachronous) diagnosed at 60\\u00a0years or younger83 (42.6)An index case with three primary cancers diagnosed at 70\\u00a0years or younger52 (26.7)Index case with colorectal polyposis defined as per InSiGHTaor eviQbtesting guidelines forMUTYHand/orAPC29 (14.9)A cancer-affected individual fulfilling two or more of the following criteria:\\u2022 Early age at onset (under 40\\u00a0years for adult cancer or\\u2009<\\u200910\\u00a0years younger than the mean age of cancer diagnosis for that tumour type and sexc)\\u2022 One or more first-degree relatives with the same kind of cancer\\u2022 Two or more first- or second-degree relatives with different cancers, where at least one cancer is very rare42 (21.5)Open in a separate windowaThe International Society for Gastrointestinal Hereditary Tumours (InSiGHT), as perhttps://www.insight-group.org/beviQ, as perhttps://www.eviq.org.au/cAustralian Institute of Health and Welfare 2017. Cancer in Australia 2017. Cancer series no. 101. Cat. no. CAN 100. Canberra: AIHWdPercent does not add to 100% as index cases may have met more than one eligibility criteriaTable 2Characteristics of the recruited index casesDemographicIndex casesNumberPercentageSexFemale14172.3Male5427.7Age at first diagnosis18 and under168.219\\u20134912564.150 and over5427.7Prior genetic testingUninformative resulta14071.8None5528.2Recruitment siteQLD4623.6VIC5829.7TAS147.2SA2110.8NSW2914.9WA2713.8Study referralMeets direct eligibility requirements15780.5Eligible via independent panel assessment3819.5Virtual panel usedPanel A (107 genes)2412.3Panel B (101 genes)17187.7Broad tumour-associated phenotypebGynaecological and breastBreast4322.1Endometrial126.2Ovarian94.6Cervical42.1Other gynaecological21.0UrogenitalRenal cell carcinoma3517.9Prostate63.1Urothelial carcinoma31.5Other urogenital136.7NeuroendocrineThyroid2914.9Non-thyroid2613.3PolyposisAdenomas or adenomatous polyposis2814.4Serrated polyps63.1Hyperplastic polyps63.1Hamartomatous or Juvenile polyps10.5Other polyps63.1GastrointestinalColorectal2814.4Upper gastrointestinal115.6Skin cancerMelanoma2110.8Non-melanoma147.2SarcomaSoft tissue105.1Bone42.1Other sarcoma31.5Respiratory126.2Head and neck84.1Brain84.1Lymphoma84.1Non-endocrine pancreatic73.6Unknown primary31.5Non-brain neurological10.5Other4322.1Open in a separate windowaInclusive of immunohistochemistry (IHC) analysis of somatic tissue and other somatic tissue testing in addition to germline genetic testingbPercentage may not add to 100% as index cases may have been diagnosed with more than one tumour-associated phenotypeFor each index case, there was either a clinical suspicion of a cancer predisposition syndrome, but there were no recommended tests for that specific presentation, or previous routine genetic testing had not identified a germline molecular genetic diagnosis for their presenting tumour-associated phenotypes. A further 19 family members of 12 index cases were also recruited to aid in the clinical interpretation of germline variants identified in the index cases. Study data were collected and managed using REDCap electronic data capture tools [10,11]. Changes in risk management strategies were recorded in REDCap [10,11] as per Additional File1: Table S2. Written informed consent was provided by all participants, or by the relevant next of kin where the index case was under 18\\u00a0years or deceased at time of recruitment.Virtual gene panelOur analysis was restricted to a comprehensive virtual panel of cancer predisposition genes. The initial panel (panel A) comprising 107 genes was compiled by the ICCon study working group members from a larger panel described previously [12]. During the ICCon study, a revised panel of 101 genes was adopted (panel B). This revision of panel A was triggered following: review at theFamilial Aspects of Cancerconference in 2018; consideration of the study conducted by Tudini et al. [12]; and acknowledgement of suggestions made during previously conducted multidisciplinary team (MDT) meetings. Panel A was used for the analysis of the first 24 participants, whilst panel B was used for the remaining 171 participants (Additional File1: Table S3 and S4).Whole-genome sequencingGermline DNA was extracted from peripheral blood, obtained as per the standard care practices of the participant\\u2019s respective FCC. DNA (1.2 to 10\\u00a0\\u00b5g) was supplied to The Kinghorn Centre for Clinical Genomics at the Garvan Institute of Medical Research (Sydney, Australia). Sequence libraries were generated from 500\\u00a0ng DNA using the KAPA Hyper PCRFree Library Preparation kit and sequenced using 150\\u00a0bp paired-end sequencing reads on the Illumina HiSeq X Ten platform (Illumina, San Diego, CA, USA) to a targeted read depth of 30X. FASTQ files were processed and analysed at the QIMR Berghofer Medical Research Institute. Sequence reads were trimmed using Cutadapt version 1.9 [13] and aligned to the GRCh37 reference genome using BWA-MEM from BWA Kit version 0.7.12 [14]. Duplicate alignments were marked with Picard (version 1.129,http://picard.sourceforge.net), and BAM files were coordinate-sorted using Samtools version 1.8 [15]. Mean read depth across each participant\\u2019s germline genome was determined using NGSCheck (genomiQa Pty. Ltd.). Mean read depth per genome was 34.3X\\u2009(range 27.8X\\u2009to 58.3X).Germline variant identification and prioritisationAnalysis was conducted in an ongoing manner as participants were recruited and sequenced. Single-nucleotide variants (SNVs) were identified using a dual calling strategy using GATK HaplotypeCaller [16] and qSNP [17]. Small insertions and deletions (indels) were identified using the GATK HaplotypeCaller. See Github public code repository under the AdamaJava project (https://github.com/AdamaJava) for details regarding software packages. SNVs and indels were annotated using the Ensembl Variant Effect Predictor (VEP) (v92.3) [18] for gene loci association and other bioinformatics predictions, with publicly available and in-house plugins. Unless present in ClinVar [19] with a likely pathogenic or pathogenic classification (as per the Clinvar May 2018 XML release), variants were disregarded as follows: variants with a minor allele frequency (AF) of\\u2009>\\u20091% in any of the major subpopulations (n\\u2265\\u20092000 individuals) of gnomAD (v2.1.1) [20], considering exome-derived and genome-derived allele frequencies separately;\\u2009>\\u20091% frequency within the ICCon study cohort; or\\u2009>\\u20091% in-house germline whole-genome dataset (n=\\u2009312). SNVs and indels with a variant allele fraction of less than 30 or 20% respectively were removed from further analysis.Structural variants (SV, inversions, insertions and intra/inter-chromosomal translocations) and copy-number variants (CNV, deletions and duplications) were detected using DELLY (downloaded 20180903) [21] with default parameters (as per the recommended germline SV calling pipeline). CNVs were detected using CNVnator (downloaded 20190820) [22], with a bin size of 100 and the \\u2018ngc\\u2019 option given. CNVs called by CNVnator with a q0 value of\\u2009<\\u20090 or\\u2009>\\u20090.5 (unless the genotype score was\\u2009<\\u20090.25) were flagged for removal. CNVs called by only a single tool were flagged for manual review. Overlap between tools was defined as the same CNV type with\\u2009\\u2265\\u200980% reciprocal overlap. SVs and CNVs that overlapped (\\u2265\\u200980% for deletions, duplications and inversion; or any overlap for translocations and insertions) with exclusionary genome regions (consisting of the ENCODE Duke and DAC poor mappability regions [23], repetitive and low complexity regions [24], GRCh37 \\u2018N\\u2019 repeat regions, and regions where mapping quality frequently falls below MAPQ of 10) were flagged for exclusion. Likewise, variants with a\\u2009>\\u20095% AF within any of the gnomAD SV dataset (version 2.1) [25], in-house germline dataset or within-cohort were also flagged for exclusion. The in-house germline dataset frequencies were calculated from variants called using the same methods used for the ICCon study cohort. Overlap for frequency assignment was\\u2009\\u2265\\u200980% for deletions, duplications and inversions, whilst the frequencies of variants within a 100-bp window were considered for translocations and insertions. Overlaps were determined using BEDTools versions 2.27.1 and 2.29.0 [26].\\u00a0SV and CNV events were associated with genes via annotation with ENSEMBL (version 75) protein-coding gene footprint region (defined as 2000\\u00a0bp upstream of the transcription start position to the transcription end position). Any degree of overlap between a CNV and gene locus was considered. For other SVs, overlap of one or both breakpoints with a gene locus was considered.Variant classification and reportingAll variants initially prioritised as described above were manually reviewed using the Integrative Genome Viewer (IGV) (version 2.3.97) [27]. These variants were then curated using ACMG/AMP guidelines [28]. Variants classified as being of uncertain significance (VUS), likely pathogenic (LP) or pathogenic (P) were reported in the form of a preliminary research report to an MDT meeting for clinical evaluation (Additional File1: Table S4). Variants that did not meet strict ACMG/AMP criteria for benign (B) or likely benign (LB) classification were not included in the preliminary research reports for MDT meeting evaluation if they were as follows: a missense variant with a REVEL [29] score less than 0.5; a synonymous variant with no predicted impact on splicing; or an in-frame indel with a PROVEAN [30] score greater than\\u2009\\u2212\\u20092.5. However, these exclusionary conditions were not applied to the variant: if there was an apparent gene-phenotype association; an alternative bioinformatics predictor (such as CADD\\u2009>\\u200920 [31]) suggesting the variant to be deleterious; and/or the variant was present in ClinVar (at the time of manual review during curation) with a LP/P classification from at least one submitter. All variants classified as LP/P, even if for recessive conditions and/or conditions that did not clearly associate with the index cases\\u2019 presenting phenotype/s were considered for MDT meeting review. The MDT meeting panel was comprised of the following: the project coordinator; two or more variant curation and/or genomics researchers; and the treating clinician(s) and/or genetic counsellor(s). Variants were critically reviewed within the context of available clinical data, family history, bioinformatic and literature-derived evidence available at the time of clinical review until a consensus regarding the variant\\u2019s pathogenicity was reached. In some cases, supporting evidence was not sufficient to clearly classify the variant and further laboratory testing and/or expertise to assist with the interpretation, was sought to reach a consensus in the classification. Once a consensus was reached, a post-MDT meeting research report, summarising reportable variants, was issued to the treating clinician. For post-MDT meeting classification, VUS were further classified into three tiers following a system similar to that adopted by the Victorian Clinical Genetics Services [7,32]. Namely, VUS:A\\u2014available evidence is highly suggestive that the variant is disease-causing; VUS:B\\u2014available evidence is insufficient to classify the variant as either disease-causing or benign; and VUS:C\\u2014available evidence is highly suggestive that the variant is likely benign. Variants downgraded to B/LB during the MDT meeting were not returned to the treating clinician (Additional File1: Table S4). A pathogenic variant identified in two participants, I129 and I134, was returned without formal MDT meeting review following consultation with their Familial Cancer Clinic and lead study team. Clinical confirmation of variants reported post-MDT meeting was conducted by a NATA-accredited laboratory where the treating clinician wished to consider implementation of changes to risk management strategies.Health economic analysisWe undertook an analysis to compare the costs and key outcome (overall diagnostic yield) associated with three hypothetical testing scenarios (Fig.1). Scenario 1 describes a typical current approach to testing where affected individuals referred to FCCs are offered targeted gene or gene panel testing when specific gene(s) are indicated. Scenario 2 describes an approach where the option of WGS plus a virtual MDT meeting is offered to affected individuals in whom no gene(s) are initially indicated by an FCC, and to affected individuals in whom gene(s) were initially indicated but targeted gene or gene panel testing identified no LP/P variants. Scenario 3 describes an approach where all affected individuals referred to FCCs are offered WGS: where non-complex cases are offered WGS with follow-up by an FCC, and complex cases are offered WGS plus a virtual MDT meeting. A scenario using an agnostic panel testing approach for all cases was not assessed.Open in a separate windowFig. 1Scenarios used as the basis of the health economic analysis. Scenario 1: Standard testing; Scenario 2: standard testing followed by whole-genome sequencing (WGS); Scenario 3 upfront WGS. FCC\\u2014familial cancer centre; MDT\\u2014multi-disciplinary team meetingAs our approach to the economic analysis is \\u2018programmatic\\u2019 (meaning we estimated the total likely cost of implementing each testing scenario as a national programme, taking into account \\u2018overheads\\u2019 to coordinate the programme as well as the resources required to conduct each test, including staff time and sequencing costs, Additional File1: Table S5), we took a national, public health system perspective. All monetary values are expressed in Australian dollars. All resource utilisation estimates were costed using 2020 prices and pay rates, and inflation or deflation of costs was not applied. As all costs and benefits for a case were defined as accruing within the same 12-month period, no discounting was applied.Estimation of costsThe estimated costs included in the analysis account for staff time within FCCs to provide counselling before and after testing (where it is assumed that post-test costs are higher for individuals with positive test results, Additional File1: Table S5), targeted gene and gene panel costs (expressed as an average cost per tested individual based on the relative distribution of cancer presentations to FCCs), WGS costs (based on data from the ICCon study), virtual MDT meeting costs (based on utilisation data from the ICCon study) and other costs associated with administering the WGS plus virtual MDT meeting testing approach for complex versus non-complex cases (Additional File1: Table S5).When identifying relevant costs from the ICCon study, we have been careful to exclude resource use that is specifically associated with a research setting (for example, ethics approvals and reporting), and to include resource use that would be expected to occur in clinical practice, outside of a research setting. The latter costs include the steps involved in genomic sequencing (that is, obtaining the appropriate biological sample(s), DNA extraction, library preparation, sequencing, analysis, data storage, and preparation of reports for clinicians [33]), clinician time to discuss and interpret the reports (whether that is time within or outside an MDT meeting), and programme co-ordination (such as, communication with FCCs, and co-ordination of virtual expert panel members).The units of resource use observed in ICCon were averaged per genome, and then unit costs were applied. For the labour items, the monetary values used represent true costs to the health system, whereas the non-labour costs associated with genomic sequencing represent the prices charged by providers. Labour costs have been estimated by applying relevant pay rates from the University of Queensland and Queensland Health, noting that these are representative of pay scales in other Australian States and Territories. Capital costs associated with genomic sequencing are not included as a separate line item as it is assumed these would be captured in the prices charged by the providers.Uptake and diagnostic yield of testing scenariosThe uptake and diagnostic yield of testing applied in each scenario are shown in Fig.1and the total costs associated with each scenario are shown in Additional File1: Tables S6 and S7. The proportion of affected individuals who are offered targeted gene or gene panel testing, and the proportion of those who accept this testing, were taken from State-wide de-identified and aggregated data from the FCCs in Queensland (provided by Queensland Health directly to the research team). The diagnostic yield for targeted gene or gene panel testing was taken from the same source.The proportions of affected individuals who are offered and who undergo WGS in Scenario 2 were taken from the ICCon study. Separate proportions have been derived for (a) affected individuals who initially had gene(s) indicated and underwent targeted gene or gene panel testing with no LP/P variants identified, and (b) affected individuals who initially had no gene(s) indicated and had not been offered targeted gene or gene panel testing. The diagnostic yield for each of these patient subgroups ((a) and (b)) was also derived from the ICCon study. For affected individuals who would undergo upfront WGS (Scenario 3) assumptions have been made regarding the rate of uptake and the diagnostic yield, informed by results from the Queensland TrakGene data study and the literature.The total number of affected individuals referred to all FCCs in Queensland in a year was available, but corresponding data for other States and Territories was not available. Consequently, the total number of affected individuals referred to all FCCs in Australia in a year has been estimated by scaling up the data from Queensland, using the populations from each jurisdiction.Sensitivity analysesOne-way sensitivity analyses were undertaken for selected parameters (Additional File1: Table S6): applying a lower cost for WGS in Scenarios 2 and 3 ($1366 instead of $1750; based on a recent quote from a sequencing provider for a similar service); assuming all cases offered WGS in Scenario 3 will require a virtual MDT meeting (virtual MDT meetings only applied to complex cases in the base case); applying a higher rate of uptake of WGS in Scenarios 2 (96.8% instead of 79.6%); applying lower diagnostic yields for upfront WGS in Scenario 3 (17.3% or 28.2%, versus 38.4% in the base case); and assuming a lower and higher number of affected individuals are referred to an FCC, across all scenarios (50% and 150% of base case).\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cluster_id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1,\n        \"min\": 0,\n        \"max\": 4,\n        \"num_unique_values\": 5,\n        \"samples\": [\n          4,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 27,\n        \"samples\": [\n          \"Molecular profiling and feasibility using a comprehensive hybrid capture panel on a consecutive series of non-small-cell lung cancer patients from a single centre\",\n          \"NTHL1is a recessive cancer susceptibility gene\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_cluster = df.loc[df['cluster_id'] == 0]\n",
        "df_cluster_1 = df.loc[df['cluster_id'] == 1]\n",
        "df_cluster_2 = df.loc[df['cluster_id'] == 2]\n",
        "df_cluster_3 = df.loc[df['cluster_id'] == 3]\n",
        "df_cluster_4 = df.loc[df['cluster_id'] == 4]"
      ],
      "metadata": {
        "id": "Q_ZA9oYj4VEJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df['cluster_id'].unique()"
      ],
      "metadata": {
        "collapsed": true,
        "id": "m3wbR6h_3IdD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e420f5a6-2837-4a88-c4db-dddf990860f7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([0, 4, 3, 2, 1])"
            ]
          },
          "metadata": {},
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_cluster.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BU9ag7595Z8j",
        "outputId": "e947591d-efaa-44a3-ff39-bca6f535fe90"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-30-6ba4ab6b41a8>:1: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_cluster.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_cluster_1.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n",
        "df_cluster_2.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n",
        "df_cluster_3.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n",
        "df_cluster_4.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "s2RwaV5Daigd",
        "outputId": "3062385e-38a3-4157-ff93-8dbaa98bf9a1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-39-c03283b3da61>:1: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_cluster_2.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n",
            "<ipython-input-39-c03283b3da61>:2: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_cluster_3.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n",
            "<ipython-input-39-c03283b3da61>:3: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  df_cluster_4.rename(columns={'abstract':'some', 'introduction': 'abstract'}, inplace=True)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_cluster.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 414
        },
        "id": "chCbc-4x7Ioo",
        "outputId": "a302e5d8-3131-4b33-8605-2a7a5c0bfa5b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                 some  \\\n",
              "0    Germline pathogenic variants associated with ...   \n",
              "4                                                 NaN   \n",
              "5                                                 NaN   \n",
              "11                                                NaN   \n",
              "13   Head and neck squamous cell carcinomas (HNSCC...   \n",
              "\n",
              "                                             abstract  \\\n",
              "0    With the rise of the genomic era, a growing n...   \n",
              "4    Characterising the genomic landscape of somat...   \n",
              "5                                                 NaN   \n",
              "11   Prostate cancer (PCa) is characterised by sub...   \n",
              "13   Head and neck squamous cell carcinoma (HNSCC)...   \n",
              "\n",
              "                                              results  \\\n",
              "0    pCPS genes with monoallelic risk phenotypes s...   \n",
              "4                                                 NaN   \n",
              "5    A framework for measuring enrichment of germl...   \n",
              "11   Known risk alleles in Black South Africans wi...   \n",
              "13                                                NaN   \n",
              "\n",
              "                                           discussion  \\\n",
              "0    Here we have explored the theory that AD(LoF)...   \n",
              "4                                                 NaN   \n",
              "5    We noted an enrichment of rare germline varia...   \n",
              "11   In this study, motivated by limited studies h...   \n",
              "13                                                NaN   \n",
              "\n",
              "                                              methods  cluster_id  \\\n",
              "0    In order to ensure reproducibility the full, ...           0   \n",
              "4                                                 NaN           0   \n",
              "5    Tumor sequencing data from PCAWGThe final con...           0   \n",
              "11   Ethics approvals and recruitmentWritten infor...           0   \n",
              "13                                                NaN           0   \n",
              "\n",
              "                                                title  \n",
              "0                                                 NaN  \n",
              "4   COSMIC: a curated database of somatic variants...  \n",
              "5   Systematic analysis of Mendelian disease-assoc...  \n",
              "11  Prostate cancer genetic risk and associated ag...  \n",
              "13  Saliva as a potential non-invasive liquid biop...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e0d69513-94e1-4f44-b829-8c77b6f2e62f\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>some</th>\n",
              "      <th>abstract</th>\n",
              "      <th>results</th>\n",
              "      <th>discussion</th>\n",
              "      <th>methods</th>\n",
              "      <th>cluster_id</th>\n",
              "      <th>title</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Germline pathogenic variants associated with ...</td>\n",
              "      <td>With the rise of the genomic era, a growing n...</td>\n",
              "      <td>pCPS genes with monoallelic risk phenotypes s...</td>\n",
              "      <td>Here we have explored the theory that AD(LoF)...</td>\n",
              "      <td>In order to ensure reproducibility the full, ...</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Characterising the genomic landscape of somat...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>COSMIC: a curated database of somatic variants...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>A framework for measuring enrichment of germl...</td>\n",
              "      <td>We noted an enrichment of rare germline varia...</td>\n",
              "      <td>Tumor sequencing data from PCAWGThe final con...</td>\n",
              "      <td>0</td>\n",
              "      <td>Systematic analysis of Mendelian disease-assoc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Prostate cancer (PCa) is characterised by sub...</td>\n",
              "      <td>Known risk alleles in Black South Africans wi...</td>\n",
              "      <td>In this study, motivated by limited studies h...</td>\n",
              "      <td>Ethics approvals and recruitmentWritten infor...</td>\n",
              "      <td>0</td>\n",
              "      <td>Prostate cancer genetic risk and associated ag...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Head and neck squamous cell carcinomas (HNSCC...</td>\n",
              "      <td>Head and neck squamous cell carcinoma (HNSCC)...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "      <td>Saliva as a potential non-invasive liquid biop...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e0d69513-94e1-4f44-b829-8c77b6f2e62f')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-e0d69513-94e1-4f44-b829-8c77b6f2e62f button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-e0d69513-94e1-4f44-b829-8c77b6f2e62f');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-221800a2-d87f-428e-87b0-869a29bebcb0\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-221800a2-d87f-428e-87b0-869a29bebcb0')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-221800a2-d87f-428e-87b0-869a29bebcb0 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_cluster",
              "summary": "{\n  \"name\": \"df_cluster\",\n  \"rows\": 7,\n  \"fields\": [\n    {\n      \"column\": \"some\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \" Head and neck squamous cell carcinomas (HNSCCs) are the most devastating diseases in India and southeast Asia. It is a preventable and curable disease if detected early. Tobacco and alcohol consumption are the two major risk-factors but infection of high-risk HPVs are also associated with development of predominantly oral and oropharyngeal carcinomas. Interestingly, unlike cervical cancer, HPV-induced HNSCCs show good prognosis and better survival in contrast, majority of tobacco-associated HPV\\u2212veHNSCCs are highly aggressive with poor clinical outcome. Biomarker analysis in circulatory body-fluids for early cancer diagnosis, prognosis and treatment monitoring are becoming important in clinical practice. Early diagnosis using non-invasive saliva for oral or other diseases plays an important role in successful treatment and better prognosis. Saliva mirrors the body's state of health as it comes into direct contact with oral lesions and needs no trained manpower to collect, making it a suitable bio-fluid of choice for screening. Saliva can be used to detect not only virus, bacteria and other biomarkers but variety of molecular and genetic markers for an early detection, treatment and monitoring cancer and other diseases. The performance of saliva-based diagnostics are reported to be highly (\\u226595\\u00a0%) sensitive and specific indicating the test's ability to correctly identify true positive or negative cases. This review focuses on the potentials of saliva in the early detection of not only HPV or other pathogens but also identification of highly reliable gene mutations, oral-microbiomes, metabolites, salivary cytokines, non-coding RNAs and exosomal miRNAs. It also discusses the importance of saliva as a reliable, cost-effective and an easy alternative to invasive procedures.Keywords:Head and neck cancer, Saliva, Non-invasive, Liquid biopsy, Biomarker, Early detection, HPV screening\",\n          \" Germline pathogenic variants associated with increased childhood mortality must be subject to natural selection. Here, we analyze publicly available germline genetic metadata from 4,574 children with cancer [11 studies; 1,083 whole exome sequences (WES), 1,950 whole genome sequences (WGS), and 1,541 gene panel] and 141,456 adults [125,748 WES and 15,708 WGS]. We find that pediatric cancer predisposition syndrome (pCPS) genes [n\\u2009=\\u200985] are highly constrained, harboring only a quarter of the loss-of-function variants that would be expected. This strong indication of selective pressure on pCPS genes is found across multiple lines of germline genomics data from both pediatric and adult cohorts. For six genes [ELP1,GPR161,VHLandSDHA/B/C], a clear lack of mutational constraint calls the pediatric penetrance and/or severity of associated cancers into question. Conversely, out of 23 known pCPS genes associated with biallelic risk, two [9%,DIS3L2andMSH2] show significant constraint, indicating that they may monoallelically increase childhood cancer risk. In summary, we show that population genetic data provide empirical evidence that heritable childhood cancer leads to natural selection powerful enough to have significantly impacted the present-day gene pool.Subject terms:Cancer genomics, Evolutionary genetics, Paediatric cancer, Cancer genetics\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"abstract\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \" Characterising the genomic landscape of somatic mutations in cancer is a challenge that has been approached from several different directions. It includes carrying out large sequencing projects like TCGA (1), which characterized over 20\\u00a0000 primary cancer and matched normal samples spanning 33 cancer types. An extension of this work was the ICGC/TCGA Pan-Cancer Analysis of Whole Genomes project (PCAWG) (2), which entailed the whole genome analysis of 2658 tumours and matched normal tissues, representing 38 cancer types and seeking to identify common patterns of mutation. An example of a more recent initiative is AACR Project GENIE (3), which is a publicly accessible cancer registry of real-world clinico-genomic data assembled through data sharing between 19 leading international cancer centres.COSMIC integrates somatic data from multiple sources published around the world and allows researchers to access and scrutinise information about somatic mutations and their impact in cancer. Over the past two decades, COSMIC has been diligently collecting, cleaning, and organising genomic data and associated metadata from cancer studies published in scientific literature and various bioinformatics sources. This data is then translated into a standardised format, integrated, and made available to the research community through well-structured datasets and user-friendly data exploration websites and tools (4,5). COSMIC aligns with the mission and values of the Wellcome Sanger Institute that are dedicated to advancing our understanding of human biology and improving human health. By collating and disseminating this wealth of cancer-related data, COSMIC empowers scientists worldwide to leverage existing research findings for the benefit of future studies and, ultimately, cancer patients.Started in 2004 as a census of somatic mutations in four cancer genes (6), COSMIC has broadened its approach to cover the entire human genome and all types of cancer. Due to the huge complexity of acquired data, numerous ways of reporting it in publications, and a vast number of data classification standards and formats, data curation in COSMIC is performed by an in-house team of doctoral level curators with developed expertise in cancer genetics, assuring high accuracy, quality, and biological relevance of the datasets.Somatic mutations and their impact on cancer development remain central to COSMIC. Furthermore, the catalogue of genetic alterations is supported by data from other omics analyses, including gene expression and DNA methylation variants. In addition to the main catalogue of somatic mutations, a further 6 accompanying resources focus on different aspects of oncology (Figure\\u200b(Figure1).1). The Cancer Gene Census (CGC) (7) and Cancer Mutation Census (CMC) provide additional annotations regarding the roles of genes and mutations in oncogenesis, which are based on a defined set of rules and sufficient evidence obtained through dedicated literature curation and analysis of the content of the core catalogue. COSMIC 3-D visualises mutation prevalence in the context of the protein structure. Mutational Signatures explore the molecular origins of mutations at genome level. Finally, Actionability and the catalogue of mutations causing drug resistance address the availability of therapeutic options for cancer and how their efficacy depends on the genetic profile of the tumour. Jointly, these complementary resources are intended to add layers of additional knowledge, show the biological impact of somatic mutations in the context of the broader spectrum of processes driving cancer, and inform about ways of targeting molecular alterations in clinical practice.Open in a separate windowFigure 1.COSMIC\\u2019s 7 key resources for understanding cancer and improving cancer patient care. The main catalogue of somatic mutations is supported by further six resources that together lay additional layers of knowledge helping to interpret the impact of somatic mutations on cancer development and presenting available therapeutic options.These curation efforts have resulted in a FAIR-compliant (8) resource that is available through a dedicated website (https://cancer.sanger.ac.uk/cosmic) and as set of data export files. The resource brings together data and metadata from more than 1.5 million patient samples and over 1000 cancer cell lines, acquired from over 29000 scientific publications and expanded by open-access data from large studies such as TCGA (1) and ICGC (9).\",\n          \" Thyroid cancer is common, with both incidence and mortality rates increasing.1,2Most malignant thyroid lesions are indolent, well-differentiated tumors such as papillary or follicular thyroid carcinomas (PTC and FTC, respectively). These tumors can be successfully treated with surgical resection of the thyroid followed by radioactive iodine.3,4However, many thyroid cancers recur or progress after such standard-of-care treatment. Rarely, recurrence can even occur decades after initial therapy and may involve distant metastasis or transformation to poorly differentiated (PDTC) or anaplastic thyroid cancer (ATC).5ATC is one of the most lethal cancers in existence. The median survival of patients with ATC is just 3\\u20135\\u00a0months\\u2014indicating a dire need to better understand drivers of thyroid cancer progression.6DNA and RNA sequencing have revolutionized our understanding of many tumors, providing insight into underlying biology and drivers of aggressive disease. However, the genomic understanding of thyroid cancer has lagged behind other tumors. Until recently, thyroid cancer molecular testing was used almost exclusively for diagnostic purposes at the time of initial biopsy.3,7Genomic studies have identified common driver mutations in\\u00a0well-differentiated thyroid cancers, allowing commercial genomic classifiers to distinguish malignant from benign lesions with high sensitivity, specificity, and accuracy.8,9,10,11,12,13BRAFV600E andRASmutations are mutually exclusive and represent the most common driver alterations in well-differentiated thyroid cancers. The frequency of these alterations has led to the molecular classification of thyroid cancers as eitherBRAF-like orRAS-like, based on the gene expression patterns ofBRAFV600E andRAS-mutant PTCs.14Other alterations detected by diagnostic molecular tests include gene fusions such asNTRK1/3,PAX8/PPARG, andRET; copy number alterations; microRNA dysregulation; and gene expression abnormalities.15,16,17Despite these advances in molecular diagnostics, thyroid cancer prognostication and post-surgical treatment are still largely guided by clinical and histopathologic features. This cancer management is in stark contrast to the biomarker-driven personalized management of many other cancers, such as non-small cell lung cancer.18,19The lack of molecular biomarker testing for thyroid cancer is due to our limited understanding of the drivers of advanced disease. While patients withBRAF-like tumors have slightly worse outcomes (5% mortality inBRAF-mutant PTC versus 1% inRAS-mutant PTC),20the majority ofBRAF-like thyroid cancers have an excellent prognosis. Recent large sequencing studies have implicatedTP53,PIK3CA, andTERTpromoter (TERTp) mutations in aggressive thyroid cancer, particularly in combination withBRAFV600E.21,22,23,24,25,26This work has led to the development of the first commercial molecular-based test for high-risk disease.27Despite this tremendous advance, there are still patients with recurrent, metastatic, and de-differentiated thyroid cancer that lack high-risk mutations. Additional biomarkers are needed to further enhance the risk-stratification and management of patients.The role of the tumor microenvironment in cancer progression is an active area of investigation and has led to many advances in prognostication and therapy.28However, there has been limited research in the microenvironment of thyroid cancer. To this end, recent work identified a subgroup ofBRAF-like lesions enriched in cancer-associated fibroblasts (CAFs) that may have more aggressive behavior.29Another study showed that in thyroid tumors driven byBRAFV600E mutation andPTENloss, fibroblasts may promote progression by remodeling collagen in the tumor microenvironment.30Other research has suggested roles of infiltrating immune cells such as macrophages in supporting aggressive thyroid cancer behaviors.31,32As such, clinical trials are ongoing to evaluate the efficacy of checkpoint inhibitor therapy for anaplastic thyroid carcinoma.33,34,35,36,37,38However, our knowledge of the stromal microenvironment across thyroid tumors remains limited.In this study, we utilize a large patient cohort, enriched for aggressive thyroid tumors, from two tertiary medical centers and perform DNA/RNA sequencing, spatial transcriptomics, and multiplex immunofluorescence in order to identify and visualize biomarkers of aggressive thyroid malignancy. In doing so, we identify high-risk mutations as well as key stromal elements including CAFs and macrophages that are strongly associated with aggressive tumors and can be detected at the initial thyroid surgery. We generate a molecular signature, the Molecular Aggression and Prediction (MAP) score, which identifies tumor, stromal, and immune microenvironmental factors associated with aggressive disease. We demonstrate the successful use of this score, in combination with mutational profiling, to identify patients at risk for future aggressive thyroid cancer. Finally, we show the utility of this score in subdividing aggressive tumors that may respond to immunotherapy. Future clinical evaluation of the stromal microenvironment could potentially inform patient management and ultimately reduce thyroid cancer-related mortality.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"results\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \" A framework for measuring enrichment of germline variants in cancerTo estimate the contribution of pathogenic variants in OMIM genes to cancer systematically, we developed a statistical method that tests the excess of pathogenic variants in cancer patients relative to healthy control samples.We compared the frequency of pathogenic variants in 2,642 cancer patients from the PCAWG [12] study to 2,504 healthy control exomes from the 1\\u00a0KG project [14] (Fig.1b and Additional file1: Table S1). Rigorous quality control steps were applied to the two distinct datasets, and we only considered variants in regions with sufficient and comparable sequencing coverage in both PCAWG and 1\\u00a0KG samples (seeMethods; Additional file2: Fig. S1). Despite applying filtering steps to both datasets, some technical differences might remain due to the different variant-calling pipelines.Germline variants were categorized as rare pathogenic variants if they exhibited a minor allele frequency (MAF)\\u2009<\\u20090.5% in the Genome Aggregation Database (gnomAD) exome [20] (seeMethods; Fig.1b). Subsequently, we gathered protein truncation variants (including splicing variants, frameshift indels, and nonsense variants; PTVs) and variants clinically validated in the ClinVar database [19] (ClinVar pathogenic). We also considered the overall PTVs and ClinVar pathogenic (incorporating either PTVs or ClinVar pathogenic) to estimate the maximum possible contribution of pathogenic variants to cancer (seeMethods; Fig.1b). We focused on three gene sets: (i) previously known cancer predisposition genes (CPGs) based on a recent study [23], (ii) OMIM genes, and (iii) somatic driver genes (SODs) that frequent somatic alterations are associated with an increased cancer risk in multiple databases. Ultimately, we identified 7,012 pathogenic variants (either PTVs or ClinVar pathogenic variants) in 2,534 genes that have at least one defined pathogenic variant. Among these variants, 77.8% (5,452 out of 7,012) were PTV variants, and 32.5% (2,280 out of 7,012) were clinically pathogenic as ClinVar pathogenic, including 10.3% overlap between PTVs and ClinVar pathogenic variants (N=\\u2009720).We gathered single nucleotide variants (SNVs) and indels for each gene. The case\\u2013control analysis was designed to test for an excess of pathogenic variants in cancer patients (cases) compared to controls (non-cancer patients). This analysis utilized a burden test with population adjustment, which accounted for the first four principal component (PC) values (explaining\\u2009>\\u200990% of variances; Additional file2: Fig. S2a-b) in order to control for population structure. This structure was estimated from common variants of case\\u2013control samples (seeMethods; Fig.1b and Additional file2: Fig. S2c-d). We set a threshold to define the genes to be tested based on the total number of carriers with pathogenic variants in both case and control samples. This was done to ensure there was no inflation in the distribution of observedP-values (Additional file2: Fig. S4). Consequently, the genes to be tested were defined as those carrying at least three pathogenic variants (either PTVs or ClinVar pathogenic) in the pan-cancer (all cancer samples together) case\\u2013control analysis.Enrichment of pathogenic OMIM gene variants genes across cancersIn our pan-cancer analysis, we found that cancer predisposition genes (CPGs) showed significant enrichment of PTVs in cases (median Log2 odds ratio (OR) for the enrichment of PTVs in cases\\u2009=\\u20091.25,P=\\u20091.07\\u2009\\u00d7\\u200910\\u22123by Wilcoxon rank sum test, Fig.2a and Additional file1: Table S2). The distribution of ORs (without logarithmic transformation) is also presented in the Additional file2: Fig. S5. Notably, we also observed that OMIM genes exhibited significant enrichment for PTVs in cases compared to controls (median Log2OR\\u2009=\\u20090.65,P=\\u20091.54\\u2009\\u00d7\\u200910\\u22124by Wilcoxon rank sum test, Fig.2a). This enrichment was also consistent or stronger when considering OMIM sub-groups separately, OMIM-Autosomal Recessive genes (OMIM-AR; median Log2OR\\u2009=\\u20090.69,P=\\u20094.08\\u2009\\u00d7\\u200910\\u22125by Wilcoxon rank sum test) and Autosomal Dominant genes (OMIM-AD; median Log2OR\\u2009=\\u20090.63,P=\\u20097.62\\u2009\\u00d7\\u200910\\u22123by Wilcoxon rank sum test). Somatic driver genes (SODs;N=\\u2009116) did not show significant enrichment in cases compared to controls (median Log2OR\\u2009=\\u20090.09), indicating that rare protein truncating variants (PTVs) from OMIM genes could contribute to an increased cancer risk.Considering ClinVar pathogenic variants, we observed marginal enrichment of pathogenic variants in CPGs in the pan-cancer case\\u2013control analysis (N=\\u200925; median Log2 OR\\u2009=\\u20090.99;P=\\u20092.39\\u2009\\u00d7\\u200910\\u22121by Wilcoxon rank sum test, Fig.2a and Additional file1: Table S2). This lower enrichment might be due to the smaller number of ClinVar pathogenic variants (which are clinically validated variants; 5,452 PTVs vs. 2,280 ClinVar pathogenic variants) compared to PTVs. OMIM genes (N=\\u2009456) also showed marginal enrichment (median Log2 OR\\u2009=\\u20090.31;P=\\u20094.52\\u2009\\u00d7\\u200910\\u22121by Wilcoxon rank sum test) and only two ClinVar pathogenic variants were detected in SODs (median Log2 OR\\u2009=\\u2009-0.29), as expected based on their definition (cancer genes discovered by recurrent somatic alterations but not inherited variants).When analyzing both PTVs and ClinVar pathogenic variants together in CPGs (N=\\u200936), we saw robust enrichment in the pan-cancer dataset (median Log2 OR\\u2009=\\u20091.06;P=\\u20094.66\\u2009\\u00d7\\u200910\\u22124by Wilcoxon rank sum test; Fig.2a). We also found that pathogenic variants in OMIM genes consistently show significant enrichment in cancer compared to control (N=\\u20091,072; median Log2 OR\\u2009=\\u20090.51;P=\\u20091.53\\u2009\\u00d7\\u200910\\u22122by Wilcoxon rank sum test). This included OMIM-AR (median Log2 OR\\u2009=\\u20090.46;P=\\u20091.18\\u2009\\u00d7\\u200910\\u22122by Wilcoxon rank sum test) and OMIM-AD (median Log2 OR\\u2009=\\u20090.60;P=\\u20094.67\\u2009\\u00d7\\u200910\\u22122by Wilcoxon rank sum test). In contrast, SODs (N=\\u2009116) did not show any difference in pathogenic variant enrichment between cancer patients and the controls, consistent with the analysis of PTVs and ClinVar pathogenic variants (median Log2 OR\\u2009=\\u20090.09). To ensure the analysis of as many potential pathogenic variants as possible, from this point forward, our definition of rare pathogenic variants will encompass both PTVs and clinically validated variants.We identified 109 genes (6 CPGs, 103 OMIM genes) that individually showed an enrichment for pathogenic variants (PTVs\\u2009+\\u2009ClinVar pathogenic variants) in pan-cancer at a false discovery rate (FDR) of less than 20% (Fig.2b and Additional file1: Table S2). These data included six previously known CPGs:BRCA2(Log2 OR\\u2009=\\u20094.21,FDR=\\u20094.51\\u2009\\u00d7\\u200910\\u22124),BRCA1(Log2 OR\\u2009=\\u20093.01,FDR=\\u20091.07\\u2009\\u00d7\\u200910\\u22122),FAH(Log2 OR\\u2009=\\u20093.33,FDR=\\u20095.33\\u2009\\u00d7\\u200910\\u22122),ATM(Log2 OR\\u2009=\\u20092.32,FDR=\\u20096.60\\u2009\\u00d7\\u200910\\u22122),BUB1B(Log2 OR\\u2009=\\u20093.14,FDR=\\u20098.69\\u2009\\u00d7\\u200910\\u22122), andGJB2(Log2 OR\\u2009=\\u20091.28,FDR=\\u20098.71\\u2009\\u00d7\\u200910\\u22122). Their targeting cancer types were presented in Additional file1: Table S3. Pathogenic variants of 103 OMIM genes (10.2% of tested OMIM genes) were found to be significantly enriched in cancer patients. This includes 63 OMIM-ARs, 24 ADs, 3 ADARs, and 13 Others. For example,PRSS12displayed the highest significant enrichment of pathogenic variants in cancer compared to control (Log2 OR\\u2009=\\u20093.74,FDR=\\u20099.86\\u2009\\u00d7\\u200910\\u22124), next toOTOG(Log2 OR\\u2009=\\u20092.68,FDR=\\u20091.70\\u2009\\u00d7\\u200910\\u22123),ATP7B(Log2 OR\\u2009=\\u20092.61,FDR=\\u20092.00\\u2009\\u00d7\\u200910\\u22123),AAGAB(Log2 OR\\u2009=\\u20093.45,P=\\u20097.12\\u2009\\u00d7\\u200910\\u22123), andPAX4(Log2 OR\\u2009=\\u20091.98,FDR=\\u20099.31\\u2009\\u00d7\\u200910\\u22123). Hereafter, we will refer to these 109 genes as CPG-like OMIM genes (including 6 CPGs). They show significant enrichment of pathogenic variants in cancers, as determined by using PTVs and ClinVar pathogenic variants in the pan-cancer analysis at an FDR of 20% (Fig.2b).Furthermore, we confirmed that pathogenic variants show specific enrichment in CPG-like OMIM genes by examining independent data sets for cancer data and control sets. For the independent cancer set, The Cancer Genome Atlas (TCGA) was used, which covers 10,389 cancer patients across 33 cancer types, providing a comprehensive independent cancer data set [23]. In addition, we used gnomAD [20], currently the largest collection of human sequencing data, as a control data set after removing allele frequencies from cancer samples (i.e., gnomAD-nonCancer). For validation of our case\\u2013control analysis, we focused on a European population to reduce ethnically biased sequence variability in both the case and control datasets. By applying two combinations of case\\u2013control analysis (TCGA vs. 1\\u00a0KG and PCAWG vs. gnomAD), we observed robust enrichment of pathogenic variants in CPG-like OMIM genes in cancer cases compared to controls. We confirmed that 52.7% to 95.7% of CPG-like OMIM genes presented enriched pathogenic variants in cancer compared to control (Additional file2: Fig. S6).Enrichment of rare germline variants of OMIM genes in specific cancersNext, we applied a case\\u2013control analysis to the 9 cancer types with more than 100 samples (comprising 60% of the samples in a pan-cancer; Additional file1: Table S1). To test for enrichment in single-cancer types versus controls, we selected genes based on the total number of pathogenic variant carriers from cases and control. This helped avoid inflation or depletion in each cancer type separately (Additional file2: Fig. S4). In total, 246 genes (15 CPGs and 231 OMIM genes with 394 gene-cancer associations) were significantly associated with at least one individual cancer type. Of these, 41 genes were also significantly enriched in the pan-cancer dataset (Odds ratio\\u2009=\\u20092.67, Fisher\\u2019s exact testP=\\u20098.15\\u2009\\u00d7\\u200910\\u22126atFDR20%; Additional file1: Table S4). Among the CPGs,BRCA1was significantly enriched in ovarian adenocarcinoma (Ovary-AdenoCA; Log2OR\\u2009=\\u20096.91,FDR=\\u20099.63\\u2009\\u00d7\\u200910\\u22126) and breast adenocarcinoma (Breast-AdenoCA; Log2OR\\u2009=\\u20094.81,FDR=\\u20094.76\\u2009\\u00d7\\u200910\\u22123) andBRCA2was enriched in Breast-AdenoCA (Log2OR\\u2009=\\u20094.70,FDR=\\u20096.93\\u2009\\u00d7\\u200910\\u22125), pancreatic adenocarcinoma (Panc-AdenoCA; Log2OR\\u2009=\\u20094.12,FDR=\\u20096.45\\u2009\\u00d7\\u200910\\u22124) and Ovary-AdenoCA (Log2OR\\u2009=\\u20093.76,FDR=\\u20099.03\\u2009\\u00d7\\u200910\\u22123). From OMIM genes, the top-ranked gene-cancer association wasEEF2(Log2OR\\u2009=\\u20095.38,FDR=\\u20091.40\\u2009\\u00d7\\u200910\\u22123) with renal cell carcinoma (Kidney-RCC). Additionally, variants ofPEX1(Log2OR\\u2009=\\u20095.13,FDR=\\u20093.07\\u2009\\u00d7\\u200910\\u22123) were enriched in B-cell non-Hodgkin\\u2019s lymphoma (Lymph-BNHL),FLG2(Log2OR\\u2009=\\u20095.04,FDR=\\u20093.68\\u2009\\u00d7\\u200910\\u22123) in Kidney-RCC,DHCR7(Log2OR\\u2009=\\u20094.81,FDR=\\u20094.76\\u2009\\u00d7\\u200910\\u22123) in Breast-AdenoCA andFAM111A(Log2OR\\u2009=\\u20094.68,FDR=\\u20097.70\\u2009\\u00d7\\u200910\\u22123) in medulloblastoma (CNS-Medullo) also showed significant enrichment.Functional annotation and disease classes represented in CPG-like OMIM genesOMIM genes that cause different genetic diseases can also be grouped into disease classes based on the physiological system affected. To complement single-gene-based case\\u2013control analysis, we measured the enrichment of pathogenic variants in 49 genetic diseases assigned to 10 representative disease classes using a regression model (seeMethods). From 49 genetic diseases with at least one pathogenic variant from CPG-like OMIM genes (N=\\u2009109), 38 diseases showed significant enrichment in at least one cancer type or pan-cancer compared to the control (77.6% atFDR<\\u20095%; Fig.3a). The results remain consistent when we incorporate an additional set of disease genes from the Clinical Genome Resource, which includes 463 newly identified disease-associated genes for 59 genetic diseases (43 out of 59 diseases, 72.9% at FDR\\u2009<\\u20095%; see Additional file2: Fig. S7), including four new disease classes (Aminoacidopathy, RASopathy, Intellectual disability, and Motile ciliopathy). The metabolic disease class had the highest frequency of significantly enriched pathogenic OMIM variants (22.4%; 11 diseases). The neurological disease class had the second-highest frequency (16.3%; 8 diseases), followed by the immune disease class (14.3%; 7 diseases).Open in a separate windowFig. 3Pathogenic variants enriched in (a) diseases and (b) functional clusters across pan-cancer and single-cancer types using the linear regression model. Circle size indicates the excess of pathogenic variants per gene set, and color represents the significance of the enrichment in a specific cancer type compared to the control (1KG). Numbers in parentheses indicate the number of samples in each cancer type (left) and the number of tested genes in each disease (or functional annotation and pathway; bottom). The bar plot on the right shows the number of detected diseases (or functional terms and pathway) in each cancer type at a false discovery rate (FDR) of less than 0.05Examining this in more detail, glycogen storage disease from the metabolic disease class in Lymph-BNHL showed the strongest enrichment (Log2OR\\u2009=\\u20095.13,FDR=\\u20092.61\\u2009\\u00d7\\u200910\\u22123). This is consistent with previous data showing that glycogen metabolism is associated with an increased risk of malignancy and tumorigenesis [47,48]. Moreover, the next most significant enrichments highlighted the following associations: peroxisomal disorder to Lymph-BNHL (Log2OR\\u2009=\\u20095.13,FDR=\\u20092.61\\u2009\\u00d7\\u200910\\u22123) and Ovary-AdenoCA (Log2OR\\u2009=\\u20094.76,FDR=\\u20097.33\\u2009\\u00d7\\u200910\\u22123), bone marrow failure syndrome to Ovary-AdenoCA (Log2OR\\u2009=\\u20094.83,FDR=\\u20096.53\\u2009\\u00d7\\u200910\\u221210), autism to Breast-AdenoCA (Log2OR\\u2009=\\u20094.81,FDR=\\u20093.88\\u2009\\u00d7\\u200910\\u22123), abnormal mineralization to Ovary-AdenoCA (Log2OR\\u2009=\\u20094.76,FDR=\\u20097.33\\u2009\\u00d7\\u200910\\u22123) and Fanconi anemia to Breast-AdenoCA (Log2OR\\u2009=\\u20094.70,FDR=\\u20096.16\\u2009\\u00d7\\u200910\\u22125). Interestingly, the comorbidity relationship between Fanconi anemia and breast cancer-associated pathway has previously been described [49,50].We also performed a deeper enrichment analysis of functional annotation using biological processes in gene ontology (GO) [51] and KEGG signaling pathways [52] (seeMethods; Fig.3b). Pathogenic variants in CPG-like OMIM genes were significantly enriched for 11 GO terms and 3 KEGG signaling pathways in at least one individual cancer type or a pan-cancer dataset compared to the control (FDR\\u2009<\\u20090.05). Overall, peroxisome-related pathways showed the strongest enrichment, including the peroxisome pathway from KEGG in Ovary-AdenoCA (Log2OR\\u2009=\\u20095.53,FDR=\\u20099.42\\u2009\\u00d7\\u200910\\u22124) and Lymph-BNHL (Log2OR\\u2009=\\u20095.13,FDR=\\u20092.89\\u2009\\u00d7\\u200910\\u22123) and protein targeting to peroxisome from GO in Lymph-BNHL (Log2OR\\u2009=\\u20095.13,FDR=\\u20093.03\\u2009\\u00d7\\u200910\\u22123). As expected, the classical cancer predisposition genes-associated pathway, homologous recombination showed significant enrichment in pan-cancer (Log2 OR\\u2009=\\u20094.21,FDR=\\u20094.45\\u2009\\u00d7\\u200910\\u22124), Ovary-AdenoCA (Log2 OR\\u2009=\\u20093.76,FDR=\\u20097.25\\u2009\\u00d7\\u200910\\u22123), Breast-AdenoCA (Log2 OR\\u2009=\\u20094.70,FDR=\\u20096.42\\u2009\\u00d7\\u200910\\u22125), and Panc-AdenoCA (Log2 OR\\u2009=\\u20094.12,FDR=\\u20095.89\\u2009\\u00d7\\u200910\\u22124). Moreover, epidermis morphogenesis in Kidney-RCC (Log2OR\\u2009=\\u20095.53,FDR=\\u20099.42\\u2009\\u00d7\\u200910\\u22124), response to ionizing radiation (Log2OR\\u2009=\\u20094.83,FDR=\\u20096.53\\u2009\\u00d7\\u200910\\u221210), and bone mineralization (Log2OR\\u2009=\\u20094.76,FDR=\\u20097.58\\u2009\\u00d7\\u200910\\u22123) in Ovary-AdenoCA also showed significant enrichment.Four possible classes in OMIM genesMany canonical CPGs are known to be constitutively expressed in tissues connected to their general biological function [4]. Moreover, they often adhere to the \\u2018two-hit\\u2019 hypothesis, where a pathogenic variant occurs in one allele of the gene and a \\u2018second hit\\u2019 occurs somatically in another allele, leading to an increased risk of cancer [4,23]. However, exceptions to this broad tissue expression have been observed in several CPGs with tissue-specific functional roles [4], and diverse mechanisms that extend beyond the canonical CPG model are being proposed [6,53]. Furthermore, it is been suggested that a single copy of an inherited CPG variant could be sufficient to increase cancer risk [6]. We hypothesized that pathogenic variant carriers might have an increased cancer risk through a variety of mechanisms beyond the classical understanding of CPGs. To systematically investigate how CPG-like OMIM genes contribute to tumorigenesis in diverse ways, we considered three defining features of the classes: (i) two-hit preference, (ii) tissue-specificity (TAU score), and (iii) loss-of-function intolerant value (LOEUF; Fig.4a).Open in a separate windowFig. 4Four classes from 109 genes.aSchematic representation of gene classification by integrating three features: (i) two-hit preference: the excess of pathogenic variants in samples with LOH compared to samples without LOH (a more positive value indicates stronger two-hit preferences), (ii) TAU score: indicating the gene expression patterns across multiple tissues (from 0 to 1, close to 1 indicates tissue-specifically expressed gene), (iii) LOEUF: the scale of predicted loss-of-function (pLoF) constraint (a higher value indicates a pLOF tolerant gene). A LEOUF score below 0.35 indicates pLOF intolerance.bPrincipal components analysis with three features for 109 genes. Each color represents different clusters.cTernary plot showing the distribution of the three features across four classes. CPGs are highlighted in each plot. The size presents the excess of pathogenic variants in cases compared to control samples (Odds ratio).dBox plot presenting the distribution of assigned values for each feature across clusters. CPGs are depicted as red dots with the name of each cluster. The black dots indicate outliers, which are values between 1.5 and 3 times the interquartile rangesFirst, we applied Knudson\\u2019s two-hit hypothesis to the genes in the pan-cancer analysis. The two-hit preference was designed to measure the enrichment of pathogenic variants (PTVs and ClinVar pathogenic variants) in samples with loss-of-heterozygosity (LOH) compared to samples without LOH within a gene (seeMethods; Additional file2: Fig. S8). We again used a regression model adding four PCs to control for population structure, estimated from common variants from cancer samples only (Additional file2: Fig. S2e). Next, we considered the gene expression pattern to reflect whether the gene is involved in a fundamental function or tissue-specific function [54]. We applied a TAU score, indicating the gene expression pattern from 27 tissues from GTEx [34], ranging from 0 (broadly expressed) to 1 (tissue specifically expressed). Lastly, LOEUF from gnomAD [20] which describes the tolerance to inactivation for predicted loss-of-function (pLoF) was used (genes below 0.35 LOEUF is considered as intolerant gene) [20].We performed a principal component analysis (PCA) using the three features mentioned above as inputs for the genes. We defined four clusters using an enhancedk-means clustering algorithm (Additional file2: Fig. S9a), suggesting that diverse mechanisms predispose OMIM genes to act like CPGs (Fig.4b-c). The first cluster showed both low two-hit prevalence in a pan-cancer analysis and the lowest TAU score distribution, indicating that these genes are broadly expressed across tissues (Fig.4d and Additional file2: Fig. S9b). We, therefore, named them the \\u2018broadly expressed\\u2019 group. This cluster includes two CPGs (ATMandFAH) and 33 OMIM genes, includingABCA7,CYP1B1,LRRK2,MVK,MMAA, andCOL18A1(Additional file1: Table S7). The second cluster, including one CPG (GJB2), and 21 OMIM genes, is characterized by high TAU score and LOEUF scores but relatively low two-hit preferences in a pan-cancer analysis compared to single cancer types (Fig.4d, Additional file2: Fig. S9b-c). We refer to this cluster as the \\u2018Tissue-specific strong pLOF tolerant\\u2019 group of CPG-like OMIM genes. For example,PRSS12in Panc-AdenoCA,ABCC11in liver hepatocellular carcinoma (Liver-HCC), andNOD2in skin melanomas (Skin-Melanoma) show tissue-specific expression patterns (Additional file1: Table 7 and Additional file2: Fig. S9c). Genes in the third cluster have a high TAU score like the second cluster but have the lowest LOEUF values. This cluster includes one CPG (BUB1B) and 30 OMIM genes and is called the \\u2018Tissue-specific less pLOF tolerant\\u2019 group (Fig.4d). The last cluster presents strong \\u2018two-hit\\u2019 preferences in a pan-cancer analysis as well as in single cancer types (Fig.4d and Additional file2: Fig. S9b-c) and includes two classical CPGs (BRCA1andBRCA2)and 19 OMIM genes. We therefore named this cluster the \\u2018Canonical two-hit\\u2019 group.PAHis a potential cancer-predisposing geneFor a deeper functional analysis of OMIM genes as potential novel CPGs, we focused on phenylalanine hydroxylase (PAH; assigned to Cluster 2, \\u2018Tissue specific strong pLOF tolerant gene\\u2019) as it exhibited the highest frequency of pathogenic variants (PTVs and ClinVar pathogenic variants) among the 109 genes in the pan-cancer analysis (1.82%; 48 carriers out of 2,642 PCAWG cancer patients; Additional file1: Table S2).PAHplays a pivotal metabolic role, converting L-phenylalanine to L-tyrosine. InheritedPAHvariants result in conversion to L-phenylpyruvate (Fig.5a) causing phenylketonuria (PKU), an autosomal recessive genetic disease with symptoms that include intellectual disability, seizures, nausea, and vomiting [55]. To gain a better understand of howPAHvariants contribute to metabolic processes, we performed a metabolic simulation using a patient-specific genome-scale metabolic model (GEM) [39] (Additional file2: Fig. S10a). SincePAHshows liver-specific expression in GTEx, PCAWG Liver-HCC samples (N=\\u2009312; including 3PAHcarriers and 309PAHvariant non-carriers) were applied to the model. The input data, defined as metabolite uptake rates, were obtained from GEM simulations (seeMethods). As expected, the GEMs predicted that L-phenylalanine was more actively secreted (P=\\u20091.42\\u2009\\u00d7\\u200910\\u22123by Student\\u2019s t-test) and actively converted to L-tyrosine in thePAHcarriers than the non-carriers (Additional file2: Fig. S10a).Open in a separate windowFig. 5Contribution ofPAHvariants to the metabolic process.aDiagram of the phenylalanine hydroxylase (PAH) pathway and profiling of 220 metabolites for the 8 individuals heterozygous for pathogenicPKUvariants and 22 healthy controls. In phenylalanine metabolism, L-phenylpyruvate can be generated from L-phenylalanine through two enzymatic reactions. The heatmap presents three representative metabolites (red: higher metabolite peaks inPAHcarriers, blue: lower metabolite peaks inPAHcarriers). PCA using the 220 metabolites clearly distinguished thePAHcarriers from thePAHnon-carriers.bHallmark GSEA signatures from PCAWG lung cancer RNA-sequencing data ranked by normalized enrichment score (NES), representing the enrichment of each cancer hallmark in PAH carriers compared to non-carriers. Hallmark signatures are ranked byP-values (at nominalP<\\u20090.05).cBox plot showing the transcriptional score (TS) using an immune checkpoint-related gene set (N=\\u200938) across three groups:PAHcarrier in Lung-SCC PCAWG,PAHnon-carrier in Lung-SCC PCAWG/TCGA, and normal lung samples from GTEx. A high TS indicates increased immune microenvironment correlation within a groupTo evaluate metabolic changes associated with pathogenicPAHvariants, we conducted a CE-TOFMS-based metabolome analysis on 8 individuals heterozygous for the pathogenicPAHvariants (from one parent of a PKU patient), and 22 healthy samples (i.e.,PAHvariant non-carriers without disease phenotype; seeMethodsand Fig.5a). We annotated 220 peaks in the CE-TOFMS data using metabolite libraries. Intriguingly, PCA analysis using the 220 measured metabolite peaks clearly distinguishedPAHvariant carriers from non-carriers, indicating thatPAHvariants contribute to metabolic changes (Fig.5a). There are 40 metabolites that showed significantly different concentrations inPAHvariant carriers compared to healthy donors, including 8 up-regulated and 32 down-regulated metabolites (P<\\u20090.05 by Welch\\u2019s T-test, Additional file1: Table S9). As expected, we observed increased L-phenylalanine levels (Phe; 1.3-fold,P=\\u20096.39\\u2009\\u00d7\\u200910\\u22124by Welch\\u2019s T-test) in the blood ofPAHcarriers compared to the non-carriers, which is a well-known characteristic ofPAHdeficiency.Next, to understand the detailed tumorigenic mechanism ofPAH, we selected Lung-SCC, which shows the highest frequencies of pathogenic variants in both PCAWG (4.3%) and TCGA (3.5%; Additional file2: Fig. S10b). We also confirmed the prevalence ofPAHpathogenic variants in Lung cancer using independent whole-exome sequencing (WES) data of Korean Lung-SCC samples (2.9%; 7 out of 245, Additional file2: Fig. S10b). To test whether tumorigenesis depends on disease-linkedPAHvariants, we performed gene set enrichment analysis (GSEA) on thePAHcarrier and non-carrier group using transcript expression levels from RNA-seq data of Lung-SCC PCAWG-TCGA (seeMethods). Interestingly, immune-response related cancer hallmarks were strongly enriched inPAHcarriers compared to non-carriers, including interleukin-2 and STAT5 signaling, inflammatory response, coagulation, and TNF-\\u03b1 signaling via NF-k(P=\\u20099.00\\u2009\\u00d7\\u200910\\u22123,P=\\u20093.40\\u2009\\u00d7\\u200910\\u22122,P=\\u20092.10\\u2009\\u00d7\\u200910\\u22122, andP=\\u20094.90\\u2009\\u00d7\\u200910\\u22122, respectively; Fig.5b). Furthermore, we conducted a disease-based enrichment analysis by using 110 pre-defined metabolites (seeMethods). ThePAHdysfunction group shows the most significant association with inflammatory disease and myocardial infarction (P=\\u20091.55\\u2009\\u00d7\\u200910\\u22128, Additional file2: Fig. S10c), followed by a metabolic disorder that affects weakened heart, muscle, and growth delay (3-Methylglutaconic aciduria type 2;P=\\u20095.26\\u2009\\u00d7\\u200910\\u22128). This indicates that metabolic changes linked toPAHpathogenic variants, in addition to causing PKU, could lead to multiple diseases, especially those associated with immune system dysfunction.In addition, we evaluated the transcriptional score (TS), the summation of gene expression correlation coefficients between 38 immune checkpoint modulatory genes, within each group to assess the contribution ofPAHvariants to gene expression in the immune system (seeMethods) [44]. We compared TSs across three groups: (i)PAHcarrier (PTVs or ClinVar pathogenic variants;N=\\u200915), (ii)PAHnon-carrier in PCAWG-TCGA Lung-SCC (N=\\u2009478), and (iii) normal lung samples from GTEx [34] (N=\\u2009578).PAHpathogenic variant carriers (median TS\\u2009=\\u200919.14) demonstrated significantly higher TS compared to non-carrier (median TS\\u2009=\\u200914.69,P=\\u20091.05\\u2009\\u00d7\\u200910\\u22123by Wilcoxon rank sum test) and normal lung samples (median TS\\u2009=\\u20099.25,P=\\u20096.14\\u2009\\u00d7\\u200910\\u22128by Wilcoxon rank sum test, Fig.5c). This indicates that the expression levels of immune check-point modulatory genes inPAHvariant carriers tend to be more positively correlated with each other than non-carriers and normal lung samples, suggesting a potentially higher response rates to immune checkpoint inhibitors. Overall, our data suggest thatPAHcould modulate immune-associated pathways to contribute to cancer and that this gene is a potential new CPG.\",\n          \" 3.1. Clinicopathological characteristics and preoperative ctDNA detectionA total of 100 patients with pathological stage II/III resectable GC were enrolled (Figure1). The patients\\u2019 baseline characteristics are summarized in Supplementary TableS2. Sixty\\u2010three patients were diagnosed with stage III GC, and 90 patients received at least one course of ACT. The median duration of ACT was 6.2 months (range: 0.8\\u201013.1 months). At a median follow\\u2010up of 52.2 (95% CI = 50.5\\u201054.3) months, 33 patients ultimately relapsed, and 27 patients died from any cause.In the primary tumors, tumor protein p53 (TP53), AT\\u2010rich interactive domain\\u2010containing protein 1A (ARID1A), and Cadherin 1 (CDH1) were the three most frequently mutated genes (Figure2). A total of 855 variants were detected, and 83 (9.7%) variants presented in only one tumor. More importantly, only 4 of the 83 variants were detected in 3 patients\\u2019 corresponding preoperative plasma samples.Open in a separate windowFIGURE 2Genomic profiling of the most commonly mutated genes (mutation frequency \\u22655%) presented in the primary gastric tumors, and the corresponding clinicopathological features and biomarker status. Abbreviations: AD, adenocarcinoma; SRCC, signet ring cell carcinoma; ctDNA, circulating tumor DNA; Pre\\u2010op, preoperative; Post\\u2010op, postoperative; ACT, adjuvant chemotherapy; CA, cancer antigen, including CEA, CA72\\u20104, CA199.Preoperatively, 223 variants were detected in 42 of the 100 patients, 28 (12.6%) variants called in plasma were presented in corresponding primary tumors, and only 33 preoperative plasma samples were considered positive for ctDNA. Besides, preoperative ctDNA status of the patients was not associated with baseline clinical features (Supplementary TableS2), and no significant RFS or OS difference was observed between patients with and without detectable preoperative ctDNA (RFS: HR = 1.10, 95% CI = 0.55\\u20132.22,P= \\u20090.785; OS: HR = 1.13, 95% CI = 0.53\\u20132.43,P= \\u20090.754; Supplementary FigureS1).3.2. Postoperative ctDNA detection and its association with recurrence riskAfter surgical resection (median, 4 days; range, 1\\u20107 days), 25.0% (25/100) of patients were positive for ctDNA. Patients with postoperative positive ctDNA had a higher proportion of preoperative elevated CEA compared with those with postoperative negative ctDNA (P= 0.046; Supplementary TableS2). No difference in other clinicopathological characteristics was observed between patients with detectable versus undetectable postoperative ctDNA (Supplementary TableS2). Patients with postoperative residual ctDNA had moderately higher risk of recurrence compared to those with undetectable ctDNA (Figure3A). Patients with positive ctDNA also had inferior OS compared to those without (Figure3B). Multivariable Cox analysis confirmed postoperative ctDNA as an independent factor for RFS (Supplementary TableS3). Higher cutoffs for circulating tumor\\u2010specific mutations per ng of cfDNA and per mL of plasma were not associated with significantly higher risk of relapse (Supplementary FigureS2), suggesting that presence of ctDNA rather than the total load of ctDNA is more related to high recurrence risk.Open in a separate windowFIGURE 3Prognostic value of postoperative and post\\u2010adjuvant chemotherapy ctDNA. (A, B) Kaplan\\u2013Meier estimates of recurrence\\u2010free survival (RFS) and overall survival (OS) among patients evaluated for postoperative ctDNA. (C, D) Kaplan\\u2013Meier estimates of RFS and OS among patients evaluated for post\\u2010ACT ctDNA.Pvalue was calculated using a two\\u2010sided log\\u2010rank test. Abbreviations: ctDNA, circulating tumor DNA; HR, hazard ratio; CI, confidence interval; ACT, adjuvant chemotherapy.3.3. Post\\u2010ACT ctDNA predicted significantly higher recurrence and death risksAmong 41 patients who had plasma samples collected within 3 months after ACT, 10 (24.4%) patients were positive for ctDNA. Patients with preoperative elevated CEA had a higher proportion of post\\u2010ACT ctDNA positivity compared to those with preoperative normal CEA (P= 0.036; Supplementary TableS2). No difference in other clinicopathological characteristics was observed between patients with positive versus negative post\\u2010ACT ctDNA (Supplementary TableS2). After ACT, ctDNA\\u2010positive patients had remarkably higher risk of recurrence than ctDNA\\u2010negative patients (Figure3C). Similarly, ctDNA\\u2010positive patients had significantly reduced OS compared to ctDNA\\u2010negative patients (Figure3D). Multivariate Cox regression analysis showed that post\\u2010ACT ctDNA status was still the strongest prognostic factor associated with recurrence and death risk (Supplementary TableS4).Next, we evaluated the dynamic ctDNA changes from preoperative to postoperative status and from postoperative to post\\u2010ACT. We found that those patients with both preoperative and postoperative ctDNA negative experienced the least disease relapse (Supplementary TableS5). In addition, compared with postoperative ctDNA status, both the 2 patients who experienced ctDNA clearance and 27 (93.1%) of 29 patients who remained ctDNA\\u2010negative were disease\\u2010free at last follow\\u2010up; 3 of 5 patients who were initially ctDNA\\u2010negative but turned positive and 4 of 5 patients who remained ctDNA\\u2010positive ultimately had disease relapse (Supplementary TableS6).3.4. Association of recurrence site and ctDNA statusWe also evaluated the association between recurrence site and ctDNA status. Of note, the peritoneum was the most common recurrence site among patients with negative postoperative or post\\u2010ACT ctDNA (Supplementary TableS7), which indicated that ctDNA detection rate is rather low in patients with peritoneal disease.3.5. Post\\u2010ACT ctDNA outperformed other biomarkers in predicting recurrenceNext, we tested the predictive power of postoperative ctDNA, post\\u2010ACT ctDNA and postoperative serial cancer antigen (Figure4A\\u2010C). We found that post\\u2010ACT ctDNA was the most related biomarker to disease relapse (P< 0.001). Further, post\\u2010ACT ctDNA consistently achieved better predictive performance than the other two biomarkers (Figure4D). Taken together, residual ctDNA after ACT could effectively predict recurrence of disease with higher performance metrics, suggesting its potential significance in prognosis evaluation.Open in a separate windowFIGURE 4Performance of recurrence prediction across different biomarkers. (A\\u2010C) The bar plot depicts the association between different biomarkers (A, postoperative ctDNA; B, post\\u2010ACT ctDNA; C, postoperative serial cancer antigen) and recurrence status.Pvalue was calculated using a two\\u2010sided Fisher's exact test. (D) Performance measurements of postoperative, post\\u2010ACT ctDNA status, and postoperative serial cancer antigen (i.e., CEA, CA199, CA72\\u20104) demonstrated by sensitivity, specificity, accuracy, PPV and NPV. Abbreviations: ctDNA, circulating tumor DNA; pos, positive; neg, negative; ACT, adjuvant chemotherapy; CA, cancer antigen, indicated CEA, CA72\\u20104, CA199; PPV, positive predictive value; NPV, negative predictive value.3.6. Correlations ofERBB4mutational status in primary tumors with RFSAlthough both residual postoperative and post\\u2010ACT ctDNA could screen out patients with high risk of recurrence, some patients (22.2%, 16/72) who were negative for ctDNA after surgery or adjuvant chemotherapy still experienced disease relapse. To investigate the biological explanations of this phenomenon, we performed an exploratory analysis of molecular alterations of primary tumors. For patients with undetectable ctDNA during surveillance, theERBB4mutation rate of primary tumors was significantly higher in patients who relapsed than those who were disease\\u2010free (25.0% vs. 3.6%,P= 0.020). Besides,ERBB4mutational status could also discriminate patients with and without disease recurrence in the whole cohort (P= 0.038; Supplementary FigureS3A). Patients who harboredERBB4mutation in tumor tissues had significantly reduced RFS compared to those who wereERBB4wildtype (Supplementary FigureS3B). The external cohort of STAD from TCGA similarly indicated that GC patients who harboredERBB4mutation had shorter DFS than patients withoutERBB4mutation (Supplementary FigureS3C\\u2010D).3.7. Incorporation of ctDNA after definitive treatment improved model's performance for RFS predictionTo assess the overall probability of recurrence, we combined all the risk factors, including pathological stage, lymphovascular invasion, nerve invasion,ERBB4mutational status presented in primary tumors and ctDNA status (postoperative or post\\u2010ACT if available), and constructed a nomogram for RFS prediction (Figure5A). In particular, the prediction of RFS generated by the model with ctDNA status (C\\u2010index = 0.78, 95% CI = 0.71\\u20130.84) was more accurate than that of the model without ctDNA status (C\\u2010index = 0.71, 95% CI = 0.64\\u20130.79,P= \\u20090.009; Figure5B). Additionally, we separately plotted time\\u2010dependent ROC curves using the recurrence probabilities computed by the predictive model with and without ctDNA status. The AUC of the model with ctDNA status versus the model without were 0.83 versus 0.74 at 2\\u2010year RFS (P= 0.006), and 0.84 versus 0.75 at 3\\u2010year RFS (P= 0.002), respectively (Figure5C). Altogether, these findings indicated that the intergration of tissue\\u2010based and circulating tumor features could achieve better risk prediction.Open in a separate windowFIGURE 5ctDNA after definitive treatment improved model's performance for RFS prediction. (A) The nomogram was constructed by selected clinicopathological factors, ERBB4 mutational status, and ctDNA status (post\\u2010operation or post\\u2010ACT if available) for predicting 2\\u2010year and 3\\u2010year RFS. mut, mutant; wt, wild\\u2010type; pos, positive; neg, negative. (B) Comparison of C\\u2010index and 95% CI for predicting RFS between the model with and without ctDNA status. (C) Time\\u2010dependent ROC curves and corresponding AUC values of 2\\u2010year (left) and 3\\u2010year (right) RFS prediction among the model with and without ctDNA status. Abbreviations: RFS, recurrence\\u2010free survival; ctDNA, circulating tumor DNA; ROC, receiver operating characteristic; AUC, area under the curve; AJCC, American Joint Committee on Cancer.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"discussion\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \" We noted an enrichment of rare germline variants in autosomal recessive OMIM genes through our case\\u2013control analysis. Genetically, this might be associated with the potential for diverse clinical phenotypes in heterozygous carriers of rare germline variants. The prevailing assumption for heterozygotes with pathogenic variants is that carriers generally experience a normal lifespan, given that the wild-type allele maintains gene function [56]. However, with extended life expectancy, unexpected phenotypes attributable to rare germline variants may become evident in heterozygotes. For instance, heterozygous carriers might exhibit an altered metabolite profile or a lower penetrance that does not provoke classical disease phenotypes, including cancers [57].Nonetheless, long-term exposure to this altered metabolism may adversely affect cellular fitness and could potentially trigger cancer.In the four decades since Alfred Knudson proposed the \\u2018two-hit\\u2019 hypothesis, several classical cancer predisposition genes (CPGs) conforming to this model have been identified. These CPGs have been validated in specific tissues or in pan-cancer analyses [4,58,59]. Our study not only reaffirms the utility of the \\u2018two-hit\\u2019 hypothesis in classifying CPGs but also identifies exceptions where pathogenic variants are not linked to additional somatic alterations, even in known CPGs. Notably, genes in Cluster 2 exhibited low bi-allelic inactivation in the pan-cancer dataset along with significant tissue-specific gene expression levels, when compared to other clusters (mean tissue-specificity value in Cluster 2\\u2009=\\u20090.87, other clusters\\u2009=\\u20090.58,P<\\u20094.59\\u2009\\u00d7\\u200910\\u22129by Student\\u2019s t-test; Fig.4d). This result might suggest the existence of a non-classical mechanism wherein bi-allelic inactivation is not commonly observed in pan-cancer analyses.In our study, we highlightPAHas a novel candidate CPG that frequently harbors pathogenic variants in the pan-cancer dataset, and specifically in single-cancer types such as Lung-SCC. Our transcriptomic analysis reveals that Lung-SCC patients with pathogenic PAH variants demonstrate associations with immune-related pathways. This could be influenced by prolonged dysregulated metabolic stress [60,61]. Indeed, patient-specific genome-scale metabolic models (GEMs) derived from PAH carriers reveal metabolic profiles that markedly deviate from those of the normal population (PAH non-carriers). These findings provide insights into a possible mechanism of cancer progression tied to altered metabolic profiles in individuals with pathogenic PAH variants. Importantly, comorbidity between phenylketonuria and various cancer types has been previously observed [62,63], which further supports our findings.In our study, we anticipate that sequencing an even larger collection of tumors will further refine these identified clusters and enhance our understanding of the diverse mechanisms contributing to cancer. Despite utilizing a well-balanced, large-scale cancer and control dataset, and controlling for population differences, our study still faces several limitations. First, although we employed a uniform pipeline for variant annotations in both case and control sets and observed no clear distinctions between sequencing centers in our PCA (Additional file2: Fig. S2c-e), other potential technical factors remain unaddressed. These include sequencing kits, and age of samples. Such factors could influence the number of detected PTVs due to technical or biological discrepancies between the case and control sets. We observed nominal differences in the number of PTVs for genes that are neither CPGs nor OMIM genes (two-sample Kolmogorov\\u2013Smirnov test,P=\\u20090.02). However, no such differences were discerned in CPGs (two-sample Kolmogorov\\u2013Smirnov test,P=\\u20091; Additional file2: Fig. S12). Given the limitations in the sample annotations for both the case and control datasets, a thorough confirmation remains technically challenging. Richer sample annotations in the future could bolster the regression model and more accurately quantify differences between case and control groups. The second potential limitation of this study arises from the unbalanced population composition, given that the PCAWG dataset is primarily derived from European/American patients (72.1%). This means that the full spectrum of pathogenic variants in non-European cancer patients may not have been entirely explored due to the limited sample size. For instance, we observed differences in the location of pathogenic variants ofPAHbetween PCAWG-TCGA (mainly European) and Korean Lung-SCC patients, indicating potential ethnic-dependent variations in tumorigenesis mechanisms (Additional file2: Fig. S10d). The sequencing of a larger cohort of both tumor samples and healthy control individuals will allow for further refinement and validation of PAH variant enrichment in Korean lung cancer data. Moreover, additional data will help elucidate the diverse roles of OMIM genes that contribute to cancer via different mechanisms.\",\n          \" This prospective cohort study evaluated the power of ctDNA detection to predict risk of recurrence in patients with stage II/III GC. The results underlined that residual ctDNA after definitive treatment could identify patients with higher likelihood of relapse. Compared to recent studies addressing the same issues within relatively small cohorts [18,19,20], we confirmed that both postoperative and post\\u2010ACT ctDNA positivity were still independent predictors of RFS after adjusting for all risk factors in multivariate analyses.In the present study, we used a 425\\u2010gene panel\\u2010based NGS approach to capture tumor\\u2010specific somatic mutations in primary tumors and applied them to detect ctDNA and we regarded a variant detected in plasma as cancer\\u2010derived only when it also presents in at least 2 primary gastric tumors. The population\\u2010based rather than highly personalized tumor\\u2010informed strategy was applied due to strong genomic heterogeneity within GCs, the mutation feature of a single tumor tissue cannot represent the whole mutational spectrum [36,37]. Besides, the European Society for Medical Oncology (ESMO) Precision Medicine Working Group recently recommends to sample at least 1 week after surgery to analyze ctDNA MRD [38]. Here, liquid biopsy was performed within 1 week after surgery considering the short half\\u2010life (\\u223c2 h) of ctDNA [39]. Undoubtedly, the population\\u2010based strategy for defining ctDNA positivity and sampling within 1 week after surgery might lead to decrease in specificity and produce false\\u2010positives, in addition to the lack of an independent validation cohort, which became the limitation of the present study. However, ctDNA evaluation performed as early as 1\\u20107 days postoperatively in this study still showed comparable performance of sensitivity with many reported tumor\\u2010informed ctDNA analyses between 4 and 10 weeks postoperatively [13,18,19,40]. In clinical practice, high sensitivity of MRD detection will be critical if MRD assays are applied for guiding a patient with ctDNA MRD to receive adjuvant therapy.Preoperatively, ctDNA positive rate was 38.0% in our cohort, which was in the low end of preoperative detection rates ranging from 37% to 57% in localized gastroesophageal cancer [18,19]. Besides, preoperative ctDNA did not demonstrate a strong correlation with baseline clinicopathological features and showed limited value for predicting recurrence. These results showed discrepancies with reports [19,41], which indicated that ctDNA shedding rates were affected by factors such as T stage and tumor volume. Further, similar to our findings, other studies also illustrated consistency in the clinical features between preoperative positive and negative ctDNA subgroups [14,42]. These data portray that our understanding of the mechanisms of ctDNA shedding in GC actually remains unclear [43].Through serial ctDNA monitoring, we highlighted the importance of post\\u2010ACT ctDNA as a prognostic biomarker for it reflected much higher risk of relapse than postoperative ctDNA (HR = 14.99 vs. 2.74), which was not emphasized in previous researches assessing the association between ctDNA detection and recurrence risk in resectable GC. Yang et\\u00a0al. [19] reported similar results with quite a small sample size (n= 23), and the other two studies lacked evaluation of residual ctDNA after ACT for predicting recurrence [18,20]. Actually, postoperative ctDNA was not invariable during the whole clinical course as we showed. After accomplishing ACT with curative intent, dynamic shifting of ctDNA status was observed and in good concordance with treatment response. More importantly, post\\u2010ACT ctDNA demonstrated a predictive advantage over postoperative ctDNA (sensitivity = 77.8% vs. 42.4%; specificity = 90.6% vs. 83.6%). Therefore, ctDNA detection after adjuvant therapy deserves more attention, as their existence indicated an extremely high risk of recurrence and the reminder of resistance to adjuvant therapy.Of note, a subset of patients who were ctDNA\\u2010negative after definitive treatment still experienced relapse (22.2% in the present study and 32.0% in a study by Yang et\\u00a0al. [19]). Exploratory analysis of primary tumors revealed thatERBB4\\u2010mutant GC was probably a more malignant subtype with poor prognosis in both our and external cohorts, as evidenced by the finding thatERBB4mutation played an oncogenic role in GC from our group [44]. Besides, a comprehensive model established with all risk factors including ctDNA status showed a high accuracy in recurrence risk prediction. Consequently, the utility of ctDNA detection for predicting recurrence should be combined with the clinicopathological and biological features (e.g., molecular alterations, pathological stage) presented in tumors to achieve better predictive accuracy.For a long time, it is debatable whether we should perform ctDNA testing for postoperative patients. In 2022, the DYNAMIC study (ACTRN12615000381583) provided prospective evidence that ctDNA\\u2010based adjuvant therapies for colon cancer reduced the use of adjuvant chemotherapy without comprising RFS [45]. Other ctDNA MRD\\u2010based adjuvant therapies were explored in the IMvigor011 study ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT04660344\\\",\\\"term_id\\\":\\\"NCT04660344\\\"}}NCT04660344) [46] for muscle\\u2010invasive urothelial cancer, the PEGASUS trial ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT04259944\\\",\\\"term_id\\\":\\\"NCT04259944\\\"}}NCT04259944) [47] for stage II\\u2010III colon cancer and the MERMAID\\u20101 study ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT04385368\\\",\\\"term_id\\\":\\\"NCT04385368\\\"}}NCT04385368) [48] for stage II\\u2010III non\\u2010small cell lung cancer. Moreover, for patients with residual ctDNA after completion of adjuvant therapies, whether they should extend treatment period or receive other treatment regimens are still worthy of further exploration. Reassuringly, we find that several prospective studies are exploring \\u201cmolecular metastatic\\u201d treatment after completing standard adjuvant therapy with molecular relapse monitoring in breast cancer [TRAK\\u2010ER ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT04985266\\\",\\\"term_id\\\":\\\"NCT04985266\\\"}}NCT04985266), DARE ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT04567420\\\",\\\"term_id\\\":\\\"NCT04567420\\\"}}NCT04567420) and LEADER ({\\\"type\\\":\\\"clinical-trial\\\",\\\"attrs\\\":{\\\"text\\\":\\\"NCT03285412\\\",\\\"term_id\\\":\\\"NCT03285412\\\"}}NCT03285412) trials].\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"methods\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \" In order to ensure reproducibility the full, commented code used for data wrangling, analysis, and visualization have been uploaded to GitHub (see Code Availability in main text). The primary code was divided under headers and named chunks, which are referenced when relevant below.Empirical data on the variation of the human pangenomeSingle nucleotide variant metadata from 141,456 adult humans published by Karczewski et al. in 202036, including their calculations of mutational constraint, was incorporated in this study by downloading their Supplementary Dataset11(publicly available as open-access). Only canonical gene transcripts were used. In this data, specific metrics on both empirically observed variants and theoretically expected variants are available. Predicted loss-of-function (pLoF) variants observed vs. expected (LOE) scores refer to the specific ratio of observed vs. expected number of pLoF variants. pLoF observed vs. expected upper/lower bound fraction (LOEUF/LOELF) scores refer to the lower and upper bounds of the 90% confidence interval of the LOE score, respectively. In keeping with the literature36,94, a gene was designated asconstrainedwhen it showed a LOEUF score lower than 0.35 based on random distribution across the genome. For the purposes of our study, genes with pLoF observed vs. expected (LOE) ratio under 0.35 were designatedlikely constrained; genes with pLoF observed vs. expected (LOE) ratio over 0.35 were designatedlikely not constrained; and genes with an observed vs. expected lower bound fraction (LOELF) score higher than 0.35 were designatednot constrained(Fig.1B). In essence,not constrainedrefers to genes where the current data does not support that the gene is, or is likely to be,constrainedusing the 35% LOEUF cut-off. Throughout, higher constraint refers to lower LOEUF score.GnomAD v2.1 data was also implemented in analyses using publicly available variant-resolution data from the 125,748 exomes (\\u201cAll chromosomes sites VCF\\u201d,https://storage.googleapis.com/gcp-public-data--gnomad/release/2.1.1/vcf/exomes/gnomad.exomes.r2.1.1.sites.vcf.bgz, see primary code, header: Fig.4(incl. gnomad data import), chunk: gnomad_data_load_and_coverage_filter). This data was subsetted using filter (pLoF == \\u201cHC\\u201d) on a separate high-performance compute cluster, extracting only variants only considered high confidence by LOFTEE (https://github.com/konradjk/loftee).Size-matched gene setIn order to account for gene size (included as a variable, coding sequence (CDS) lengths, in Supplementary Dataset11by Karczewski et al.36) we arranged pCPS genes according to gene size and grouped genes into nine subsets of equal size. Using the size ranges from these nine groups of pCPS genes, we randomly sampled all genes (not excluding pCPS genes), to create the maximum size-matched subset, satisfying the criteria that (1) genes appeared only once and (2) each group contained an equal number of genes (see primary code, header: Data cleaning, annotation and preparation, chunk: sizematching). The resultant group of size-matched genes was compared to the genes of interest to ensure that the two groups were no longer significantly different.Selecting cancer predisposition syndrome genes with pediatric-onsetThe paperSelection criteria for assembling a pediatric cancer predisposition syndrome gene panel(2021) by Byrjalsen et al.41lists a total of 85 genes deemedCategory 1for childhood cancer risk (Supplementary Data3); i.e. a pathogenic variant (i) in a minimum of five children (aged 0\\u201318 at diagnosis) with (ii) a cancer diagnosis in at least two independent families. These minimum requirements are independent of the prevalence in adult-onset cancer patients41.In our work, these 85 genes are referred to as the pediatric-onset cancer predisposition syndrome (pCPS) genes. Throughout, we distinguish between biallelic phenotypes, caused by two pathogenic variants sitting in trans in a gene (i.e. autosomal recessive), and monoallelic phenotypes, caused by one single variant in a gene (i.e. autosomal dominant and X-linked recessive), as well as whether the associated pediatric phenotype was driven by loss-of-function or gain-of-function genotypes. (Supplementary Data4).A separate set of cancer predisposition syndrome (CPS) genes were designated as adult-onset cancer predisposition syndrome (aCPS) genes using the following methodology, also employed in our previous study56. Using a list of all CPS genes published by Rahman (2014)95, we filtered out all gene/mode-of-inheritance pairs that were designated as associated with pCPS (see above). The remaining 62 genes/mode-of-inheritance combinations were designated as aCPS genes (Supplementary Data4, see primary code, header: Data cleaning, annotation and preparation, chunk: annotating_constraint).Combined germline mutational spectrum in patients with childhood cancerAll pediatric pan-cancer studies, published in June 2022 or earlier and including germline sequencing of 100 or more participants were included irrespective of the sequencing strategy (targeted, WES, or WGS). First, the data were loaded into R in their original published format; the \\u2018rbind()\\u2019 function was used to combine data from the following studies: Zhang et al. (2015)33, Parsons et al. (2016)46, Mody et al. (2016)47, Oberg et al. (2016)48, Gr\\u00f6bner et al. (2018)10, Wong et al. (2020)49, Byrjalsen et al. (2020)34, Fiala et al. (2021)50, Newman et al. (2021)51, Stedingk et al. (2021)52, and Wagener et al. (2021)53(Supplementary Data5\\u201315). Similarly, gene panels from each study were combined using \\u2018rbind()\\u2019 (Supplementary Data16\\u201326). A patient and data overlap between Zhang et al. (2015)33and Gr\\u00f6bner et al. (2018)10was noted by Gr\\u00f6bner et al. (2018);10\\u201c[out of]n=\\u2009914 individual patients, about 25% of samples overlapping with the previous study\\u201d. While not stated explicitly in the publication, it was assumed that this refers to the 259 samples [28.4%, 259/914] that shared a St. Jude ID (format; SJxxxnnn). Hence, this overlap was accounted for by filtering out patient IDs from Zhang et al. (2015)33that also appeared in Gr\\u00f6bner et al. (2018)10(see primary code, header: Data cleaning, annotation and preparation, chunk: metadata_homogen).Next, nomenclature cleaning was performed on the mutational ontology, pathogenicity classification, and diagnosis type categories using the \\u2018mutate()\\u2019 function from the \\u2018dplyr\\u2019 package in R. As an example, likely pathogenic variants, reported as \\u2018PP\\u2019, \\u2018likely pathogenic\\u2019, \\u2018C4: Likely pathogenic\\u2019, \\u2018Likely Pathogenic\\u2019, \\u2018Likely pathogenic\\u2019, and \\u2018LP\\u2019 were homogenized to simply \\u2018LP\\u2019 (full details for all 300+ renamings are in primary code, header: Data cleaning, annotation and preparation, chunk: metadata_homogen). All reported variants were retained and no variants were filtered out, with the exception of the c.3920\\u2009T\\u2009>\\u2009A (p.Ile1307Lys) allele in theAPCgene, that while reported as pathogenic in one50of the 11 studies, is considered a risk factor for cancer rather than a precipitator of the CPS associated with the gene. The c.3920\\u2009T\\u2009>\\u2009A (p.Ile1307Lys) allele has not been shown to cause childhood cancer (see primary code, header: Fig.5, chunk: pancan_plots).Any germline variant reported as either likely pathogenic or pathogenic may be referred to simply as pathogenic in the manuscript text. Variants were considered pLoF if the variant ontology was frameshift, nonsense or splice donor/acceptor regardless of pathogenicity. To make variation observed in children with cancer comparable to those seen in gnomAD, we used pre-computed downsamples of the gnomAD data (see under header \\u2018Data Availability\\u2019 in the main manuscript text; see primary code, header: Fig.4(incl. gnomad data import), chunk: downsampling_import) filtered to variation affecting canonical transcripts. Importing this data (https://storage.googleapis.com/gcp-publicdata--gnomad/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.downsamplings.txt.bgz), we generated plots comparing the observed number of LoF variation in the two groups by downsampling the combined germline data reports from pediatric pancancer studies (downsampling was done, as allowed by sample size, to the same levels as those precomputed in the gnomAD data). A total of 38 possible downsampling steps were used [10, 20, 50, 100, 200, 500, 1000, 2000, 3070, 5000, 5040, 8128, 9197, 10000, 10824, 15000, 15308, 17296, 20000, 25000, 30000, 35000, 40000, 45000, 50000, 55000, 56885, 60000, 65000, 70000, 75000, 80000, 85000, 90000, 95000, 100000, 110000, 120000]. In the gnomAD dataset, at each of these downsampling steps the observed number of distinct LoF variants were pre-computed, as were the expected number of distinct LoF variants, calculated using the same methods employed for the gnomAD-wide caluculations, as detailed by Karczewski et al.36. In our downsampling of the pediatric pancancer cohort data, we employed an iterative for-loop (see primary code, header: Fig.4(incl. gnomad data import), chunk: downsampling_for_pancan). This meant that the for-loop began by counting all distinct LoF variants (as defined below) across all available studies (variable by number of studies with the given gene included on panel (see Supplementary Data16\\u201326)). For genes included in all 11 pediatric pancancer studies, the total number of childhood cancer cases with data was 4,574 (9,148 alleles). Each unique patient was labeled either with their specific variant or as having no variant detected. From this set, using the sample_n function in the R package dplyr, a random sample, corresponding to the nearest lower downsample step (n=\\u20093,070 in the example), was extracted. LoF variants in the extracted random sample were counted, and then, a random sample corresponding to the next downsample step (n=\\u20092,000 in the example) was taken from the previous step (n=\\u20093,070 in the example). This process ran iteratively through all downsampling steps (9 steps [10, 20, 50, 100, 200, 500, 1000, 2000, 3070] in the example).Statistical testingWe used R version 4.0.2 (2020-06-22) for all statistical tests (see primary code, header: All statistical tests, chunk: all_stat_tests). All reported P values are two-sided. No data samples were analyzed more than once. Confidence parameters are reported in-text for each test.Inclusion and ethicsThis study did not include original data and, as such, neither qualified for nor required ethical oversight.Software and toolsExternal data was downloaded in.xlsx or.csv format and compiled in sheets, without edits, in a Microsoft Excel workbook. All subsequent data wrangling, analysis, calculations and some visualizations were done using R (version 4.2.2) through RStudio (version 2022.07.1). Non-computational illustrations were made using the browser-based graphical tool BioRender. Illustrations of exon/protein and variants were made using the browser-based graphical tool ProteinPaint96on the PeCan platform. Scripts were written in Rmarkdown and are available on github (see Code Availability in main text). The scripts are preambled by the required packages, which, byname(version number), are; readxl(1.4.1), backports(1.4.1), systemfonts(1.0.4), plyr(1.8.8), lazyeval(0.2.2), splines(4.2.2), TH.data(1.1-1), digest(0.6.31), htmltools(0.5.4), fansi(1.0.3), memoise(2.0.1), googlesheets4(1.0.1), tzdb(0.3.0), penxlsx(4.2.5.1), remotes(2.4.2), modelr(0.1.10), matrixStats(0.63.0), sandwich(3.0-2), prettyunits(1.1.1), colorspace(2.0-3), vest(1.0.3), textshaping(0.3.6), haven(2.5.1), xfun(0.36), callr(3.7.3), crayon(1.5.2), jsonlite(1.8.4), libcoin(1.0-9), Exact(3.2,) survival(3.4-0), glue(1.6.2), polyclip(1.10-4), gtable(0.3.1), gargle(1.2.1), pkgbuild(1.4.0), clipr(0.8.0), Quandl(2.11.0), mvtnorm(1.1-3), DBI(1.1.3), miniUI(0.1.1).1, Rcpp(1.0.9), xtable(1.8-4), gridtext(0.1.5), foreign(0.8-83), bit(4.0.5), proxy(0.4-27), stats4(4.2.2), profvis(0.3.7), htmlwidgets(1.6.0), httr(1.4.4), modeltools(0.2-23), ellipsis(0.3.2), farver(2.1.1), urlchecker(1.0.1), pkgconfig(2.0.3), ultcompView(0.1-8), dbplyr(2.2.1), utf8(1.2.2), labeling(0.4.2), tidyselect(1.2.0), rlang(1.0.6), later(1.3.0), munsell(0.5.0), cellranger(1.1.0), tools(4.2.2), cachem(1.0.6), generics(0.1.3), broom(1.0.2), evaluate(0.19,) fastmap(1.1.0), ragg(1.2.4), yaml(2.3.6), processx(3.8.0), bit64(4.0.5), fs(1.5.2), zip(2.2.2), coin(1.4-2), rootSolve(1.8.2).3, mime(0.12,) xml2(1.3.3), compiler(4.2.2), rstudioapi(0.14,) curl(4.3.3), e1071(1.7-12), gt(0.8.0), reprex(2.0.2), tweenr(2.0.2), broom.helpers(1.11.0), DescTools(0.99.47), stringi(1.7.8), ps(1.7.2), lattice(0.20-45), Matrix(1.5-1), vctrs(0.5.1), pillar(1.8.1), lifecycle(1.0.3), lmtest(0.9-40), ata.table(1.14.6), lmom(2.9,) httpuv(1.6.7), R6(2.5.1), promises(1.2.0).1, gld(2.6.6), sessioninfo(1.2.2), codetools(0.2-18), boot(1.3-28), MASS(7.3-58.1), assertthat(0.2.1), pkgload(1.3.2), withr(2.5.0), nortest(1.0-4), multcomp(1.4-20), expm(0.999-6), parallel(4.2.2), hms(1.1.2), quadprog(1.5-8), class(7.3-20), rmarkdown(2.19,) googledrive(2.0.0), shiny(1.7.4).Reporting summaryFurther information on research design is available in theNature Portfolio Reporting Summarylinked to this article.\",\n          \" Tumor sequencing data from PCAWGThe final consensus set of merged germline mutation calls in variant call format (VCF) generated by the Pan-cancer Analysis of Whole Genomes (PCAWG) [12] consortium was obtained from the International Cancer Genome Consortium Portal (https://dcc.icgc.org/releases/PCAWG) with authorization. The final dataset contains 2,642 high-quality samples after excluding 192 samples due to potential technical issues such as contamination or low-quality from the original pool 2,834 donors [12]. The germline variant calling file was divided into two germline VCF files: one containing 1,823 donors from the International Cancer Genome Consortium (ICGC), excluding US donors, and the other containing 819 donors from The Cancer Genome Atlas (TCGA). These separate VCF files were then integrated using bcftools-1.9, as recommended by the ICGC data coordination center (DCC) [13]. Germline variant calling was originally performed by the ICGC working group [12]. This group used six different variant callers: GATK HaplotypeCaller, FreeBayes, Real Time Genomics (RTG), Delly, TraFiC mobile element insertion caller (https://gitlab.com/mobilegenomes/TraFiC), and Eagle2 to assemble the SNVs, indels, and structural variations (SV). The final set of PCAWG germline calls comprises 2,642 donors who are composed of four subpopulations as described in Additional file1: Table S1 European/American (1,904 donors; 72.1%), East Asian (396 donors; 15.0%), African (125 donors; 4.7%), and South Asian (36 donors; 1.4%). This also contains 181 donors (6.9%) whose ethnic information was not available.Control exomes from the 1000 Genomes ProjectExome sequences of healthy controls were collected from the 1000 Genomes Project (1\\u00a0KG; Phase III high-coverage whole-exome sequences) [14] VCF from the FTP server (ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/) under the following authorization (dbGaP phs000710.v1.p1). The data covers 2,504 individuals from four subpopulations (European,N=\\u2009503; American,N=\\u2009347; African,N=\\u2009661; East Asian,N=\\u2009504; South Asian,N=\\u2009489, Additional file2: Fig. S2a). Data generation and variant calling procedures are detailed in the 1\\u00a0KG flagship paper [14].Clinical informationClinical and histological annotation for PCAWG donors were obtained from the ICGC Portal [12] (https://dcc.icgc.org/releases/PCAWG/clinical_and_histology). PCAWG data included 1,462 males (55.3%) and 1,180 females (44.7%). The Pan-cancer cohort comprises 37 distinct tumor types (Additional file1: Table S1). Cancer patients\\u2019 normal samples were mainly derived from blood and non-tumor tissues adjacent to the primary stie or other sites such as bone marrow and lymph node. The ethnicity and gender annotation for 1\\u00a0KG samples is available from the 1000 Genome portal (https://www.internationalgenome.org/data). No additional phenotype information was collected since 1\\u00a0KG is a freely available genomic data set.Variant annotation and filtering stepTo generate a single-standard functional annotation pipeline, we compiled each VCF file from PCAWG (N=\\u20092,642) and 1\\u00a0KG (N=\\u20092,504). First, we discarded variants in ENCODE/DUKE [15] and DAC blacklist [16] regions to filter out variants in low mappability regions and we selected variants only in the ENCODE/CRG GEM mappability region (75mers) [17]. Next, we annotated the collected variants in the VCF file using Annovar (version 2014 Apr 14) [18]. Of the data Annovar reports, we used (i) the clinical and phenotypic effect of variants via ClinVar [19] (accessed on 18 June 2019), (ii) minor allele frequencies (MAFs) of variants across eight subpopulations: African/African American (AFR), South Asian (SAS), East Asian (EAS), Latino/Admixed American (AMR), non-Finnish European (NFE), Finnish European (FIN), Ashkenazi Jewish (ASJ) and Other (OTH) and also globally in The Genome Aggregation Database (gnomAD) exome (v2.1.1) [20]. Simultaneously, we also used Variant Effect Predictor (VEP)-release-96 [21] to provide the gene-based information of canonical transcripts. Of the data VEP reports, we used the consequences of the protein-coding variants: synonymous, missense, stop gain and loss, splice site, frameshift insertion/deletion (indel), and in-frame indel. Next, we discarded all variants marked as potential technical artifacts in the gnomAD exome including excess heterozygosity at a variant site (InbreedingCoeff), allele count zero variants after filtering out low-confidence genotypes (AC0) and failed random forest filtering threshold (RF). The differences in the number of detected variants between the two groups were not observed across three types of variants (median number of variants per sample in PCAWG had 17,320 of total variant, 7,788 missense, and 364 PTVs, while 17,464 total, 7,843 missense and 369 PTVs in 1\\u00a0KG; Additional file2: Fig. S1; Pearson correlation coefficient between PCAWG and control for three types of variants\\u2009=\\u20090.96, respectively).Principal component analysis using common variantsWe expect many germline variants from whole-exome sequencing data or genome-wide association studies to vary according to the different ethnicities within the cohort. To account for the potential confounding effects from population stratifications, we conducted a principal component analysis (PCA) using only the common germline variants (non-synonymous) with\\u2009\\u2265\\u20095% MAF, both globally and within each ethnic group in the gnomAD exome and 90% genotyping rate using PLINK [22] version 2.0. We performed two separate PCAs: (i) a case\\u2013control analysis using PCAWG and 1\\u00a0KG data, and (ii) a two-hit preference analysis using only PCAWG data (Additional file2: Fig. S2).Rare pathogenic variantsWe defined rare variants as those with a frequency\\u2009<\\u20090.5% globally and within each of the eight subpopulations in the gnomAD exome. We confirmed the distribution of MAFs of clinically validated pathogenic variants. The excess of pathogenic variants in cancer was robust when applying a 0.1% MAF globally and within each of the eight subpopulations in the gnomAD exome (Additional file2: Fig. S11). For variants not present in the gnomAD exome, we checked the detection frequency of variants in our cohorts (PCAWG, 1\\u00a0KG separately) to avoid including TCGA or 1\\u00a0KG biased variants, and discarded variants detected in more than 1% of samples in either the PCAWG or 1\\u00a0KG cohort.From the collected rare variants, we defined pathogenic variants using two conditions: (1) Protein truncation variants (PTVs; as potentially deleterious variants) were included splice site, frameshift indel, and stop gain/lost variants as annotated by VEP (release-96) [21]. Additionally, we included predicted splicing donor or acceptor loss variants with a score greater than 0.8 by SpliceAI (https://github.com/Ensembl/VEP_plugins/blob/release/109/SpliceAI.pm). To collect only highly confident PTVs, we applied several filtering steps: (i) removing PTVs annotated as \\u201cBenign\\u201d or \\u201cLikely benign\\u201d in ClinVar [19], and (ii) excluding PTVs located in either terminal exon or without functional domains (https://www.ebi.ac.uk/interpro/). (2) ClinVar pathogenic variants were defined with \\u2018Pathogenic\\u2019, \\u2018Likely pathogenic\\u2019, \\u2018association\\u2019, and \\u2018risk factors\\u2019 with related clinical evidence developed by the American College of Medical Genetics and Genomics (ACMG) in ClinVar [19].Gene set classificationWe retrieved the genes with known to cause in clinical diseases and phenotypes from the Online Mendelian Inheritance in Man (OMIM) databases [8] (https://www.omim.org/downloads; Jan 21, 2020) via the Gene Map (cytogenetic locations of genes;genemap2.txt) and Morbid Map (cytogenetic locations of disorders;mim2gene.txt). The OMIM database provides 17,076 genes linked to 5,392 diseases represented in the Gene Map and Morbid Map. We selected 5,460 disease-associated genes, each associated with at least one phenotype (genetic disease), fromgenemap2.txtfor further analysis. Using the R packageannotables(https://github.com/stephenturner/annotables), we annotated these genes to their Ensemble gene names, resulting in 4,095 unique genes. Furthermore, we collected a total of 152 known CPGs from recent studies [23] including 114 genes from a recently published review paper [4], 11 genes from Cancer Gene Census-Germline (http://cancer.sanger.ac.uk/census/), 12 genes from references search (details in from Huang et al. [23]), and 15 genes from the St. Jude PCGP germline study [24].To compile a list of highly confident somatic driver genes, we first collected 1,196 genes across four data sources: IntOGen [25], MutSig [26], MutPan [27], and Cancer Gene Census-Somatic [28]. To avoid overlapping gene sets, we first defined the CPGs (N=\\u2009152), then identified 3,952 OMIM genes, excluding 143 that overlapped with CPGs. Next, we defined 697 SODs after removing 499 genes that overlapped with either CPGs or OMIM genes. Subsequently, we excluded 53 CPGs from the final CPG set, which had no clinically proven pathogenic variants from ClinVar found in cancer consortium data sets (either TCGA or PCAWG). We considered only those genes with at least three pathogenic variant carriers in a pan-cancer study to increase statistical power. Ultimately, we defined 1,224 genes as input genes, including 36 CPGs, 116 SODs, and 1,072 OMIM genes, which comprised 658 autosomal-recessive [AR] genes, 229 autosomal-dominant [AD] genes, 85 AD-AR genes (i.e., variants from the same gene having dual effects, either dominant or recessive), 5 X-linked genes, 7 somatic, 3 digenic, 2 isolated, 3 multifactorial, and 80 genes with unknown inheritable phenotypes, based on distinctive inheritance annotation (Additional file1: Table S2).Statistical analysis of pathogenic variant enrichmentWe performed a burden test to compare the enrichment of pathogenic variants within genes between cancer cases (PCAWG) and controls (1\\u00a0KG). This method aims to measure gene-level cancer risk enrichment by consolidating potentially rare pathogenic variants using a generalized linear regression model (GLM) via a stats package in R. We implemented this model on each gene in a pan-cancer study as follows:glm(N\\u223cGermlineVariants+PC1+PC2+PC3+PC4,family=``binomial\\\")Here,N=\\u2009case (1) or control (0), Germline Variants\\u2009=\\u2009the number of samples carrying rare pathogenic variants for each gene-tissue pair, and PC1-PC4 are principal component values derived from PCA analysis of PCAWG and 1\\u00a0KG data, utilized to control population structures. We applied PTVs or ClinVar pathogenic variants separately and together (Fig.2a). We computed the regression coefficient andP-value for individual gene-tissue pairs using the summary function in R. More positive coefficient values signify stronger enrichment of pathogenic variants in the case compared to the control. Additionally, we examined if including gender as a factor would influence the regression results. We confirmed that the model's output remains consistent, regardless of the presence or absence of the gender effect (see Additional file2: Fig. S3). For single-cancer type analysis, we only considered the single, most prominent population due to the statistical issue of vastly biased population discrepancies with a small sample size compared to pan-cancer.Open in a separate windowFig. 2Enrichment of pathogenic variants in 2,642 cases compared to 2,504 control samples.aEnrichment of pathogenic variants in cases compared to control samples for three gene sets and four OMIM subgroups (*P<\\u20090.05, **P<\\u20090.01, ***P<\\u20090.001). The median value of each gene set is displayed as a circle. The length of each whisker represents 1.5 times the interquartile range (shown as the height of each box).bExcess of pathogenic variants (PTVs and ClinVar pathogenic variants) in case samples compared to control samples for 1,265 individual genes. The dashed horizontal line indicates a Log2(odds ratio) equal to zero, and the dashed vertical line represents the statistical significance threshold (FDR=\\u200920%). Color indicates significance, shape represents the type of gene, and size presents the frequency of pathogenic variants. Significantly enriched CPGs are highlighted with a yellow boundaryPathogenic variant enrichment in disease class and pathwaysTo systematically estimate the enrichment of pathogenic variants in disease classes compared to controls, we compiled 10 diseases classes (Cardiovascular, Endocrine, Epilepsy, Hematology, Immune, Liver, Lung, Metabolism, Muscular/Skeletal, and Neurological) encompassing 96 hereditary diseases based on clinically relevant disease diagnosis from Centogene (https://www.centogene.com/diagnostics/ngspanels.html) and Blueprint Genetics (https://blueprintgenetics.com/tests/panels/).To investigate the biological knowledge associated with the 109 genes and their pathogenic variants, we first conducted a web-based functional annotation analysis using the Database for Annotation, Visualization and Integrated Discovery (DAVID;https://david.ncifcrf.gov/) [29]. Using DAVID, we mapped gene-level functional annotations (GO) and signaling pathway information (KEGG) to individual genes (see Additional file1: Table S5). Next, we performed an enrichment analysis for each disease or gene ontology/pathway in a pan-cancer study as follows:glm(N\\u223cPathwayorDiseaseclasses+PC1+PC2+PC3+PC4,family=``binomial\\\")where:N=\\u2009case (1) or control (0),Pathway or Disease=\\u2009the number of samples carrying a rare pathogenic variant for each disease or pathway-related gene. PC values from PCA analysis of PCAWG and 1\\u00a0KG served as input for the regression model.Loss-of-heterozygosity event from PCAWGWe retrieved genomic copy-number alteration (CNA) data, which includes loss-of-heterozygosity (LOH) information, from the International Cancer Genome Consortium (ICGC) data portal [30,31]. This dataset provides gene-level calls across 2,642 samples and can be accessed at ICGC (https://dcc.icgc.org/releases/PCAWG/consensus_cnv/gene_level_calls). It encompasses summarized intersections of genomic region calls from six CNA callers: ABSOLUTE, ACEseq, Battenberg, cloneHD, JaBbA, and Sclust, as detailed in Dentro et al. [32]. LOH was determined from the CNA data (all_samples.consensus_CN.minor_allele.by_gene.170313.txt; Additional file1: Table S6) using the profile of the minor copy number, defined as the least amplified allele, following the methodology described in the PCAWG paper [31]. Additionally, we compared our LOH classifications (either LOH or no-LOH) against observed copy number changes (either loss or gain) from the dataset (all_samples.consensus_level_calls.by_gene.170214.txt) for all examined genes. Our detected LOH events were categorized as losses (77.2%), wild-type (18.9%), or gains (3.4%). In comparison, the proportions for no-LOH events were 5.2%, 83.1%, and 11.8%, respectively (Additional file2: Fig. S8).Statistical analysis of two-hit preferenceTo test the two-hit hypothesis, we performed an analysis to check for an excess of pathogenic variants within a gene in samples with LOH events compared to samples without such events across cancer types. We applied the model to each gene-tissue pair as follows:glm(N\\u223cGermlineVariants+PC1+PC2+PC3+PC4,family=``binomial\\\")In this model,Nrepresents samples with LOH event (1) or samples without LOH event (0),Germline Variantsindicates the number of samples carrying rare pathogenic variants (PTVs\\u2009+\\u2009ClinVar pathogenic variants) for each gene-tissue pair. The principal component values derived from PCA analysis, performed solely on cancer samples to control for population structures (Additional file2: Fig. S2e), were used as an input for the regression. We computed the regression coefficient andP-value for each gene-tissue pairs using thesummaryfunction in R. More positive coefficient values indicate a stronger enrichment of pathogenic variants in samples with LOH compared to samples without LOH.Case\\u2013control analysis in a single cancerFor the analysis of individual cancer types, we focused only on the nine cancer types with a sample size greater than 100 to ensure robust analysis. These include Breast-AdenoCA (N=\\u2009194), CNS-Medullo (N=\\u2009145), Kidney-RCC (N=\\u2009143), Liver-HCC (N=\\u2009317), Lymph-BNHL (N=\\u2009107), Ovary-AdenoCA (N=\\u2009113), Panc-AdenoCA (N=\\u2009238), Prost-AdenoCA (N=\\u2009210), and Skin-Melanoma (N=\\u2009107). We then applied the regression model for pathogenic variant enrichment, disease/pathway analysis, and the two-hit preference analysis to the major ethnic group for each cancer type (East Asian for Liver-HCC and European for the remaining eight cancer types, as detailed in Additional file1: Table S1). The threshold for the number of samples carrying a pathogenic variant in each cancer type was carefully adjusted to avoid inflation or depletion dependent on the cancer type. We included genes with more than two carriers either in cases (samples with LOH) or controls (samples without LOH) for CNS-Medullo and Liver-HCC, more than three for Kidney-RCC, Lymph-BNHL, Ovary-AdenoCA, Panc-AdenoCA, Prost-AdenoCA, and Skin-Melanoma, and more than four for Breast-AdenoCA (Additional file2: Fig. S4).PCA for the gene classificationTo classify significantly enriched genes (N=\\u2009109), we integrated three features: (i) Two-hit preferences in a pan-cancer (excess of pathogenic variants in samples with LOH event compared to samples without LOH event, as determined by the regression model; measured as a Log2 odds ratio), (ii) TAU score [33], which indicates gene expression patterns across various tissues from GTEx data [34] (from 0 to 1: close to 1 indicates a tissue-specifically expressed gene), (iii) The loss-of-function tolerance value, the loss of function observed/expected upper bound fraction (LOEUF). This scale measures genes predicted for loss-of-function (pLoF) constraint, as determined by gnomAD [20]. We used three values as input for PCA using theprcompfunction in the R package stats and defined four clusters using the enhancedK-means clustering method in the R package factoextra package (https://cran.r-project.org/web/packages/factoextra/index.html) [35].Independent cancer cohort from TCGA and gnomADFor independent validation, we applied two different databases: TCGA (as an independent case set; for TCGA vs. 1\\u00a0KG) and gnomAD (as an independent control set; for PCAWG vs. gnomAD). Germline variants for 10,389 TCGA samples [23] were obtained from the National Cancer Institute (NCI) Genomic Data Commons (GDC) [36] legacy archive (dbGaP phs000178; PCA.r1.TCGAbarcode.merge.tnSwapCorrected.10389.vcf). After excluding 782 samples overlapping with the PCAWG study, we collected a final sample of 9,607 TCGA cases. This sample included four populations: European (N=\\u20097,072; 73.6%), Asian (N=\\u2009603; 6.3%), African (N=\\u2009793; 8.3%), and other or unknown ethnicity (N=\\u20091,139; 11.9%). We also used gnomAD v2 exome [20] as an independent control set, performing additional variant annotation steps with both Annovar and VEP to align with the annotation used in the PCAWG data. We extracted pathogenic variants in gnomAD using the same variant collection steps as previously described. To circumvent issues with population differences, our independent validation focused on the European population. Among the 109 CPG-like OMIM genes, pathogenic variants were also found in 93 genes within the European population. We calculated pathogenic variant frequencies across genes by summing all non-cancer European allele frequencies, since individual variant information is not available in gnomAD. Finally, we measured the differences between PCAWG_European and non-cancer European gnomAD data.Metabolome analysis ofPAHcarriersTo investigate metabolite changes due to PAH alteration, we collected blood samples from 8 PAH heterozygote carriers (parents of Phenylketonuria patients) and 22 healthy controls (without the phenylketonuria phenotype) at the Seoul National University Hospital, South Korea approved by the Seoul National University Hospital Review Boards (IRB No. H-2101\\u2013196-1193). Phenylketonuria (PKU) is caused by genetic variants in both alleles of the PAH gene (i.e., homozygotes). Therefore, we assumed that the parents of PKU patients would be heterozygote carriers. All carriers and controls were of Korean ethnicity. The metabolic changes associated with PAH heterozygote carriers were assessed at Human Metabolome Technologies, Inc. (HMT), with the samples being processed according to the HMT protocol (Document ID: BLB.1.0.0). Metabolic changes were evaluated using the Cation and Anion modes of CE-TOFMS (Capillary Electrophoresis Time-of-Flight Mass Spectrometry) for metabolome profiling. We extracted 220 annotated peaks from the CE-TOFMS analysis using automatic integration software (MasterHands ver.2 [37]), which included the mass-to-charge ratio (m/z), migration time (MT), and peak area. We used the following equation to convert the peak area to the relative peak area.RelativePeakArea=MetabolitePeakAreaInternalStandardPeakArea\\u00d7SampleAmountWe then assigned putative metabolites from HMT's standard library and Known-Unknown peak library using the m/z and MT values. All metabolite concentrations were measured using standard curves derived from single-point (100\\u03bcM) calibrations, and the peak area of each metabolite was normalized to the area of the internal standard. We selected 110 metabolites with quantified concentrations for disease-related enrichment analysis (Additional file2: Fig. S10c) by MetaboAnalyst 5.0 [38] (https://www.metaboanalyst.ca/), a web-based metabolome data analysis tool.Metabolic simulations using genome-scale metabolic modelsWe used a genome-scale metabolic model (GEM) [39] for metabolic simulations in PAH carriers and non-carriers within PCAWG. GEM is a computational model that incorporates information on the biochemical reactions of all metabolic genes in a cell, and it can be simulated using optimization techniques for various metabolic studies [40]. In this study, a previously developed generic human GEM Recon 2M.2 [39] was transformed into a context-specific (or patient-specific) GEM by integration it with RNA-Sequencing data from PCAWG-TCGA Liver-HCC and Lung-SCC samples. The Task-driven Integrative Network Inference for Tissues (tINIT) method, complemented by a rank-based weight function, was employed to generate patient-specific GEMs. We first predicted metabolites secreted from the PAH carriers (N=\\u20093) and the non-carriers (N=\\u2009309) using the Liver-HCC GEMs. The Liver-HCC GEMs predicted PAH carriers and non-carriers to secrete 16 and 40 metabolites, respectively. The predicted metabolite secretion rates from the PAH carriers with Liver-HCC were then used as input for the GEMs representing the PAH carriers with Lung-SCC (N=\\u200915). Similarly, the predicted metabolite secretion rates from the PAH non-carriers with Liver-HCC were used as input for the GEMs representing the PAH non-carriers with Lung-SCC (N=\\u2009478). To simulate the Lung-SCC GEMs, we arbitrarily limited the maximum uptake rate for the predicted metabolites from the Liver-HCC GEMs to 1.1 times the average metabolite secretion rates. Exchange reactions for inorganic nutrients were set to have the maximum secretion and uptake rates of 1,000 and -1,000 mmol/gDCW/h, respectively. We did not allow uptake of other nutrients when simulating the Lung-SCC GEMs. We predicted metabolic fluxes of the patient-specific GEMs using the least absolute deviation (LAD) method, which aims to minimize the distance between transcript expression levels from RNA-seq data and target reaction fluxes to be calculated. Both the tINIT with the rank-based weight function and the LAD method were implemented using in-house Python scripts [39].Functional analysis withPAHWe used 493 RNA sequencing (RNA-seq) data from the PCAWG-TCGA Lung-SCC via the GDC Data portal (https://portal.gdc.cancer.gov), for the PAH analysis. Gene expression values were calculated post-transcriptomic read alignment to the human reference genome (hg19) using spliced transcripts alignment to a reference (STAR) version 2.5.3a [41] and RSEM v1.3.0 (RNA-seq by Expectation Maximization) [42]. Gene set enrichment analysis (GSEA) was performed using the Java application version 4.1.0 [43] to evaluate the functional enrichment in Cancer Hallmark with 1,000 permutations. In addition, we calculated the transcriptional score (TS) [44], defined as an absolute sum of gene expression correlation coefficient values using 38 immune checkpoint modulatory genes [45]. We compared TS values among three groups of samples (1)PAHcarrier in Lung-SCC, (2)PAHnon-carrier in Lung-SCC, and (3) normal lung samples in GTEx [34]. Given the significantly smaller number of PAH carriers in Lung-SCC compared to either PAH non-carriers in Lung-SCC or GTEx normal lung samples, we randomly selected an equal number of samples 1,000 times (N=\\u200945 samples of PAH non-carriers and GTEx normal lung group, separately). The mean TS values calculated from random sampling with non-PAH and GTEx samples were then compared with the observed TS value of PAH carriers.Korean Lung-SCC germline variant callingFor the independent validation ofPAHpathogenic variant frequency in lung cancer, we collected 245 peripheral blood mononuclear cell (PBMC) specimens from Korean patients with lung squamous cell carcinoma. This was carried out with informed consent from Samsung Medical Center (SMC) under the IRB No. 2013\\u201310-112 and 2008\\u201306-033. The samples were processed using the Swift 2S Turbo DNA library kit, and whole exome sequencing (WES) data was generated with the IDT xGen Exome Research Panel v1.0 kit on the Illumina NovaSeq 6000 platform. The variant calling step was conducted using the Genome Analysis Toolkit (GATK) Best Practices Workflow pipeline, such as the HaplotypeCaller [46]. We followed the same variant annotation process as described above.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cluster_id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 0,\n        \"max\": 0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"COSMIC:\\u00a0a curated database of somatic variants and clinical data for cancer\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_cluster.shape"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EbvBfmMR_ma5",
        "outputId": "61738d24-7979-41a1-e242-1e09f310bb04"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(7, 7)"
            ]
          },
          "metadata": {},
          "execution_count": 41
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Wrapping inference as a class"
      ],
      "metadata": {
        "id": "Suk2VvlfUNAL"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GnB7ap31cbGH"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "57e2435e-b941-4908-b65b-ad08d7d101cd",
        "id": "LPscTwrhcbGH"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n"
          ]
        }
      ],
      "source": [
        "drive.mount('/content/gdrive')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "package_directory = r'/content/gdrive/MyDrive/CS_546_Project_Dependencies/lib/python3.10/site-packages'"
      ],
      "metadata": {
        "id": "5JzqQYCCcG14"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import sys\n",
        "sys.path.append(package_directory)"
      ],
      "metadata": {
        "id": "3brPEjHGalEt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import torch\n",
        "import pandas as pd\n",
        "import spacy\n",
        "import torch\n",
        "from fastcoref import FCoref\n",
        "from transformers import AutoTokenizer, AutoModelForSeq2SeqLM"
      ],
      "metadata": {
        "id": "ITUVBbXccVME"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from spacy.tokens import Doc, Span"
      ],
      "metadata": {
        "id": "2Jnqig8sloAc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class RelationExtractor:\n",
        "    def __init__(self, nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, data=None):\n",
        "      \"\"\"\n",
        "      Initialize the RelationExtractor class with models for coreference, triplet extraction, and data handling.\n",
        "      \"\"\"\n",
        "      self.nlp_scispacy = nlp_scispacy\n",
        "      self.fcoref_model = fcoref_model\n",
        "      self.rebel_model = rebel_model\n",
        "      self.rebel_tokenizer = rebel_tokenizer\n",
        "      self.device = device\n",
        "      self.data = data\n",
        "\n",
        "    def filter_data(self):\n",
        "      \"\"\"\n",
        "      Filter and clean the abstract text.\n",
        "      \"\"\"\n",
        "      if self.data is None:\n",
        "          raise ValueError(\"No data loaded. Please load data using first.\")\n",
        "\n",
        "      self.data['title'] = self.data['title'].fillna('')\n",
        "      df_filtered = self.data[['title', 'abstract']]\n",
        "      df_filtered['abstract_updated'] = df_filtered['abstract'].apply(self.filter_text)\n",
        "      df_filtered['input'] = ( df_filtered['title'] + \" \" + df_filtered['abstract_updated'])\n",
        "      self.data = df_filtered\n",
        "\n",
        "    ### Utility Methods ###\n",
        "    def get_fast_cluster_spans(self, doc, clusters):\n",
        "      \"\"\"\n",
        "      Convert fast coreference model output clusters into spans.\n",
        "      \"\"\"\n",
        "      fast_clusters = []\n",
        "      for cluster in clusters:\n",
        "          new_group = []\n",
        "          for span_tuple in cluster:\n",
        "              (start, end) = span_tuple\n",
        "              span = doc.char_span(start, end)\n",
        "              if span:\n",
        "                  new_group.append([span.start, span.end - 1])\n",
        "          fast_clusters.append(new_group)\n",
        "      return fast_clusters\n",
        "\n",
        "    def get_fastcoref_clusters(self, doc, text):\n",
        "      \"\"\"\n",
        "      Extract coreference clusters using the FastCoref model.\n",
        "      \"\"\"\n",
        "      try:\n",
        "          preds = self.fcoref_model.predict(texts=[text])\n",
        "          fast_clusters = preds[0].get_clusters(as_strings=False)\n",
        "          fast_cluster_spans = self.get_fast_cluster_spans(doc, fast_clusters)\n",
        "      except Exception as e:\n",
        "          print(f'Error while getting coreference clusters: {e}')\n",
        "          return []\n",
        "      return fast_cluster_spans\n",
        "\n",
        "    def core_logic_part(self, document: Doc, coref: list[int], resolved: list[str], mention_span: Span):\n",
        "      \"\"\"\n",
        "      Replace a coreference mention with the cluster's head mention.\n",
        "      \"\"\"\n",
        "      final_token = document[coref[1]]\n",
        "      if final_token.tag_ in [\"PRP$\", \"POS\"]:\n",
        "          resolved[coref[0]] = mention_span.text + \"'s\" + final_token.whitespace_\n",
        "      else:\n",
        "          resolved[coref[0]] = mention_span.text + final_token.whitespace_\n",
        "      for i in range(coref[0] + 1, coref[1] + 1):\n",
        "          resolved[i] = \"\"\n",
        "      return resolved\n",
        "\n",
        "    def get_span_noun_indices(self, doc: Doc, cluster: list[list[int]]) -> list[int]:\n",
        "      \"\"\"\n",
        "      Get indices of noun spans within a coreference cluster.\n",
        "      \"\"\"\n",
        "      try:\n",
        "          spans = [doc[span[0]:span[1] + 1] for span in cluster]\n",
        "          spans_pos = [[token.pos_ for token in span] for span in spans]\n",
        "          span_noun_indices = [\n",
        "              i for i, span_pos in enumerate(spans_pos)\n",
        "              if any(pos in span_pos for pos in ['NOUN', 'PROPN'])\n",
        "          ]\n",
        "          return span_noun_indices\n",
        "      except Exception as e:\n",
        "          print(f'Error while getting span noun indices: {e}')\n",
        "          return []\n",
        "\n",
        "    def get_cluster_head(self, doc: Doc, cluster: list[list[int]], noun_indices: list[int]):\n",
        "      \"\"\"\n",
        "      Determine the head mention for a coreference cluster.\n",
        "      \"\"\"\n",
        "      head_idx = noun_indices[0]\n",
        "      head_start, head_end = cluster[head_idx]\n",
        "      head_span = doc[head_start:head_end + 1]\n",
        "      return head_span, [head_start, head_end]\n",
        "\n",
        "    def is_containing_other_spans(self, span: list[int], all_spans: list[list[int]]):\n",
        "      \"\"\"\n",
        "      Check if a span contains other spans within a cluster.\n",
        "      \"\"\"\n",
        "      return any([s[0] >= span[0] and s[1] <= span[1] and s != span for s in all_spans])\n",
        "\n",
        "    def filter_text(self, text):\n",
        "      \"\"\"\n",
        "      Filter and clean the abstract text.\n",
        "      \"\"\"\n",
        "      try:\n",
        "          if type(text) != str:\n",
        "              return ''\n",
        "          if 'abstract' in text:\n",
        "              return text.replace('abstract ', '').strip()\n",
        "          return text.strip()\n",
        "      except Exception as e:\n",
        "          print(f'Error filtering text: {e}')\n",
        "          return ''\n",
        "\n",
        "    ### Coreference Resolution Methods ###\n",
        "    def improved_replace_corefs(self, document, clusters):\n",
        "      \"\"\"\n",
        "      Replace coreferences with their head mentions in the document.\n",
        "      \"\"\"\n",
        "      resolved = list(tok.text_with_ws for tok in document)\n",
        "      all_spans = [span for cluster in clusters for span in cluster]\n",
        "\n",
        "      for cluster in clusters:\n",
        "          noun_indices = self.get_span_noun_indices(document, cluster)\n",
        "\n",
        "          if noun_indices:\n",
        "              mention_span, mention = self.get_cluster_head(document, cluster, noun_indices)\n",
        "\n",
        "              for coref in cluster:\n",
        "                  if coref != mention and not self.is_containing_other_spans(coref, all_spans):\n",
        "                      self.core_logic_part(document, coref, resolved, mention_span)\n",
        "\n",
        "      return \"\".join(resolved)\n",
        "\n",
        "    def perform_coreference_resolution(self, text):\n",
        "      \"\"\"\n",
        "      Perform coreference resolution on the input text.\n",
        "      \"\"\"\n",
        "      try:\n",
        "          doc = self.nlp_scispacy(text)\n",
        "          clusters = self.get_fastcoref_clusters(doc, text)\n",
        "          coref_text = self.improved_replace_corefs(doc, clusters)\n",
        "          return coref_text\n",
        "      except Exception as e:\n",
        "          print(f\"Error in coreference resolution: {e}\")\n",
        "          return text\n",
        "\n",
        "    ### Triplet Extraction Methods ###\n",
        "    def extract_triplets(self, decoded_string):\n",
        "      \"\"\"\n",
        "      Extract triplets from the decoded string.\n",
        "      \"\"\"\n",
        "      decoded_str_copy = decoded_string\n",
        "      if '<s>' in decoded_string:\n",
        "          decoded_str_copy = ''.join(decoded_string.split('<s>')[1:])\n",
        "          decoded_str_copy = decoded_str_copy.split('</s>')[0]\n",
        "\n",
        "      triplets = []\n",
        "      parts = decoded_str_copy.split(\"<triplet>\")\n",
        "\n",
        "      for part in parts[1:]:\n",
        "          head, *relations = part.split(\"<subj>\")\n",
        "          head = head.strip()\n",
        "\n",
        "          for relation in relations:\n",
        "              if \"<obj>\" in relation:\n",
        "                  subj, obj = relation.split(\"<obj>\", 1)\n",
        "                  triplets.append((head, subj.strip(), obj.strip()))\n",
        "\n",
        "      return triplets\n",
        "\n",
        "    def perform_inference(self, text):\n",
        "      \"\"\"\n",
        "      Perform inference to extract triplets from the input text.\n",
        "      \"\"\"\n",
        "      try:\n",
        "          extracted_triplets = []\n",
        "          for sentence in text.split('. '):\n",
        "              inputs = self.rebel_tokenizer(\n",
        "                  sentence, return_tensors=\"pt\", truncation=True, max_length=512\n",
        "              ).to(self.device)\n",
        "              outputs = self.rebel_model.generate(**inputs)\n",
        "              decoded_output = self.rebel_tokenizer.decode(outputs[0], skip_special_tokens=False)\n",
        "              triplets = list(set(self.extract_triplets(decoded_output)))\n",
        "              if triplets:\n",
        "                  extracted_triplets.extend(triplets)\n",
        "          return extracted_triplets\n",
        "      except Exception as e:\n",
        "          print(f\"Error during inference: {e}\")\n",
        "          return []\n",
        "\n",
        "    ### Data Handling and Aggregation Methods ###\n",
        "    def load_data(self, file_path):\n",
        "      \"\"\"\n",
        "      Load and preprocess data from a JSON file.\n",
        "      \"\"\"\n",
        "\n",
        "      if self.data is None:\n",
        "        df = pd.read_json(file_path)\n",
        "        df['title'] = df['title'].fillna('')\n",
        "        df_filtered = df[['title', 'abstract']]\n",
        "        df_filtered['abstract_updated'] = df_filtered['abstract'].apply(self.filter_text)\n",
        "        df_filtered['input'] = ( df_filtered['title'] + \" \" + df_filtered['abstract_updated'])\n",
        "\n",
        "        self.data = df_filtered\n",
        "\n",
        "        '''\n",
        "        else:\n",
        "          print('Data already loaded')\n",
        "          self.data['title'] = self.data['title'].fillna('')\n",
        "          df_filtered = self.data[['title', 'abstract']]\n",
        "          df_filtered['abstract_updated'] = df_filtered['abstract'].apply(self.filter_text)\n",
        "          df_filtered['input'] = ( df_filtered['title'] + \" \" + df_filtered['abstract_updated'])\n",
        "          self.data = df_filtered\n",
        "\n",
        "        return self.data\n",
        "        '''\n",
        "      #return pubmed_inference_filtered\n",
        "\n",
        "    def process_coreference_resolution(self):\n",
        "      \"\"\"\n",
        "      Process coreference resolution for the loaded data.\n",
        "      \"\"\"\n",
        "      if self.data is None:\n",
        "          raise ValueError(\"No data loaded. Please load data using `load_data` method.\")\n",
        "      self.data['input_coref'] = self.data['input'].apply(self.perform_coreference_resolution)\n",
        "\n",
        "    def process_triplet_extraction(self):\n",
        "      \"\"\"\n",
        "      Process triplet extraction for the loaded data.\n",
        "      \"\"\"\n",
        "      if 'input_coref' not in self.data.columns:\n",
        "          raise ValueError(\"Coreference resolution not performed. Please run `process_coreference_resolution` first.\")\n",
        "      self.data['triplets'] = self.data['input_coref'].apply(self.perform_inference)\n",
        "\n",
        "    def aggregate_triplets(self):\n",
        "      \"\"\"\n",
        "      Aggregate all extracted triplets into a single set.\n",
        "      \"\"\"\n",
        "      if 'triplets' not in self.data.columns:\n",
        "          raise ValueError(\"Triplet extraction not performed. Please run `process_triplet_extraction` first.\")\n",
        "      aggregated_triplets = list(set().union(*self.data['triplets']))\n",
        "      return aggregated_triplets\n",
        "\n",
        "    def write_triplets_to_file(self, aggregated_triplets, file_path):\n",
        "      \"\"\"\n",
        "      Write aggregated triplets to a text file.\n",
        "      \"\"\"\n",
        "      with open(file_path, \"w\") as f:\n",
        "          for triplet in aggregated_triplets:\n",
        "              f.write(f\"{triplet}\\n\")\n"
      ],
      "metadata": {
        "id": "LuhOqhAv1hOT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "model_path = \"/content/gdrive/MyDrive/CS_546_Project_Dependencies/BioRED_Results/checkpoint-118\""
      ],
      "metadata": {
        "id": "HtSX1auntmNQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "file_path = '/content/research_text_pubmed_new.json'"
      ],
      "metadata": {
        "id": "1aWlEHolu0ca"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "device = \"cuda:0\" if torch.cuda.is_available() else \"cpu\""
      ],
      "metadata": {
        "id": "SBw4cQu7tw49"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_model = AutoModelForSeq2SeqLM.from_pretrained(model_path)"
      ],
      "metadata": {
        "id": "klLwgi6ntmNS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_model.to(device)"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7074d6eb-327e-4b66-9e08-e3c171f13320",
        "id": "8FM-Wdc9tw49"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "BartForConditionalGeneration(\n",
              "  (model): BartModel(\n",
              "    (shared): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "    (encoder): BartEncoder(\n",
              "      (embed_tokens): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024)\n",
              "      (layers): ModuleList(\n",
              "        (0-11): 12 x BartEncoderLayer(\n",
              "          (self_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (activation_fn): GELUActivation()\n",
              "          (fc1): Linear(in_features=1024, out_features=4096, bias=True)\n",
              "          (fc2): Linear(in_features=4096, out_features=1024, bias=True)\n",
              "          (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "        )\n",
              "      )\n",
              "      (layernorm_embedding): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "    )\n",
              "    (decoder): BartDecoder(\n",
              "      (embed_tokens): BartScaledWordEmbedding(50272, 1024, padding_idx=1)\n",
              "      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024)\n",
              "      (layers): ModuleList(\n",
              "        (0-11): 12 x BartDecoderLayer(\n",
              "          (self_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (activation_fn): GELUActivation()\n",
              "          (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (encoder_attn): BartSdpaAttention(\n",
              "            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
              "          )\n",
              "          (encoder_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "          (fc1): Linear(in_features=1024, out_features=4096, bias=True)\n",
              "          (fc2): Linear(in_features=4096, out_features=1024, bias=True)\n",
              "          (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "        )\n",
              "      )\n",
              "      (layernorm_embedding): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
              "    )\n",
              "  )\n",
              "  (lm_head): Linear(in_features=1024, out_features=50272, bias=False)\n",
              ")"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "rebel_tokenizer = AutoTokenizer.from_pretrained(model_path)"
      ],
      "metadata": {
        "id": "3x_M_joOtw4-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "nlp_scispacy = spacy.load('en_core_sci_md')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qXwMwPBBuJB9",
        "outputId": "7ac2716d-4684-419d-9884-6e05ce2b405d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/spacy/language.py:2195: FutureWarning: Possible set union at position 6328\n",
            "  deserializers[\"tokenizer\"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "fcoref_model = FCoref(device=device)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 400,
          "referenced_widgets": [
            "78f7e9faeb354d41b5097c1e906d8926",
            "1cc3431de0da4a17b6150b434d898c9e",
            "b770309788914ef0bcdb6a58c99618c1",
            "fa45d2ca6653413a81eeae3693700f52",
            "e49623ff75bf497ca4985618e1262196",
            "1798e88119d74ad6bf071fa849858c14",
            "fe4e6ab40eb54099bc062472fd98ef22",
            "8fa9181848f740c3b2c49b3320846937",
            "818f7c9f135e43d884635af9eeec31a8",
            "0f58231c0a9c4ab0b0613da8d8fc760e",
            "a03d4879da2d4c5fae644e7ebb9f248b",
            "b59f144ff36c43e9a738900e9173f08b",
            "dfb71ef971ef4ae3af14321be9283b56",
            "e09ab65f039d4d7a91575c409021fc61",
            "0257cd89fa3f45129c9c7bfa184f2049",
            "2d419d6719844396bcac4000f4dcb138",
            "7124f8c491f24363ae5fdd3364dc8a07",
            "7cb630959f714cd088e202c7ed901e55",
            "3818cd316f5040c9b8d522aba4e50a89",
            "3703030778574de288ae5073475281cf",
            "42a944d62949408c943ddb0e234ab6a2",
            "7145b20514bd40cdbbbb82109c67b1bf",
            "e8e48ed233964d758c79f2241f2af3f4",
            "f9fe372bcad345d19620c2a3862b589a",
            "dcbfda2f8d204427a2ff0ee756d9c702",
            "59600d0699bd43678f98a5274e0cb62c",
            "501fefbf21ac4d44bcf7e17674a1c566",
            "e47be1b5b9ba4b9e915f77945c051732",
            "468bb1c0ecbd49bab8de977cd6c7f597",
            "cc69579b708049d38a1f4ea317ae7983",
            "68f8924445224b9ebe0986cee669839d",
            "4372c458d49047ae96859be793763031",
            "3758484c358f444382629747ea7f998e",
            "ed35a3babab049d6b4153c224b0475c3",
            "4cb48f81521a4047b8a4fd00cefadcc5",
            "ac763f13a48d46028f989fba83402b22",
            "6635573513fa4ef0ad38840dea43c066",
            "fe203bbf3df4445eb657d22b4507fb41",
            "75b45012a2b14af98e69ecbbf0bf6a6f",
            "d54c7d9641904136aa44c43cee575fb4",
            "5ae23f3db17445e19d562d8f09ceae41",
            "c939eb235a104687bf70c140359fa7cc",
            "7673334c0d3c417888fabcfa9f55aedf",
            "5b938c57fb364adda61d8e2482238de5",
            "e5d458db2aac4a11a13f1c934b1e8dcf",
            "8cd4bca9d48b4f8f98369a2dd5adf17b",
            "a66e9995344648e5b77b0674156cc2fb",
            "3a1a1947fdd54457a59da862268ba6fd",
            "3fd30ba8c8534b5dbc29c7396597be95",
            "ea28a193d2024bff8463cbb11d09d755",
            "7587cd9037644ccb8925ee0524857f21",
            "8aedcb205cb14c5ebc878e4c2689a10c",
            "f5cc2a5959464e0fb72e3b420a58e5d9",
            "21906dc6b3434bf29f62460890ec6288",
            "582f2ef1a2714f59912cd0bcdbf1bb75",
            "d176bbfd2d464d7585bb65be65c31262",
            "a94a3cec76fe47748a9a4db32b924cfa",
            "2817e9a3ac6447e8981188717e927c9d",
            "0141e4bc61c247c680fa4c05b3d04903",
            "ad5edbd6ee20493f9a8e0833cfac7db5",
            "534c09960afd40b0b3ccf4ec620aba51",
            "d7ebd277b99143fdad80e1f4473d73fb",
            "c06916c9e574464ebaa45d54b3aa2c92",
            "9f900c0c253d461eaeeaede238093346",
            "01ff09afb1af40bbb72a62826b81fc64",
            "ef1e840a9f3d4a048c95c3865a5da922",
            "c4d57890e664453aa90a45c4bbb1d17f",
            "b4e8802f985e443caee7f82f9e1d642f",
            "e1bd9242da8043cda8a465415842d182",
            "2013b5dea4794109b5d4e6932a3166a9",
            "6a954028886f40c39ebbad3d33e81036",
            "22ffba8398a048e98c2628e2770f3141",
            "5ce8a4e265374bdc87de7dc7ff0a6613",
            "d0a34c02e88a4d3fb2841e7fc696d5a0",
            "cd389c7bca4a4ad39dce662ffb8e2300",
            "02186d45c99a463e82a3439bbd1b535a",
            "41626856c41f40d58cee4207bdf9aee1"
          ]
        },
        "outputId": "092f03b1-6c67-46a7-bd2b-5a0bc11c433d",
        "collapsed": true,
        "id": "XEgDFezhuJB9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "config.json:   0%|          | 0.00/819 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "78f7e9faeb354d41b5097c1e906d8926"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer_config.json:   0%|          | 0.00/393 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "b59f144ff36c43e9a738900e9173f08b"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "e8e48ed233964d758c79f2241f2af3f4"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "ed35a3babab049d6b4153c224b0475c3"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "e5d458db2aac4a11a13f1c934b1e8dcf"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "d176bbfd2d464d7585bb65be65c31262"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/spacy/util.py:1740: UserWarning: [W111] Jupyter notebook detected: if using `prefer_gpu()` or `require_gpu()`, include it in the same cell right before `spacy.load()` to ensure that the model is loaded on the correct device. More information: http://spacy.io/usage/v3#jupyter-notebook-gpu\n",
            "  warnings.warn(Warnings.W111)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "pytorch_model.bin:   0%|          | 0.00/362M [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "c4d57890e664453aa90a45c4bbb1d17f"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor = RelationExtractor(nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, df_cluster)"
      ],
      "metadata": {
        "id": "gwtkKVFbUKrN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_1 = RelationExtractor(nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, df_cluster_1)"
      ],
      "metadata": {
        "id": "dI4iT3x9YXxm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_2 = RelationExtractor(nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, df_cluster_2)\n",
        "relation_extractor_3 = RelationExtractor(nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, df_cluster_3)\n",
        "relation_extractor_4 = RelationExtractor(nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device, df_cluster_4)"
      ],
      "metadata": {
        "id": "2X7gJ9qKaweT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##### Extractor - Cluster 0"
      ],
      "metadata": {
        "id": "pmgoLl7Bh4jd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor.filter_data()"
      ],
      "metadata": {
        "collapsed": true,
        "id": "cGhqWw8PFnB1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor.process_coreference_resolution()"
      ],
      "metadata": {
        "collapsed": true,
        "id": "CwH28lsw1t2T"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor.process_triplet_extraction()"
      ],
      "metadata": {
        "id": "JR4kILAj1wEI",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets = relation_extractor.aggregate_triplets()"
      ],
      "metadata": {
        "id": "-Z3s5x_Z1xhl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets"
      ],
      "metadata": {
        "id": "7tTPDUeTX3zT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor.write_triplets_to_file(aggregated_triplets, \"aggregated_triplets_cluster0.txt\")"
      ],
      "metadata": {
        "id": "jsCZ5y4wX2pc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##### Extractor - Cluster 1"
      ],
      "metadata": {
        "id": "ZANitfwYhuZr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_1.filter_data()"
      ],
      "metadata": {
        "id": "TO-TeBKmYaeP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_1.process_coreference_resolution()"
      ],
      "metadata": {
        "id": "3hhj-W92YekQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_1.process_triplet_extraction()"
      ],
      "metadata": {
        "id": "DGhelI76Yira"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets_1 = relation_extractor_1.aggregate_triplets()"
      ],
      "metadata": {
        "id": "3v5erAVBYl0x"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_1.write_triplets_to_file(aggregated_triplets_1, \"aggregated_triplets_cluster1.txt\")"
      ],
      "metadata": {
        "id": "CTLEQarcZBWF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##### Extractor - Cluster 2"
      ],
      "metadata": {
        "id": "mYweqwfEhygw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_2.filter_data()"
      ],
      "metadata": {
        "id": "wPRA1tkBbOuG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_2.process_coreference_resolution()"
      ],
      "metadata": {
        "id": "RhiDx9kFcDNl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_2.process_triplet_extraction()"
      ],
      "metadata": {
        "id": "5BNSBtoMcDKJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets_2 = relation_extractor_2.aggregate_triplets()"
      ],
      "metadata": {
        "id": "eYX2SiTQcDEH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_2.write_triplets_to_file(aggregated_triplets_2, \"aggregated_triplets_cluster2.txt\")"
      ],
      "metadata": {
        "id": "uTkJPYDXccSu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##### Extractor - Cluster 3"
      ],
      "metadata": {
        "id": "R-HDk6sXicSq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_3.filter_data()"
      ],
      "metadata": {
        "id": "Z0dSlXFPbOls"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_3.process_coreference_resolution()"
      ],
      "metadata": {
        "id": "zlJWF-fkccOu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_3.process_triplet_extraction()"
      ],
      "metadata": {
        "id": "qyJIhfpnccLy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets_3 = relation_extractor_3.aggregate_triplets()"
      ],
      "metadata": {
        "id": "V2vaP4vwccE6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_3.write_triplets_to_file(aggregated_triplets_3, \"aggregated_triplets_cluster3.txt\")"
      ],
      "metadata": {
        "id": "wXlUNDyycb6Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##### Extractor - Cluster 4"
      ],
      "metadata": {
        "id": "dhuj8Fj0igsF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_4.filter_data()"
      ],
      "metadata": {
        "id": "OCYlqmQMbOeK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_4.process_coreference_resolution()"
      ],
      "metadata": {
        "id": "pj4ZwHWgdwHS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_4.process_triplet_extraction()"
      ],
      "metadata": {
        "id": "jyx9y70Udv8n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "aggregated_triplets_4 = relation_extractor_4.aggregate_triplets()"
      ],
      "metadata": {
        "id": "6IONCkgyd-9_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "relation_extractor_4.write_triplets_to_file(aggregated_triplets_4, \"aggregated_triplets_cluster4.txt\")"
      ],
      "metadata": {
        "id": "g9X0M0i2cbzq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Experimental"
      ],
      "metadata": {
        "id": "IE5fRRyeQ3Ev"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "text5 = \"Jennifer Doudna discovered the CRISPR-Cas9 system. She later won a Nobel Prize.\"\n",
        "text5_a = \"Jennifer Doudna and Emmanuelle Charpentier discovered the CRISPR-Cas9 system. She later won a Nobel Prize.\""
      ],
      "metadata": {
        "id": "9UcjGxjpdnnb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "text6 = \"Chandragupta Maurya was the founder of the Maurya Empire and dynasty, based in Magadha. His rise to power began after the death of Alaxander in 323 BCE.\"\n",
        "text6_a = \"Chandragupta Maurya was the founder of the Maurya Empire and dynasty. His rise to power began after the death of Alaxander in 323 BCE.\"\n",
        "text6_b = \"Chandragupta Maurya's rise to power began after the death of Alaxander in 323 BCE.\"\n",
        "text6_c = \"Chandragupta Maurya's rise to power began in 323 BCE.\"\n",
        "text6_d = \"Chandragupta Maurya's rise to power began after he won the war.\""
      ],
      "metadata": {
        "id": "QdVthMohnKI9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "inputs = rebel_tokenizer(text5_a, return_tensors=\"pt\", max_length=1024, truncation=False).to(device)"
      ],
      "metadata": {
        "id": "DG6FamV1jqLF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "outputs = rebel_model.generate(**inputs)"
      ],
      "metadata": {
        "id": "geb3b3FC-flY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "extracted_text = rebel_tokenizer.decode(outputs[0], skip_special_tokens=True)"
      ],
      "metadata": {
        "id": "GPG-i7edq21s"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "extracted_text"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 36
        },
        "id": "dRmFkYLr-fWO",
        "outputId": "e6a1f0da-7761-4c74-c3c3-1210ba22327f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "' CRISPR-Cas9  Jennifer Doudna  discoverer or inventor  Emmanuelle Charpentier  discoverer or'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "test_str = pubmed_inference_filtered['input_coref'][1]"
      ],
      "metadata": {
        "id": "qPJQbMiG9ayu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "test_str1 = 'LSD1-mediated epigenetic modification is required for TAL1 function and hematopoiesis TAL1 is a critical transcription factor required for hematopoiesis'"
      ],
      "metadata": {
        "id": "pIH5b9uG-0g2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "extracted_triplets = []\n",
        "for sentence in test_str1.split('. '):\n",
        "  inputs = rebel_tokenizer(sentence, return_tensors=\"pt\", truncation=True, max_length=512).to(device)\n",
        "  outputs = rebel_model.generate(**inputs)\n",
        "  decoded_output = rebel_tokenizer.decode(outputs[0], skip_special_tokens=False)\n",
        "  triplets = list(set(extract_triplets(decoded_output)))\n",
        "  if len(triplets)==0:\n",
        "    continue\n",
        "  else:\n",
        "    extracted_triplets.extend(triplets)"
      ],
      "metadata": {
        "id": "jo48m6RnElCM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "extracted_triplets"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BbofpW-8ErBx",
        "outputId": "d82a6197-c479-4ee6-c6a4-f7ce58cccf5e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('TAL1', 'epigenetic modification', 'Association'),\n",
              " ('hematopoiesis', 'TAL1', 'Association'),\n",
              " ('TAL1', 'hematopoiesis', 'Association')]"
            ]
          },
          "metadata": {},
          "execution_count": 67
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "inputs"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AS43q96N-w_s",
        "outputId": "2dcf4a72-3c86-4ac9-f6b7-201001055206"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'input_ids': tensor([[    0, 10463,   495,   134,    12, 43728, 44323, 15557, 29685,    16,\n",
              "          1552,    13,   255,  2118,   134,  5043,     8,    37,  9244,  1517,\n",
              "           139,   918,   354,  1437,  1437,   255,  2118,   134,    16,    10,\n",
              "          2008, 37118,  3724,  1552,    13,    37,  9244,  1517,   139,   918,\n",
              "           354,     4,   635,     6, 32819, 13157,  1258,     9,    63,  1940,\n",
              "           747,  3315,     7,   255,  3551, 28837,     4,  5994,     8,   141,\n",
              "            63, 37118,   337,  1713,    32, 13588,   148,    37,  9244,  1517,\n",
              "           139,   918,   354,  1189,     7,    28,  4873,     4,  1398,     6,\n",
              "            52,   311,    14,   255,  2118,   134,    16,  3059,    19, 33945,\n",
              "          1264,  4410, 30298,  3175, 34003,  8200,   784,  2459,   833,    12,\n",
              "         14175,  4410, 30298,  3175,   112,    36, 10463,   495,   134,   238,\n",
              "          4979,   134, 11052, 17725, 37118,  3724,  2731, 13447,   368,    36,\n",
              "          8739,   500,  4923,   238, 33945,  1264,   263, 26799,  4360,  3175,\n",
              "           112,    36, 13022,  2562,   134,   238,     8, 33945,  1264,   263,\n",
              "         26799,  4360,  3175,   132,    11,  1437,  4270,   212, 35080, 44336,\n",
              "             8,   255,  3551, 28837,  4590,     4,    20, 46538,  8307,  5183,\n",
              "         11170,     9, 37619,   134,  1974,    41,   505,   774,    11,  2851,\n",
              "          9828,     5,   255,  2118,   134,    12, 25706, 37118,     9,   272,\n",
              "          2118,   306,  4439,  3307,     7,    10,  3553,  8307, 40314, 11269,\n",
              "           118,  3175,  9865, 23408,     4,  9870,     6,    52,  8085,    14,\n",
              "             5,   255,  2118,   134,    12, 38838, 37619,   134,     6,  7951,\n",
              "          2562,   134,     6,     8,    49, 46538,  8307,  5183,  1713,    32,\n",
              "         31481,  7223,   159,    12, 35908,   148,     5,   419, 17369,     9,\n",
              "          1437,  4270,   212, 36866, 37225,     4,  9051, 21464,    19,     5,\n",
              "          6379,  1022,     9,   255,  2118,   134,  2383,  7293, 13447,   368,\n",
              "          2632,   148, 37225,     6,   255,  2118,   134, 16521, 37619,   134,\n",
              "             7,     5, 30125,   181,   306,     4,   176, 23408,    11,  2432,\n",
              "         47314, 15253,     6,    53,    45,    11, 32691,     6, 22802,   833,\n",
              "          1437,  4270,   212, 35080, 44336,    36,   448,  3721,    43,  4590,\n",
              "             4,  3347,     6,  1481, 43612,    12, 43728,  7956,  3955,     9,\n",
              "         37619,   134,    11,   256,  3721,  4590,  4596,    11,   263, 12597,\n",
              "         21791,     9,     5,   255,  2118,   134,  1002, 10596,  7513,    30,\n",
              "          2284, 30979,   725,   246,   530,   306,    23,     5, 23408,   976,\n",
              "             4, 10623,     6,    84,   414,  1487,    14, 33945,  1264,   784,\n",
              "          2459,   833,  4410, 30298,  3175, 37619,   134,   189, 15708, 14244,\n",
              "           255,  2118,   134,    12, 43728, 37118,     8,  3608,    14,     5,\n",
              "          6878,  5746,     9,   255,  2118,   134,    12, 38838, 37619,   134,\n",
              "            73, 13022,  2562,   134,  2632,   189,  3094,     5, 23808,     9,\n",
              "          1437,  4270,   212, 36866, 37225,  1767,     4,     2]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
              "         1, 1, 1, 1]])}"
            ]
          },
          "metadata": {},
          "execution_count": 89
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "outputs = rebel_model.generate(**inputs)"
      ],
      "metadata": {
        "id": "RXe9heQz9_qB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "decoded_output = rebel_tokenizer.decode(outputs[0], skip_special_tokens=False)"
      ],
      "metadata": {
        "id": "0ZU10JR3-mEh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "decoded_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 36
        },
        "id": "azc76lLH-uOy",
        "outputId": "4ae6370d-ba75-4359-9845-9490a2800cbc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'<s><triplet> LSD1 <subj> TAL1 <obj> Association <triplet> TAL1 <subj> LSD1 <obj> Association</s>'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 97
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "triplets = set(extract_triplets(decoded_output))"
      ],
      "metadata": {
        "id": "QSEz05JJ-rcn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "triplets"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "400Vh7LR-se4",
        "outputId": "a13edaca-1adb-45ae-9511-37ad6678f56d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "set()"
            ]
          },
          "metadata": {},
          "execution_count": 75
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "perform_inference(test_str)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 287
        },
        "id": "cHVLqDOz9t37",
        "outputId": "68d28fed-6dee-4eee-89c9-f6f2ba6636ba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-70-1c44b06fc566>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mperform_inference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-64-275d6fee449a>\u001b[0m in \u001b[0;36mperform_inference\u001b[0;34m(text)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mperform_inference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m   \u001b[0minputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrebel_tokenizer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_tensors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"pt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtruncation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m512\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m   \u001b[0moutputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrebel_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m   \u001b[0mdecoded_output\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrebel_tokenizer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskip_special_tokens\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0mtriplets\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextract_triplets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdecoded_output\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py\u001b[0m in \u001b[0;36mdecorate_context\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    114\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdecorate_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mctx_factory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 116\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdecorate_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py\u001b[0m in \u001b[0;36mgenerate\u001b[0;34m(self, inputs, generation_config, logits_processor, stopping_criteria, prefix_allowed_tokens_fn, synced_gpus, assistant_model, streamer, negative_prompt_ids, negative_prompt_attention_mask, **kwargs)\u001b[0m\n\u001b[1;32m   2244\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2245\u001b[0m             \u001b[0;31m# 13. run beam sample\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2246\u001b[0;31m             result = self._beam_search(\n\u001b[0m\u001b[1;32m   2247\u001b[0m                 \u001b[0minput_ids\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2248\u001b[0m                 \u001b[0mbeam_scorer\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py\u001b[0m in \u001b[0;36m_beam_search\u001b[0;34m(self, input_ids, beam_scorer, logits_processor, stopping_criteria, generation_config, synced_gpus, **model_kwargs)\u001b[0m\n\u001b[1;32m   3453\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3454\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Unchanged original behavior\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3455\u001b[0;31m                 \u001b[0moutputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mmodel_inputs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_dict\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3456\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3457\u001b[0m             \u001b[0;31m# synced_gpus: don't waste resources running the code we don't need; kwargs must be updated before skipping\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1734\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compiled_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[misc]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1736\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m     \u001b[0;31m# torchrec tests the code consistency with the following code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1745\u001b[0m                 \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_pre_hooks\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_hooks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1746\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[0;32m-> 1747\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/models/bart/modeling_bart.py\u001b[0m in \u001b[0;36mforward\u001b[0;34m(self, input_ids, attention_mask, decoder_input_ids, decoder_attention_mask, head_mask, decoder_head_mask, cross_attn_head_mask, encoder_outputs, past_key_values, inputs_embeds, decoder_inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1640\u001b[0m                 )\n\u001b[1;32m   1641\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1642\u001b[0;31m         outputs = self.model(\n\u001b[0m\u001b[1;32m   1643\u001b[0m             \u001b[0minput_ids\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1644\u001b[0m             \u001b[0mattention_mask\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mattention_mask\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1734\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compiled_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[misc]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1736\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m     \u001b[0;31m# torchrec tests the code consistency with the following code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1745\u001b[0m                 \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_pre_hooks\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_hooks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1746\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[0;32m-> 1747\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/models/bart/modeling_bart.py\u001b[0m in \u001b[0;36mforward\u001b[0;34m(self, input_ids, attention_mask, decoder_input_ids, decoder_attention_mask, head_mask, decoder_head_mask, cross_attn_head_mask, encoder_outputs, past_key_values, inputs_embeds, decoder_inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1526\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1527\u001b[0m         \u001b[0;31m# decoder outputs consists of (dec_features, past_key_value, dec_hidden, dec_attn)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1528\u001b[0;31m         decoder_outputs = self.decoder(\n\u001b[0m\u001b[1;32m   1529\u001b[0m             \u001b[0minput_ids\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdecoder_input_ids\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1530\u001b[0m             \u001b[0mattention_mask\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdecoder_attention_mask\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1734\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compiled_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[misc]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1736\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m     \u001b[0;31m# torchrec tests the code consistency with the following code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1745\u001b[0m                 \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_pre_hooks\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_hooks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1746\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[0;32m-> 1747\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/models/bart/modeling_bart.py\u001b[0m in \u001b[0;36mforward\u001b[0;34m(self, input_ids, attention_mask, encoder_hidden_states, encoder_attention_mask, head_mask, cross_attn_head_mask, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1378\u001b[0m                 )\n\u001b[1;32m   1379\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1380\u001b[0;31m                 layer_outputs = decoder_layer(\n\u001b[0m\u001b[1;32m   1381\u001b[0m                     \u001b[0mhidden_states\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1382\u001b[0m                     \u001b[0mattention_mask\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mattention_mask\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1734\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compiled_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[misc]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1736\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m     \u001b[0;31m# torchrec tests the code consistency with the following code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1745\u001b[0m                 \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_pre_hooks\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_hooks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1746\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[0;32m-> 1747\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/transformers/models/bart/modeling_bart.py\u001b[0m in \u001b[0;36mforward\u001b[0;34m(self, hidden_states, attention_mask, encoder_hidden_states, encoder_attention_mask, layer_head_mask, cross_attn_layer_head_mask, past_key_value, output_attentions, use_cache)\u001b[0m\n\u001b[1;32m    702\u001b[0m         \u001b[0mhidden_states\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mactivation_fn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfc1\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhidden_states\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    703\u001b[0m         \u001b[0mhidden_states\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunctional\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhidden_states\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mactivation_dropout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraining\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtraining\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 704\u001b[0;31m         \u001b[0mhidden_states\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfc2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhidden_states\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    705\u001b[0m         \u001b[0mhidden_states\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunctional\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhidden_states\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraining\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtraining\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    706\u001b[0m         \u001b[0mhidden_states\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresidual\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mhidden_states\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1734\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compiled_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[misc]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1736\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_impl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1738\u001b[0m     \u001b[0;31m# torchrec tests the code consistency with the following code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1745\u001b[0m                 \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_pre_hooks\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_global_backward_hooks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1746\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[0;32m-> 1747\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1749\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/torch/nn/modules/linear.py\u001b[0m in \u001b[0;36mforward\u001b[0;34m(self, input)\u001b[0m\n\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mforward\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTensor\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mTensor\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 125\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlinear\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mweight\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbias\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    126\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mextra_repr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "K7sCq_HI7pmw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "pubmed_dataset = Dataset.from_pandas(pubmed_inference_filtered[['input_coref']])"
      ],
      "metadata": {
        "id": "6s-UY1uV9yJb"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "pubmed_dataset"
      ],
      "metadata": {
        "id": "8WyPsvW19yJc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "tokenized_pubmed_dataset = pubmed_dataset.map(preprocess_test_data, batched=True)"
      ],
      "metadata": {
        "id": "vFPWq0l09yJc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "class RelationExtractor:\n",
        "    def __init__(self, nlp_scispacy, fcoref_model, rebel_model, rebel_tokenizer, device):\n",
        "        \"\"\"\n",
        "        Initialize the RelationExtractor class with models for coreference, triplet extraction, and data handling.\n",
        "        \"\"\"\n",
        "        self.nlp_scispacy = nlp_scispacy\n",
        "        self.fcoref_model = fcoref_model\n",
        "        self.rebel_model = rebel_model\n",
        "        self.rebel_tokenizer = rebel_tokenizer\n",
        "        self.device = device\n",
        "        self.data = None\n",
        "\n",
        "    def filter_text(self, text):\n",
        "        \"\"\"\n",
        "        Filter and clean the abstract text.\n",
        "        \"\"\"\n",
        "        try:\n",
        "            if type(text) != str:\n",
        "                return ''\n",
        "            if 'abstract' in text:\n",
        "                return text.replace('abstract ', '').strip()\n",
        "            return text.strip()\n",
        "        except Exception as e:\n",
        "            print(f'Error filtering text: {e}')\n",
        "            return ''\n",
        "\n",
        "    ### Data Reading Methods ###\n",
        "    def load_data(self, file_path):\n",
        "        \"\"\"\n",
        "        Load and preprocess data from a JSON file.\n",
        "        \"\"\"\n",
        "        pubmed_inference_data = pd.read_json(file_path)\n",
        "        pubmed_inference_filtered = pubmed_inference_data[['title', 'abstract']]\n",
        "        pubmed_inference_filtered['abstract_updated'] = pubmed_inference_filtered['abstract'].apply(self.filter_text)\n",
        "        pubmed_inference_filtered['input'] = (\n",
        "            pubmed_inference_filtered['title'] + \" \" + pubmed_inference_filtered['abstract_updated']\n",
        "        )\n",
        "        self.data = pubmed_inference_filtered\n",
        "        return pubmed_inference_filtered\n",
        "\n",
        "    ### Coreference Resolution Methods ###\n",
        "    def core_logic_part(self, document: Doc, coref: list[int], resolved: list[str], mention_span: Span):\n",
        "        final_token = document[coref[1]]\n",
        "        if final_token.tag_ in [\"PRP$\", \"POS\"]:\n",
        "            resolved[coref[0]] = mention_span.text + \"'s\" + final_token.whitespace_\n",
        "        else:\n",
        "            resolved[coref[0]] = mention_span.text + final_token.whitespace_\n",
        "        for i in range(coref[0] + 1, coref[1] + 1):\n",
        "            resolved[i] = \"\"\n",
        "        return resolved\n",
        "\n",
        "    def get_span_noun_indices(self, doc: Doc, cluster: list[list[int]]) -> list[int]:\n",
        "        try:\n",
        "            spans = [doc[span[0]:span[1] + 1] for span in cluster]\n",
        "            spans_pos = [[token.pos_ for token in span] for span in spans]\n",
        "            span_noun_indices = [\n",
        "                i for i, span_pos in enumerate(spans_pos)\n",
        "                if any(pos in span_pos for pos in ['NOUN', 'PROPN'])\n",
        "            ]\n",
        "            return span_noun_indices\n",
        "        except Exception as e:\n",
        "            print(f'Error while getting span noun indices: {e}')\n",
        "            return []\n",
        "\n",
        "    def get_cluster_head(self, doc: Doc, cluster: list[list[int]], noun_indices: list[int]):\n",
        "        head_idx = noun_indices[0]\n",
        "        head_start, head_end = cluster[head_idx]\n",
        "        head_span = doc[head_start:head_end + 1]\n",
        "        return head_span, [head_start, head_end]\n",
        "\n",
        "    def is_containing_other_spans(self, span: list[int], all_spans: list[list[int]]):\n",
        "        return any([s[0] >= span[0] and s[1] <= span[1] and s != span for s in all_spans])\n",
        "\n",
        "    def improved_replace_corefs(self, document, clusters):\n",
        "        resolved = list(tok.text_with_ws for tok in document)\n",
        "        all_spans = [span for cluster in clusters for span in cluster]\n",
        "\n",
        "        for cluster in clusters:\n",
        "            noun_indices = self.get_span_noun_indices(document, cluster)\n",
        "\n",
        "            if noun_indices:\n",
        "                mention_span, mention = self.get_cluster_head(document, cluster, noun_indices)\n",
        "\n",
        "                for coref in cluster:\n",
        "                    if coref != mention and not self.is_containing_other_spans(coref, all_spans):\n",
        "                        self.core_logic_part(document, coref, resolved, mention_span)\n",
        "\n",
        "        return \"\".join(resolved)\n",
        "\n",
        "    def get_fast_cluster_spans(self, doc, clusters):\n",
        "        fast_clusters = []\n",
        "        for cluster in clusters:\n",
        "            new_group = []\n",
        "            for span_tuple in cluster:\n",
        "                (start, end) = span_tuple\n",
        "                span = doc.char_span(start, end)\n",
        "                if span:\n",
        "                    new_group.append([span.start, span.end - 1])\n",
        "            fast_clusters.append(new_group)\n",
        "        return fast_clusters\n",
        "\n",
        "    def get_fastcoref_clusters(self, doc, text):\n",
        "        try:\n",
        "            preds = self.fcoref_model.predict(texts=[text])\n",
        "            fast_clusters = preds[0].get_clusters(as_strings=False)\n",
        "            fast_cluster_spans = self.get_fast_cluster_spans(doc, fast_clusters)\n",
        "        except Exception as e:\n",
        "            print(f'Error while getting coreference clusters: {e}')\n",
        "            return text\n",
        "        return fast_cluster_spans\n",
        "\n",
        "    def perform_coreference_resolution(self, text):\n",
        "        \"\"\"\n",
        "        Perform coreference resolution on the input text.\n",
        "        \"\"\"\n",
        "        doc = self.nlp_scispacy(text)\n",
        "        clusters = self.get_fastcoref_clusters(doc, text)\n",
        "        coref_text = self.improved_replace_corefs(doc, clusters)\n",
        "        return coref_text\n",
        "\n",
        "    ### Triplet Extraction Methods ###\n",
        "    def extract_triplets(self, decoded_string):\n",
        "        \"\"\"\n",
        "        Extract triplets from the decoded string.\n",
        "        \"\"\"\n",
        "        decoded_str_copy = decoded_string\n",
        "        if '<s>' in decoded_string:\n",
        "            decoded_str_copy = ''.join(decoded_string.split('<s>')[1:])\n",
        "            decoded_str_copy = decoded_str_copy.split('</s>')[0]\n",
        "\n",
        "        triplets = []\n",
        "        parts = decoded_str_copy.split(\"<triplet>\")\n",
        "\n",
        "        for part in parts[1:]:\n",
        "            head, *relations = part.split(\"<subj>\")\n",
        "            head = head.strip()\n",
        "\n",
        "            for relation in relations:\n",
        "                if \"<obj>\" in relation:\n",
        "                    subj, obj = relation.split(\"<obj>\", 1)\n",
        "                    triplets.append((head, subj.strip(), obj.strip()))\n",
        "\n",
        "        return triplets\n",
        "\n",
        "    def perform_inference(self, text):\n",
        "        \"\"\"\n",
        "        Perform inference to extract triplets from the input text.\n",
        "        \"\"\"\n",
        "        extracted_triplets = []\n",
        "        for sentence in text.split('. '):\n",
        "            inputs = self.rebel_tokenizer(\n",
        "                sentence, return_tensors=\"pt\", truncation=True, max_length=512\n",
        "            ).to(self.device)\n",
        "            outputs = self.rebel_model.generate(**inputs)\n",
        "            decoded_output = self.rebel_tokenizer.decode(outputs[0], skip_special_tokens=False)\n",
        "            triplets = list(set(self.extract_triplets(decoded_output)))\n",
        "            if triplets:\n",
        "                extracted_triplets.extend(triplets)\n",
        "\n",
        "        return extracted_triplets\n",
        "\n",
        "    def aggregate_triplets(self, input_texts):\n",
        "        \"\"\"\n",
        "        Aggregate triplets for a collection of texts.\n",
        "        \"\"\"\n",
        "        aggregated_triplets = []\n",
        "        \n",
        "        for text in input_texts:\n",
        "            coref_text = self.perform_coreference_resolution(text)\n",
        "            triplets = self.perform_inference(coref_text)\n",
        "            aggregated_triplets.extend(triplets)\n",
        "        return list(set(aggregated_triplets))\n",
        "\n",
        "    def write_triplets_to_file(self, aggregated_triplets, file_path):\n",
        "        \"\"\"\n",
        "        Write aggregated triplets to a text file.\n",
        "        \"\"\"\n",
        "        with open(file_path, \"w\") as f:\n",
        "            for triplet in aggregated_triplets:\n",
        "                f.write(f\"{triplet}\\n\")\n"
      ],
      "metadata": {
        "id": "HeCU2hrSUK14"
      }
    }
  ]
}